From 1629c0138c8292fd19b3be86def3f125e875341a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Mon, 18 Oct 2021 12:44:17 +0200
Subject: [PATCH 03/12] user32: Report at least one monitor in
 QueryDisplayConfig.

---
 dlls/user32/sysparams.c | 31 ++++++++++++++++++++++++++++---
 1 file changed, 28 insertions(+), 3 deletions(-)

diff --git a/dlls/user32/sysparams.c b/dlls/user32/sysparams.c
index 11111111111..11111111111 100644
--- a/dlls/user32/sysparams.c
+++ b/dlls/user32/sysparams.c
@@ -1317,7 +1317,8 @@ static LONG query_display_device_modes( const WCHAR *device_name, UINT32 output_
     if (++mode_index == *mode_count) return -1;
 
     /* Extract the adapter index from device_name to use as the source ID */
-    adapter_index = wcstol( device_name + lstrlenW( L"\\\\.\\DISPLAY" ), NULL, 10 ) - 1;
+    if (!device_name) adapter_index = 0;
+    else adapter_index = wcstol( device_name + lstrlenW( L"\\\\.\\DISPLAY" ), NULL, 10 ) - 1;
 
     /* Multiple targets can be driven by the same source, ensure a mode
      * hasn't already been added for this source.
@@ -1347,8 +1348,8 @@ LONG WINAPI QueryDisplayConfig(UINT32 flags, UINT32 *numpathelements, DISPLAYCON
     HDEVINFO devinfo;
     SP_DEVINFO_DATA device_data = {sizeof(device_data)};
     DWORD monitor_index = 0, state_flags, type;
-    UINT32 output_id, path_index = 0, mode_index = 0;
-    LUID gpu_luid;
+    UINT32 output_id = 0, path_index = 0, mode_index = 0;
+    LUID gpu_luid = {0};
     WCHAR device_name[CCHDEVICENAME];
     DEVMODEW devmode;
     POINT origin;
@@ -1485,6 +1486,22 @@ LONG WINAPI QueryDisplayConfig(UINT32 flags, UINT32 *numpathelements, DISPLAYCON
         mode_index += ret;
     }
 
+    /* always report at least one monitor */
+    if (monitor_index == 1)
+    {
+        ret = query_display_device_modes(NULL, output_id, &gpu_luid, flags,
+                                         path_index, numpathelements, pathinfo,
+                                         mode_index, numinfoelements, modeinfo);
+        if (ret < 0)
+        {
+            ret = ERROR_INSUFFICIENT_BUFFER;
+            goto done;
+        }
+
+        path_index += 1;
+        mode_index += ret;
+    }
+
     *numpathelements = path_index;
     *numinfoelements = mode_index;
     ret = ERROR_SUCCESS;
@@ -1559,6 +1576,14 @@ LONG WINAPI DisplayConfigGetDeviceInfo(DISPLAYCONFIG_DEVICE_INFO_HEADER *packet)
             break;
         }
         SetupDiDestroyDeviceInfoList(devinfo);
+        /* always report at least one monitor */
+        if (index == 1 && source_name->header.id == 0 &&
+            source_name->header.adapterId.LowPart == 0 &&
+            source_name->header.adapterId.HighPart == 0)
+        {
+            wcscpy(source_name->viewGdiDeviceName, L"\\\\.\\DISPLAY1");
+            ret = ERROR_SUCCESS;
+        }
         release_display_device_init_mutex(mutex);
         return ret;
     }
-- 
2.36.1

