From 124666490cd070dc68500cac28ff126533515348 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Mon, 19 Apr 2021 09:48:16 +0200
Subject: [PATCH 10/12] server: Update desktop cursor window in set_window_pos.

---
 server/queue.c  | 5 +++++
 server/user.h   | 1 +
 server/window.c | 2 ++
 3 files changed, 8 insertions(+)

diff --git a/server/queue.c b/server/queue.c
index 11111111111..11111111111 100644
--- a/server/queue.c
+++ b/server/queue.c
@@ -536,6 +536,11 @@ static int update_desktop_cursor_pos( struct desktop *desktop, user_handle_t win
     return updated;
 }
 
+void update_desktop_cursor_win( struct desktop *desktop )
+{
+    update_desktop_cursor_pos( desktop, 0, desktop->shared->cursor.x, desktop->shared->cursor.y );
+}
+
 /* set the cursor position and queue the corresponding mouse message */
 static void set_cursor_pos( struct desktop *desktop, int x, int y )
 {
diff --git a/server/user.h b/server/user.h
index 11111111111..11111111111 100644
--- a/server/user.h
+++ b/server/user.h
@@ -187,6 +187,7 @@ extern void set_process_default_desktop( struct process *process, struct desktop
 extern void close_process_desktop( struct process *process );
 extern void set_thread_default_desktop( struct thread *thread, struct desktop *desktop, obj_handle_t handle );
 extern void release_thread_desktop( struct thread *thread, int close );
+extern void update_desktop_cursor_win( struct desktop *desktop );
 
 static inline int is_rect_empty( const rectangle_t *rect )
 {
diff --git a/server/window.c b/server/window.c
index 11111111111..11111111111 100644
--- a/server/window.c
+++ b/server/window.c
@@ -2482,6 +2482,8 @@ DECL_HANDLER(set_window_pos)
         reply->surface_win = top->handle;
         reply->needs_update = !!(top->paint_flags & (PAINT_HAS_PIXEL_FORMAT | PAINT_PIXEL_FORMAT_CHILD));
     }
+
+    update_desktop_cursor_win( win->desktop );
 }
 
 
-- 
2.37.1

