From 51fb6091eccfd9a88381a7c7adfa4242ecd60d31 Mon Sep 17 00:00:00 2001
From: Huw Davies <huw@codeweavers.com>
Date: Mon, 29 Jun 2020 11:12:12 +0100
Subject: [PATCH 02/15] server: Use the helper to update the cursor last change
 time.

Signed-off-by: Huw Davies <huw@codeweavers.com>
---
 server/queue.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/server/queue.c b/server/queue.c
index 11144941461..fb82bcf5e86 100644
--- a/server/queue.c
+++ b/server/queue.c
@@ -1814,7 +1814,7 @@ static int queue_mouse_message( struct desktop *desktop, user_handle_t win, cons
         WM_MOUSEHWHEEL   /* 0x1000 = MOUSEEVENTF_HWHEEL */
     };
 
-    desktop->cursor.last_change = get_tick_count();
+    update_desktop_cursor_pos( desktop, 0, desktop->cursor.x, desktop->cursor.y ); /* Update last change time */
     flags = input->mouse.flags;
     time  = input->mouse.time;
     if (!time) time = desktop->cursor.last_change;
