From 28027dafc55eb1abeabba070c00a7dc91b1fdec9 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Thu, 23 Jan 2020 14:19:27 +0100
Subject: [PATCH 10/16] user32/tests: Add broken w1064v1809 mouse input test
 result.

---
 dlls/user32/tests/win.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/user32/tests/win.c b/dlls/user32/tests/win.c
index 11111111111..11111111111 100644
--- a/dlls/user32/tests/win.c
+++ b/dlls/user32/tests/win.c
@@ -4663,6 +4663,8 @@ static void test_mouse_input(void)
        msg.hwnd, popup, msg.message);
 
     ret = wait_for_message( &msg );
+    if (broken(msg.message >= WM_USER) /* on w1064v1809 */)
+        ret = wait_for_message( &msg );
     ok(ret, "no message available\n");
     ok(msg.hwnd == popup && msg.message == WM_LBUTTONUP, "hwnd %p/%p message %04x\n",
        msg.hwnd, popup, msg.message);
-- 
0.0.0

