From 59728636a399e038e03c477adca7ec9aa596b5b6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Fri, 7 May 2021 15:34:56 +0200
Subject: [PATCH 15/16] user32/tests: Add some more layered window tests.

---
 dlls/user32/tests/win.c | 61 +++++++++++++++++++++++++++++++++++++++++
 1 file changed, 61 insertions(+)

diff --git a/dlls/user32/tests/win.c b/dlls/user32/tests/win.c
index 11111111111..11111111111 100644
--- a/dlls/user32/tests/win.c
+++ b/dlls/user32/tests/win.c
@@ -9182,6 +9182,67 @@ static void test_layered_window(void)
     hbm = CreateCompatibleBitmap( hdc, 200, 200 );
     SelectObject( hdc, hbm );
 
+    hwnd = CreateWindowExA( 0, "MainWindowClass", "message window", WS_CAPTION | WS_VISIBLE,
+                            100, 100, 200, 200, 0, 0, 0, NULL );
+    assert( hwnd );
+    flush_events( TRUE );
+
+    /* already visible window has some layered attributes but UpdateLayeredWindow succeeds */
+    SetWindowLongA( hwnd, GWL_EXSTYLE, GetWindowLongA(hwnd, GWL_EXSTYLE) | WS_EX_LAYERED );
+    ret = pGetLayeredWindowAttributes( hwnd, &key, &alpha, &flags );
+    todo_wine ok( ret, "GetLayeredWindowAttributes should succeed on layered visible window\n" );
+    ok( key == 0, "wrong color key %lx\n", key );
+    ok( alpha == 0, "wrong alpha %u\n", alpha );
+    ok( flags == 0, "wrong flags %lx\n", flags );
+    ret = pUpdateLayeredWindow( hwnd, 0, NULL, &sz, hdc, &pt, 0, NULL, ULW_OPAQUE );
+    ok( ret, "UpdateLayeredWindow should succeed on layered visible window\n" );
+
+    /* even after resetting the style */
+    SetWindowLongA( hwnd, GWL_EXSTYLE, GetWindowLongA(hwnd, GWL_EXSTYLE) & ~WS_EX_LAYERED );
+    SetWindowLongA( hwnd, GWL_EXSTYLE, GetWindowLongA(hwnd, GWL_EXSTYLE) | WS_EX_LAYERED );
+    ret = pGetLayeredWindowAttributes( hwnd, &key, &alpha, &flags );
+    todo_wine ok( ret, "GetLayeredWindowAttributes should succeed on layered visible window\n" );
+    ok( key == 0, "wrong color key %lx\n", key );
+    todo_wine ok( alpha == 0xff, "wrong alpha %u\n", alpha );
+    todo_wine ok( flags == 2, "wrong flags %lx\n", flags );
+    ret = pUpdateLayeredWindow( hwnd, 0, NULL, &sz, hdc, &pt, 0, NULL, ULW_OPAQUE );
+    ok( ret, "UpdateLayeredWindow should succeed on layered visible window\n" );
+
+    /* hiding the window before setting layered style is okay */
+    SetWindowLongA( hwnd, GWL_EXSTYLE, GetWindowLongA(hwnd, GWL_EXSTYLE) & ~WS_EX_LAYERED );
+    ShowWindow( hwnd, SW_HIDE );
+    flush_events( TRUE );
+    SetWindowLongA( hwnd, GWL_EXSTYLE, GetWindowLongA(hwnd, GWL_EXSTYLE) | WS_EX_LAYERED );
+    ret = pGetLayeredWindowAttributes( hwnd, &key, &alpha, &flags );
+    ok( !ret, "GetLayeredWindowAttributes should fail on layered visible window\n" );
+    ret = pUpdateLayeredWindow( hwnd, 0, NULL, &sz, hdc, &pt, 0, NULL, ULW_OPAQUE );
+    ok( ret, "UpdateLayeredWindow should succeed on layered hidden window\n" );
+
+    /* showing the window after setting layered style is okay */
+    SetWindowLongA( hwnd, GWL_EXSTYLE, GetWindowLongA(hwnd, GWL_EXSTYLE) & ~WS_EX_LAYERED );
+    SetWindowLongA( hwnd, GWL_EXSTYLE, GetWindowLongA(hwnd, GWL_EXSTYLE) | WS_EX_LAYERED );
+    ShowWindow( hwnd, SW_SHOW );
+    flush_events( TRUE );
+    ret = pGetLayeredWindowAttributes( hwnd, &key, &alpha, &flags );
+    ok( !ret, "GetLayeredWindowAttributes should fail on layered visible window\n" );
+    ret = pUpdateLayeredWindow( hwnd, 0, NULL, &sz, hdc, &pt, 0, NULL, ULW_OPAQUE );
+    ok( ret, "UpdateLayeredWindow should succeed on layered visible window\n" );
+
+    /* but hiding the window after setting layered style is not */
+    SetWindowLongA( hwnd, GWL_EXSTYLE, GetWindowLongA(hwnd, GWL_EXSTYLE) & ~WS_EX_LAYERED );
+    SetWindowLongA( hwnd, GWL_EXSTYLE, GetWindowLongA(hwnd, GWL_EXSTYLE) | WS_EX_LAYERED );
+    ShowWindow( hwnd, SW_HIDE );
+    flush_events( TRUE );
+    ret = pUpdateLayeredWindow( hwnd, 0, NULL, &sz, hdc, &pt, 0, NULL, ULW_OPAQUE );
+    todo_wine ok( !ret, "UpdateLayeredWindow should fail on layered hidden window\n" );
+
+    ShowWindow( hwnd, SW_SHOW );
+    flush_events( TRUE );
+    ret = pUpdateLayeredWindow( hwnd, 0, NULL, &sz, hdc, &pt, 0, NULL, ULW_OPAQUE );
+    todo_wine ok( !ret, "UpdateLayeredWindow should succeed on layered visible window\n" );
+
+    DestroyWindow( hwnd );
+
     hwnd = CreateWindowExA(0, "MainWindowClass", "message window", WS_CAPTION,
                            100, 100, 200, 200, 0, 0, 0, NULL);
     assert( hwnd );
-- 
2.40.0

