From e68fb69642669fe3bcd35f6854eeb844c0cb7b4d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Wed, 12 May 2021 15:37:00 +0200
Subject: [PATCH 13/13] winex11.drv: Do not enable/disable XInput2 when
 clipping.

---
 dlls/winex11.drv/mouse.c | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/dlls/winex11.drv/mouse.c b/dlls/winex11.drv/mouse.c
index 3eb3c7ee475..1bc874d678a 100644
--- a/dlls/winex11.drv/mouse.c
+++ b/dlls/winex11.drv/mouse.c
@@ -507,9 +507,6 @@ static BOOL grab_clipping_window( const RECT *clip )
         return FALSE;
     }
 
-    /* enable XInput2 unless we are already clipping */
-    if (!clipping_cursor) x11drv_xinput_enable( data->display, DefaultRootWindow( data->display ), PointerMotionMask );
-
     if (data->xi2_state != xi_enabled)
     {
         WARN( "XInput2 not supported, refusing to clip to %s\n", wine_dbgstr_rect(clip) );
@@ -542,7 +539,6 @@ static BOOL grab_clipping_window( const RECT *clip )
 
     if (!clipping_cursor)
     {
-        x11drv_xinput_disable( data->display, DefaultRootWindow( data->display ), PointerMotionMask );
         XUnmapWindow( data->display, clip_window );
         return FALSE;
     }
@@ -568,7 +564,6 @@ void ungrab_clipping_window(void)
     if (clipping_cursor) XUngrabPointer( data->display, CurrentTime );
     clipping_cursor = FALSE;
     data->clip_reset = GetTickCount();
-    x11drv_xinput_disable( data->display, DefaultRootWindow( data->display ), PointerMotionMask );
 }
 
 /***********************************************************************
