From df186688d9a524db219e5f906b97cfa8846d9eeb Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Mon, 25 Jan 2021 15:16:19 +0100
Subject: [PATCH 8/9] server: Update desktop cursor window on cursor move.

---
 server/queue.c | 38 +++++++++++++++++++++++++-------------
 1 file changed, 25 insertions(+), 13 deletions(-)

diff --git a/server/queue.c b/server/queue.c
index 1fa0fb20b1a..97464ea8c52 100644
--- a/server/queue.c
+++ b/server/queue.c
@@ -371,8 +371,10 @@ static struct message *alloc_hardware_message( lparam_t info, struct hw_msg_sour
     return msg;
 }
 
-static int update_desktop_cursor_pos( struct desktop *desktop, int x, int y )
+static int update_desktop_cursor_pos( struct desktop *desktop, user_handle_t win, int x, int y )
 {
+    struct thread_input *input;
+    struct thread *thread;
     int updated;
 
     x = max( min( x, desktop->cursor.clip.right - 1 ), desktop->cursor.clip.left );
@@ -382,6 +384,21 @@ static int update_desktop_cursor_pos( struct desktop *desktop, int x, int y )
     desktop->cursor.y = y;
     desktop->cursor.last_change = get_tick_count();
 
+    if (win && (thread = get_window_thread( win )))
+    {
+        input = thread->queue->input;
+        release_object( thread );
+    }
+    else input = desktop->foreground_input;
+
+    if (input && input->capture)
+        win = input->capture;
+    else if (!win || !is_window_visible( win ) || is_window_transparent( win ))
+        win = shallow_window_from_point( desktop, x, y );
+
+    if (win != desktop->cursor.win) updated = 1;
+    desktop->cursor.win = win;
+
     return updated;
 }
 
@@ -394,7 +411,7 @@ static void set_cursor_pos( struct desktop *desktop, int x, int y )
 
     if ((device = current->process->rawinput_mouse) && (device->flags & RIDEV_NOLEGACY))
     {
-        update_desktop_cursor_pos( desktop, x, y );
+        update_desktop_cursor_pos( desktop, 0, x, y );
         return;
     }
 
@@ -1366,10 +1383,10 @@ static void update_input_key_state( struct desktop *desktop, unsigned char *keys
 
 /* update the desktop key state according to a mouse message flags */
 static void update_desktop_mouse_state( struct desktop *desktop, unsigned int flags,
-                                        int x, int y, lparam_t wparam )
+                                        user_handle_t win, int x, int y, lparam_t wparam )
 {
     if (flags & MOUSEEVENTF_MOVE)
-        update_desktop_cursor_pos( desktop, x, y );
+        update_desktop_cursor_pos( desktop, win, x, y );
     if (flags & MOUSEEVENTF_LEFTDOWN)
         update_input_key_state( desktop, desktop->keystate, WM_LBUTTONDOWN, wparam );
     if (flags & MOUSEEVENTF_LEFTUP)
@@ -1476,11 +1493,9 @@ static user_handle_t find_hardware_message_window( struct desktop *desktop, stru
             if (*msg_code < WM_SYSKEYDOWN) *msg_code += WM_SYSKEYDOWN - WM_KEYDOWN;
         }
     }
-    else if (!input || !(win = input->capture)) /* mouse message */
+    else /* mouse message */
     {
-        if (is_window_visible( msg->win ) && !is_window_transparent( msg->win )) win = msg->win;
-        else win = shallow_window_from_point( desktop, msg->x, msg->y );
-
+        win = desktop->cursor.win;
         *thread = window_thread_from_point( win, msg->x, msg->y );
     }
 
@@ -1558,7 +1573,7 @@ static void queue_hardware_message( struct desktop *desktop, struct message *msg
         if (msg->msg == WM_MOUSEMOVE)
         {
             prepend_cursor_history( msg->x, msg->y, msg->time, msg_data->info );
-            if (update_desktop_cursor_pos( desktop, msg->x, msg->y )) always_queue = 1;
+            if (update_desktop_cursor_pos( desktop, msg->win, msg->x, msg->y )) always_queue = 1;
         }
         if (desktop->keystate[VK_LBUTTON] & 0x80)  msg->wparam |= MK_LBUTTON;
         if (desktop->keystate[VK_MBUTTON] & 0x80)  msg->wparam |= MK_MBUTTON;
@@ -1587,9 +1602,6 @@ static void queue_hardware_message( struct desktop *desktop, struct message *msg
     }
     input = thread->queue->input;
 
-    if (win != desktop->cursor.win) always_queue = 1;
-    desktop->cursor.win = win;
-
     if (!always_queue || merge_message( input, msg )) free_message( msg );
     else
     {
@@ -1811,7 +1823,7 @@ static int queue_mouse_message( struct desktop *desktop, user_handle_t win, cons
 
     if ((device = current->process->rawinput_mouse) && (device->flags & RIDEV_NOLEGACY))
     {
-        update_desktop_mouse_state( desktop, flags, x, y, input->mouse.data << 16 );
+        update_desktop_mouse_state( desktop, flags, win, x, y, input->mouse.data << 16 );
         return 0;
     }
 
