From 853d778a54b6126381ab1c07a02d0cd472850143 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Sun, 14 Nov 2021 14:38:45 +0100
Subject: [PATCH 04/12] winex11.drv: Reset XInput2 valuators only on
 DeviceChange events.

---
 dlls/winex11.drv/mouse.c | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/dlls/winex11.drv/mouse.c b/dlls/winex11.drv/mouse.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/mouse.c
+++ b/dlls/winex11.drv/mouse.c
@@ -466,10 +466,6 @@ void x11drv_xinput_disable( Display *display, Window window, long event_mask )
     pXISelectEvents( display, DefaultRootWindow( display ), &mask, 1 );
 
     if (!data) return;
-    data->x_valuator.number = -1;
-    data->y_valuator.number = -1;
-    data->x_valuator.value = 0;
-    data->y_valuator.value = 0;
     data->xi2_core_pointer = 0;
     data->xi2_state = xi_disabled;
 #endif
-- 
2.35.1

