From: "André Hentschel" <nerv@dawncrow.de>
Subject: [PATCH 3/8] kernelbase: Fix some spec file entries
Message-Id: <5b7ddae2-ff87-f765-9e18-f9aa507967c0@dawncrow.de>
Date: Sun, 19 Jan 2020 22:51:49 +0100

Signed-off-by: André Hentschel <nerv@dawncrow.de>
---
 dlls/kernelbase/kernelbase.spec | 20 ++++++++++----------
 1 file changed, 10 insertions(+), 10 deletions(-)

diff --git a/dlls/kernelbase/kernelbase.spec b/dlls/kernelbase/kernelbase.spec
index 11111111111..11111111111 100644
--- a/dlls/kernelbase/kernelbase.spec
+++ b/dlls/kernelbase/kernelbase.spec
@@ -215,7 +215,7 @@
 @ stdcall CreateProcessInternalW(long wstr wstr ptr ptr long long ptr wstr ptr ptr ptr)
 @ stdcall CreateProcessW(wstr wstr ptr ptr long long ptr wstr ptr ptr)
 @ stdcall CreatePseudoConsole(long long long long ptr)
-@ stdcall CreateRemoteThread(long ptr long ptr long long ptr)
+@ stdcall CreateRemoteThread(long ptr long ptr ptr long ptr)
 @ stdcall CreateRemoteThreadEx(long ptr long ptr ptr long ptr ptr)
 @ stdcall CreateRestrictedToken(long long long ptr long ptr long ptr ptr)
 @ stdcall CreateSemaphoreExW(ptr long long wstr long long)
@@ -226,7 +226,7 @@
 # @ stub CreateStateLock
 # @ stub CreateStateSubcontainer
 @ stdcall CreateSymbolicLinkW(wstr wstr long)
-@ stdcall CreateThread(ptr long ptr long long ptr)
+@ stdcall CreateThread(ptr long ptr ptr long ptr)
 @ stdcall CreateThreadpool(ptr)
 @ stdcall CreateThreadpoolCleanupGroup()
 @ stdcall CreateThreadpoolIo(ptr ptr ptr ptr)
@@ -378,7 +378,7 @@
 @ stdcall FindNextVolumeW(long ptr long)
 # @ stub FindPackagesByPackageFamily
 @ stdcall FindResourceExW(long wstr wstr long)
-@ stdcall FindResourceW(long wstr wstr)
+@ stdcall FindResourceW(ptr wstr wstr)
 @ stdcall FindStringOrdinal(long wstr long wstr long long)
 @ stdcall FindVolumeClose(ptr)
 @ stdcall FlsAlloc(ptr)
@@ -387,7 +387,7 @@
 @ stdcall FlsSetValue(long ptr)
 @ stdcall FlushConsoleInputBuffer(long)
 @ stdcall FlushFileBuffers(long)
-@ stdcall FlushInstructionCache(long long long)
+@ stdcall FlushInstructionCache(long ptr long)
 @ stdcall FlushProcessWriteBuffers() ntdll.NtFlushProcessWriteBuffers
 @ stdcall FlushViewOfFile(ptr long)
 @ stdcall FoldStringW(long wstr long ptr long)
@@ -400,7 +400,7 @@
 @ stdcall FreeEnvironmentStringsW(ptr)
 # @ stub FreeGPOListInternalA
 # @ stub FreeGPOListInternalW
-@ stdcall FreeLibrary(long)
+@ stdcall FreeLibrary(ptr)
 @ stdcall FreeLibraryAndExitThread(long long)
 @ stdcall FreeLibraryWhenCallbackReturns(ptr ptr) ntdll.TpCallbackUnloadDllOnCompletion
 @ stdcall FreeResource(long)
@@ -799,7 +799,7 @@
 @ stdcall GetWsChangesEx(long ptr ptr)
 @ stdcall -arch=i386,x86_64 GetXStateFeaturesMask(ptr ptr)
 @ stdcall GlobalAlloc(long long)
-@ stdcall GlobalFree(long)
+@ stdcall GlobalFree(ptr)
 @ stdcall GlobalMemoryStatusEx(ptr)
 # @ stub GuardCheckLongJumpTarget
 # @ stub HasPolicyForegroundProcessingCompletedInternal
@@ -962,7 +962,7 @@
 @ stdcall LoadLibraryExW(wstr long long)
 @ stdcall LoadLibraryW(wstr)
 @ stdcall LoadPackagedLibrary(wstr long)
-@ stdcall LoadResource(long long)
+@ stdcall LoadResource(ptr long)
 @ stdcall LoadStringA(long long ptr long)
 @ stub LoadStringBaseExW
 @ stub LoadStringByReference
@@ -1421,8 +1421,8 @@
 @ stdcall SHTruncateString(str long)
 # @ stub SaveAlternatePackageRootPath
 # @ stub SaveStateRootFolderPath
-@ stdcall ScrollConsoleScreenBufferA(long ptr ptr ptr ptr)
-@ stdcall ScrollConsoleScreenBufferW(long ptr ptr ptr ptr)
+@ stdcall ScrollConsoleScreenBufferA(long ptr ptr long ptr)
+@ stdcall ScrollConsoleScreenBufferW(long ptr ptr long ptr)
 @ stdcall SearchPathA(str str str long ptr ptr)
 @ stdcall SearchPathW(wstr wstr wstr long ptr ptr)
 @ stdcall SetAclInformation(ptr ptr long long)
@@ -1552,7 +1552,7 @@
 @ stdcall SetupComm(long long long)
 # @ stub SharedLocalIsEnabled
 @ stdcall SignalObjectAndWait(long long long long)
-@ stdcall SizeofResource(long long)
+@ stdcall SizeofResource(ptr long)
 @ stdcall Sleep(long)
 @ stdcall SleepConditionVariableCS(ptr ptr long)
 @ stdcall SleepConditionVariableSRW(ptr ptr long long)
