From: "André Hentschel" <nerv@dawncrow.de>
Subject: [PATCH 8/8] ntdll: Fix some spec file entries
Message-Id: <8f3ee54c-0a7e-a654-0cf5-1d1edcb1a75a@dawncrow.de>
Date: Sun, 19 Jan 2020 22:51:59 +0100

Signed-off-by: André Hentschel <nerv@dawncrow.de>
---
 dlls/ntdll/ntdll.spec | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/dlls/ntdll/ntdll.spec b/dlls/ntdll/ntdll.spec
index 11111111111..11111111111 100644
--- a/dlls/ntdll/ntdll.spec
+++ b/dlls/ntdll/ntdll.spec
@@ -96,7 +96,7 @@
 @ stdcall LdrGetProcedureAddress(ptr ptr long ptr)
 # @ stub LdrHotPatchRoutine
 @ stub LdrInitShimEngineDynamic
-@ stdcall LdrInitializeThunk(ptr long long long)
+@ stdcall LdrInitializeThunk(ptr ptr long long)
 @ stub LdrLoadAlternateResourceModule
 @ stdcall LdrLoadDll(wstr long ptr ptr)
 @ stdcall LdrLockLoaderLock(long ptr ptr)
@@ -228,7 +228,7 @@
 @ stdcall -syscall NtGetCurrentProcessorNumber()
 # @ stub NtGetDevicePowerState
 @ stdcall -syscall NtGetNextThread(ptr ptr long long long ptr)
-@ stdcall -syscall NtGetNlsSectionPtr(long long long ptr ptr)
+@ stdcall -syscall NtGetNlsSectionPtr(long long ptr ptr ptr)
 # @ stub NtGetPlugPlayEvent
 @ stdcall NtGetTickCount()
 @ stdcall -syscall NtGetWriteWatch(long long ptr long ptr ptr ptr)
@@ -918,7 +918,7 @@
 @ stub RtlQueryProperties
 @ stub RtlQueryPropertyNames
 @ stub RtlQueryPropertySet
-@ stdcall RtlQueryRegistryValues(long ptr ptr ptr ptr)
+@ stdcall RtlQueryRegistryValues(long wstr ptr ptr ptr)
 @ stdcall RtlQueryRegistryValuesEx(long ptr ptr ptr ptr) RtlQueryRegistryValues
 @ stub RtlQuerySecurityObject
 @ stub RtlQueryTagHeap
@@ -1254,7 +1254,7 @@
 @ stdcall -private -norelay -syscall ZwGetContextThread(long ptr) NtGetContextThread
 @ stdcall -private -syscall ZwGetCurrentProcessorNumber() NtGetCurrentProcessorNumber
 # @ stub ZwGetDevicePowerState
-@ stdcall -private -syscall ZwGetNlsSectionPtr(long long long ptr ptr) NtGetNlsSectionPtr
+@ stdcall -private -syscall ZwGetNlsSectionPtr(long long ptr ptr ptr) NtGetNlsSectionPtr
 # @ stub ZwGetPlugPlayEvent
 @ stdcall -private ZwGetTickCount() NtGetTickCount
 @ stdcall -private -syscall ZwGetWriteWatch(long long ptr long ptr ptr ptr) NtGetWriteWatch
@@ -1476,7 +1476,7 @@
 @ cdecl -private -arch=i386 _CIpow()
 @ cdecl -private -arch=i386 _CIsin()
 @ cdecl -private -arch=i386 _CIsqrt()
-@ stdcall -arch=x86_64,arm,arm64 __C_specific_handler(ptr long ptr ptr)
+@ stdcall -arch=x86_64,arm,arm64 __C_specific_handler(ptr ptr ptr ptr)
 @ cdecl -arch=arm,arm64,x86_64 -norelay __chkstk()
 @ cdecl __isascii(long)
 @ cdecl __iscsym(long)
