From: Francois Gouget <fgouget@free.fr>
Subject: [PATCH] kernelbase: Fix an ASCII / ANSI mixup in a comment.
Message-Id: <E1mIRaY-00058N-I4@amboise>
Date: Tue, 24 Aug 2021 10:15:18 +0200 (CEST)

Signed-off-by: Francois Gouget <fgouget@free.fr>
---
 dlls/kernelbase/registry.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/kernelbase/registry.c b/dlls/kernelbase/registry.c
index a8f0643e8bf..5f777e0f752 100644
--- a/dlls/kernelbase/registry.c
+++ b/dlls/kernelbase/registry.c
@@ -1651,7 +1651,7 @@ LSTATUS WINAPI DECLSPEC_HOTPATCH RegQueryValueExA( HKEY hkey, LPCSTR name, LPDWO
     if (status && status != STATUS_BUFFER_OVERFLOW) goto done;
 
     /* we need to fetch the contents for a string type even if not requested,
-     * because we need to compute the length of the ASCII string. */
+     * because we need to compute the length of the ANSI string. */
     if (data || is_string(info->Type))
     {
         /* retry with a dynamically allocated buffer */
@@ -2106,7 +2106,7 @@ LSTATUS WINAPI RegEnumValueA( HKEY hkey, DWORD index, LPSTR value, LPDWORD val_c
                                   buffer, total_size, &total_size );
 
     /* we need to fetch the contents for a string type even if not requested,
-     * because we need to compute the length of the ASCII string. */
+     * because we need to compute the length of the ANSI string. */
 
     /* retry with a dynamically allocated buffer */
     while (status == STATUS_BUFFER_OVERFLOW)

-- 
2.20.1

