From: Francois Gouget <fgouget@free.fr>
Subject: [PATCH] iphlpapi: Fix an ASCII / ANSI mixup in a comment.
Message-Id: <E1mIRiq-0005wh-7h@amboise>
Date: Tue, 24 Aug 2021 10:23:52 +0200 (CEST)

Signed-off-by: Francois Gouget <fgouget@free.fr>
---
 dlls/iphlpapi/tests/iphlpapi.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/iphlpapi/tests/iphlpapi.c b/dlls/iphlpapi/tests/iphlpapi.c
index d1881616d30..223d6d270cb 100644
--- a/dlls/iphlpapi/tests/iphlpapi.c
+++ b/dlls/iphlpapi/tests/iphlpapi.c
@@ -28,7 +28,7 @@
  *
  * The Name field of the IP_ADAPTER_INDEX_MAP entries returned by
  * GetInterfaceInfo is declared as a wide string, but the bytes are actually
- * an ASCII string on some versions of the IP helper API under Win9x.  This was
+ * an ANSI string on some versions of the IP helper API under Win9x.  This was
  * apparently an MS bug, it's corrected in later versions.
  *
  * The DomainName field of FIXED_INFO isn't NULL-terminated on Win98.

-- 
2.20.1

