From: "André Hentschel" <nerv@dawncrow.de>
Subject: [PATCH 1/8] user32: Fix some spec file entries
Message-Id: <61293e05-cff4-ef74-fbca-76d94dfccbd2@dawncrow.de>
Date: Sun, 19 Jan 2020 22:51:31 +0100

Signed-off-by: André Hentschel <nerv@dawncrow.de>
---
 dlls/user32/user32.spec | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/dlls/user32/user32.spec b/dlls/user32/user32.spec
index 11111111111..11111111111 100644
--- a/dlls/user32/user32.spec
+++ b/dlls/user32/user32.spec
@@ -10,7 +10,7 @@
 @ stdcall AnyPopup()
 @ stdcall AppendMenuA(long long long ptr)
 @ stdcall AppendMenuW(long long long ptr)
-@ stdcall AreDpiAwarenessContextsEqual(long long)
+@ stdcall AreDpiAwarenessContextsEqual(ptr ptr)
 @ stdcall ArrangeIconicWindows(long)
 @ stdcall AttachThreadInput(long long long) NtUserAttachThreadInput
 @ stdcall BeginDeferWindowPos(long)
@@ -257,7 +257,7 @@
 @ stdcall GetAppCompatFlags2(long)
 @ stdcall GetAsyncKeyState(long)
 @ stdcall GetAutoRotationState(ptr)
-@ stdcall GetAwarenessFromDpiAwarenessContext(long)
+@ stdcall GetAwarenessFromDpiAwarenessContext(ptr)
 @ stdcall GetCapture()
 @ stdcall GetCaretBlinkTime()
 @ stdcall GetCaretPos(ptr)
@@ -475,7 +475,7 @@
 @ stdcall IsRectEmpty(ptr)
 # @ stub IsServerSideWindow
 @ stdcall IsTouchWindow(long ptr)
-@ stdcall IsValidDpiAwarenessContext(long)
+@ stdcall IsValidDpiAwarenessContext(ptr)
 @ stdcall IsWinEventHookInstalled(long)
 @ stdcall IsWindow(long)
 @ stdcall IsWindowEnabled(long)
@@ -691,7 +691,7 @@
 @ stdcall SetPhysicalCursorPos(long long)
 @ stdcall SetProcessDPIAware()
 @ stdcall SetProcessDefaultLayout(long)
-@ stdcall SetProcessDpiAwarenessContext(long)
+@ stdcall SetProcessDpiAwarenessContext(ptr)
 @ stdcall SetProcessDpiAwarenessInternal(long)
 @ stdcall SetProcessWindowStation(long) NtUserSetProcessWindowStation
 @ stdcall SetProgmanWindow (long)
