From: Francois Gouget <fgouget@free.fr>
Subject: [PATCH] imm32: Fix an ASCII / ANSI mixup in a comment.
Message-Id: <E1mIRiz-0005wu-VM@amboise>
Date: Tue, 24 Aug 2021 10:24:01 +0200 (CEST)

Signed-off-by: Francois Gouget <fgouget@free.fr>
---
 dlls/imm32/imm.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/imm32/imm.c b/dlls/imm32/imm.c
index 9d2a09f22be..20633041871 100644
--- a/dlls/imm32/imm.c
+++ b/dlls/imm32/imm.c
@@ -1755,7 +1755,7 @@ UINT WINAPI ImmGetDescriptionA(
   /* fetch the unicode buffer */
   len = ImmGetDescriptionW( hKL, buf, len + 1 );
 
-  /* convert it back to ASCII */
+  /* convert it back to ANSI */
   len = WideCharToMultiByte( CP_ACP, 0, buf, len + 1,
                              lpszDescription, uBufLen, NULL, NULL );
 

-- 
2.20.1

