From: Francois Gouget <fgouget@free.fr>
Subject: [PATCH] dpnet: Fix an ASCII / ANSI mixup in a warning message.
Message-Id: <E1mIRjn-00061t-W7@amboise>
Date: Tue, 24 Aug 2021 10:24:51 +0200 (CEST)

Signed-off-by: Francois Gouget <fgouget@free.fr>
---
 dlls/dpnet/address.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/dpnet/address.c b/dlls/dpnet/address.c
index 6aa62a97785..b616d221b38 100644
--- a/dlls/dpnet/address.c
+++ b/dlls/dpnet/address.c
@@ -500,7 +500,7 @@ static HRESULT WINAPI IDirectPlay8AddressImpl_AddComponent(IDirectPlay8Address *
         case DPNA_DATATYPE_STRING_ANSI:
             if ((strlen((const CHAR*)lpvData)+1) != dwDataSize)
             {
-                WARN("Invalid ASCII size, returning DPNERR_INVALIDPARAM\n");
+                WARN("Invalid ANSI size, returning DPNERR_INVALIDPARAM\n");
                 return DPNERR_INVALIDPARAM;
             }
             break;

-- 
2.20.1

