From: "André Hentschel" <nerv@dawncrow.de>
Subject: [PATCH 4/8] shell32: Fix some spec file entries
Message-Id: <325cc174-7cb2-61a3-4ec8-794725e39a32@dawncrow.de>
Date: Sun, 19 Jan 2020 22:51:51 +0100

Signed-off-by: André Hentschel <nerv@dawncrow.de>
---
 dlls/shell32/shell32.spec | 26 +++++++++++++-------------
 1 file changed, 13 insertions(+), 13 deletions(-)

diff --git a/dlls/shell32/shell32.spec b/dlls/shell32/shell32.spec
index 11111111111..11111111111 100644
--- a/dlls/shell32/shell32.spec
+++ b/dlls/shell32/shell32.spec
@@ -46,7 +46,7 @@
   48 stdcall -noname PathSetDlgItemPath(long long ptr) PathSetDlgItemPathAW
   49 stdcall -ordinal PathQualify(ptr) PathQualifyAW
   50 stdcall -noname PathStripToRoot(ptr) PathStripToRootAW
-  51 stdcall PathResolve(str long long) PathResolveAW
+  51 stdcall PathResolve(str ptr long) PathResolveAW
   52 stdcall -noname PathGetArgs(str) PathGetArgsAW
   53 stdcall DoEnvironmentSubst(ptr long) DoEnvironmentSubstAW
   54 stub LogoffWindowsDialog
@@ -108,7 +108,7 @@
  113 stdcall -noname FileMenu_ReplaceUsingPidl(long long ptr long ptr)
  114 stdcall -noname FileMenu_Create(long long long long long)
  115 stdcall -noname FileMenu_AppendItem(long ptr long long long long) FileMenu_AppendItemAW
- 116 stdcall -noname FileMenu_TrackPopupMenuEx(long long long long long long)
+ 116 stdcall -noname FileMenu_TrackPopupMenuEx(long long long long long ptr)
  117 stdcall -noname FileMenu_DeleteItemByCmd(long long)
  118 stdcall -noname FileMenu_Destroy(long)
  119 stdcall -ordinal IsLFNDrive(ptr) IsLFNDriveAW
@@ -160,11 +160,11 @@
  168 stdcall -ordinal SHCreatePropSheetExtArray(long wstr long)
  169 stdcall -ordinal SHDestroyPropSheetExtArray(long)
  170 stdcall -ordinal SHReplaceFromPropSheetExtArray(long long ptr long)
- 171 stdcall -ordinal PathCleanupSpec(ptr ptr)
+ 171 stdcall -ordinal PathCleanupSpec(wstr ptr)
  172 stdcall -noname SHCreateLinks(long str ptr long ptr)
  173 stdcall -ordinal SHValidateUNC(long wstr long)
  174 stdcall -ordinal SHCreateShellFolderViewEx(ptr ptr)
- 175 stdcall -noname SHGetSpecialFolderPath(long long long long) SHGetSpecialFolderPathAW
+ 175 stdcall -noname SHGetSpecialFolderPath(long ptr long long) SHGetSpecialFolderPathAW
  176 stdcall -ordinal SHSetInstanceExplorer(ptr) shcore.SetProcessReference
  177 stub DAD_SetDragImageFromListView
  178 stdcall -ordinal SHObjectProperties(long long wstr wstr)
@@ -217,11 +217,11 @@
 #300 stub Shl3216_ThunkData32
 
  505 stdcall SHRegCloseKey (long)
- 506 stdcall SHRegOpenKeyA (long str long)
- 507 stdcall SHRegOpenKeyW (long wstr long)
+ 506 stdcall SHRegOpenKeyA (long str ptr)
+ 507 stdcall SHRegOpenKeyW (long wstr ptr)
  508 stdcall SHRegQueryValueA(long str ptr ptr)
  509 stdcall SHRegQueryValueExA(long str ptr ptr ptr ptr)
- 510 stdcall SHRegQueryValueW (long long long long)
+ 510 stdcall SHRegQueryValueW (long wstr ptr ptr)
  511 stdcall SHRegQueryValueExW (long wstr ptr ptr ptr ptr)
  512 stdcall SHRegDeleteKeyW (long wstr)
 
@@ -247,7 +247,7 @@
 
  651 stdcall -noname @(ptr long) ReadCabinetState # OldReadCabinetState
  652 stdcall -noname WriteCabinetState(ptr)
- 653 stdcall -noname PathProcessCommand(long long long long) PathProcessCommandAW
+ 653 stdcall -noname PathProcessCommand(ptr ptr long long) PathProcessCommandAW
  654 stdcall ReadCabinetState(ptr long)
 
  660 stdcall -noname FileIconInit(long)
@@ -433,9 +433,9 @@
 @ stub SheFullPathA
 @ stub SheFullPathW
 @ stub SheGetCurDrive
-@ stdcall SheGetDirA(long long)
+@ stdcall SheGetDirA(long ptr)
 @ stub SheGetDirExW
-@ stdcall SheGetDirW (long long)
+@ stdcall SheGetDirW (long ptr)
 @ stub SheGetPathOffsetW
 @ stub SheRemoveQuotesA
 @ stub SheRemoveQuotesW
@@ -448,9 +448,9 @@
 @ stdcall ShellExec_RunDLLA(long long str long)
 @ stdcall ShellExec_RunDLLW(long long wstr long)
 @ stdcall ShellExecuteA(long str str str str long)
-@ stdcall ShellExecuteEx (long) ShellExecuteExA
-@ stdcall ShellExecuteExA (long)
-@ stdcall ShellExecuteExW (long)
+@ stdcall ShellExecuteEx (ptr) ShellExecuteExA
+@ stdcall ShellExecuteExA (ptr)
+@ stdcall ShellExecuteExW (ptr)
 @ stdcall ShellExecuteW (long wstr wstr wstr wstr long)
 @ stdcall ShellHookProc(long long long)
 @ stdcall Shell_NotifyIcon(long ptr) Shell_NotifyIconA
