From: "André Hentschel" <nerv@dawncrow.de>
Subject: [PATCH 7/8] shlwapi: Fix some spec file entries
Message-Id: <766556f0-eea9-f6c0-8f62-6e0471a60a74@dawncrow.de>
Date: Sun, 19 Jan 2020 22:51:56 +0100

Signed-off-by: André Hentschel <nerv@dawncrow.de>
---
 dlls/shlwapi/shlwapi.spec | 20 ++++++++++----------
 1 file changed, 10 insertions(+), 10 deletions(-)

diff --git a/dlls/shlwapi/shlwapi.spec b/dlls/shlwapi/shlwapi.spec
index 68d0e7f8122..d0cced75370 100644
--- a/dlls/shlwapi/shlwapi.spec
+++ b/dlls/shlwapi/shlwapi.spec
@@ -219,7 +219,7 @@
 219 stdcall -import QISearch(ptr ptr ptr ptr)
 220 stdcall -noname SHSetDefaultDialogFont(ptr long)
 221 stdcall -noname SHRemoveDefaultDialogFont(ptr)
-222 stdcall -noname SHGlobalCounterCreate(long)
+222 stdcall -noname SHGlobalCounterCreate(ptr)
 223 stdcall -noname SHGlobalCounterGetValue(long)
 224 stdcall -noname SHGlobalCounterIncrement(long)
 225 stdcall -ordinal SHStripMneumonicW(wstr)
@@ -257,15 +257,15 @@
 257 stdcall -noname SHCreateWorkerWindowA(ptr ptr long long ptr long)
 258 stub -noname SHRegisterWaitForSingleObject
 259 stub -noname SHUnregisterWait
-260 stdcall -noname SHQueueUserWorkItem(long long long long long long long)
+260 stdcall -noname SHQueueUserWorkItem(ptr ptr long long ptr str long)
 261 stub -noname SHCreateTimerQueue
 262 stub -noname SHDeleteTimerQueue
 263 stdcall -noname SHSetTimerQueueTimer(long ptr ptr long long str long)
 264 stub -noname SHChangeTimerQueueTimer
 265 stub -noname SHCancelTimerQueueTimer
 266 stdcall -noname SHRestrictionLookup(long wstr ptr ptr)
-267 stdcall -noname SHWeakQueryInterface(long long long long)
-268 stdcall -noname SHWeakReleaseInterface(long long)
+267 stdcall -noname SHWeakQueryInterface(ptr ptr ptr ptr)
+268 stdcall -noname SHWeakReleaseInterface(ptr ptr)
 269 stdcall -noname GUIDFromStringA(str ptr)
 270 stdcall -noname GUIDFromStringW(wstr ptr)
 271 stdcall -noname SHGetRestriction(wstr wstr wstr)
@@ -280,7 +280,7 @@
 280 stdcall -ordinal SHRegGetIntW(ptr wstr long) shcore.SHRegGetIntW
 281 stdcall -noname SHPackDispParamsV(ptr ptr long ptr)
 282 varargs -noname SHPackDispParams(ptr ptr long)
-283 stdcall -noname IConnectionPoint_InvokeWithCancel(ptr long long long long)
+283 stdcall -noname IConnectionPoint_InvokeWithCancel(ptr long ptr long long)
 284 stdcall -noname IConnectionPoint_SimpleInvoke(ptr long ptr)
 285 stdcall -noname IConnectionPoint_OnChanged(ptr long)
 286 varargs -noname IUnknown_CPContainerInvokeParam(ptr ptr long ptr long)
@@ -354,7 +354,7 @@
 354 stdcall -ordinal SHFormatDateTimeW(ptr ptr ptr long)
 355 stdcall -noname IUnknown_EnableModeless(ptr long)
 356 stdcall -noname CreateAllAccessSecurityAttributes(ptr ptr long)
-357 stdcall -noname SHGetNewLinkInfoWrapW(wstr wstr wstr long long)
+357 stdcall -noname SHGetNewLinkInfoWrapW(wstr wstr wstr ptr long)
 358 stdcall -noname SHDefExtractIconWrapW(wstr long long ptr ptr long)
 359 stdcall -noname OpenEventWrapW(long long wstr) kernel32.OpenEventW
 360 stdcall -noname RemoveDirectoryWrapW(wstr) kernel32.RemoveDirectoryW
@@ -481,14 +481,14 @@
 481 stdcall -noname IUnknown_HasFocusIO(ptr)
 482 stub -noname SHMessageBoxHelpA
 483 stub -noname SHMessageBoxHelpW
-484 stdcall -noname IUnknown_QueryServiceExec(ptr ptr ptr long long long ptr)
+484 stdcall -noname IUnknown_QueryServiceExec(ptr ptr ptr long long ptr ptr)
 485 stub -noname MapWin32ErrorToSTG
 486 stub -noname ModeToCreateFileFlags
 487 stdcall -ordinal SHLoadIndirectString(wstr ptr long ptr)
 488 stub -noname SHConvertGraphicsFile
 489 stdcall -noname GlobalAddAtomWrapW(wstr) kernel32.GlobalAddAtomW
 490 stdcall -noname GlobalFindAtomWrapW(wstr) kernel32.GlobalFindAtomW
-491 stdcall -noname SHGetShellKey(long long long)
+491 stdcall -noname SHGetShellKey(long wstr long)
 492 stub -noname PrettifyFileDescriptionW
 493 stub -noname SHPropertyBag_ReadType
 494 stub -noname SHPropertyBag_ReadStr
@@ -663,8 +663,8 @@
 @ stdcall -import PathRenameExtensionW(wstr wstr)
 @ stdcall -import PathSearchAndQualifyA(str ptr long)
 @ stdcall -import PathSearchAndQualifyW(wstr ptr long)
-@ stdcall PathSetDlgItemPathA (long long ptr)
-@ stdcall PathSetDlgItemPathW (long long ptr)
+@ stdcall PathSetDlgItemPathA (long long str)
+@ stdcall PathSetDlgItemPathW (long long wstr)
 @ stdcall -import PathSkipRootA(str)
 @ stdcall -import PathSkipRootW(wstr)
 @ stdcall -import PathStripPathA(str)
