From b5151dfab81e95b03ce7f9dc576c16d31c9c2b19 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Mon, 25 Jan 2021 15:16:19 +0100
Subject: [PATCH 03/13] server: Notify desktop of cursor window changes.

---
 server/queue.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/server/queue.c b/server/queue.c
index 11111111111..11111111111 100644
--- a/server/queue.c
+++ b/server/queue.c
@@ -530,7 +530,11 @@ static int update_desktop_cursor_pos( struct desktop *desktop, user_handle_t win
     else if (!win || !is_window_visible( win ) || is_window_transparent( win ))
         win = shallow_window_from_point( desktop, x, y );
 
-    if (win != desktop->cursor_win) updated = 1;
+    if (win != desktop->cursor_win)
+    {
+        post_desktop_message( desktop, desktop->cursor_change_msg, win, desktop->cursor_handle );
+        updated = 1;
+    }
     desktop->cursor_win = win;
 
     return updated;
-- 
2.37.1

