From 5a727263adf56194a24c5abad8350205e81c068b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Mon, 25 Jan 2021 15:22:08 +0100
Subject: [PATCH 04/13] winex11.drv: Always set the clipping window cursor when
 clipping.

---
 dlls/winex11.drv/window.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/winex11.drv/window.c b/dlls/winex11.drv/window.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/window.c
+++ b/dlls/winex11.drv/window.c
@@ -3261,6 +3261,7 @@ LRESULT X11DRV_WindowMessage( HWND hwnd, UINT msg, WPARAM wp, LPARAM lp )
             else
                 sync_window_cursor( win );
         }
+        if (clipping_cursor) set_window_cursor( x11drv_thread_data()->clip_window, (HCURSOR)lp );
         return 0;
     }
     case WM_X11DRV_CLIP_CURSOR_NOTIFY:
-- 
2.39.0

