From 5fc9620801dc17e6864d266768b76b2f2eaf9423 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Fri, 22 Jan 2021 13:44:19 +0100
Subject: [PATCH 07/13] winex11.drv: Properly cache HCURSOR to Cursor
 associations.

---
 dlls/winex11.drv/mouse.c | 57 ++++++++++++++++++++++++----------------
 1 file changed, 35 insertions(+), 22 deletions(-)

diff --git a/dlls/winex11.drv/mouse.c b/dlls/winex11.drv/mouse.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/mouse.c
+++ b/dlls/winex11.drv/mouse.c
@@ -206,35 +206,46 @@ static Cursor get_empty_cursor(void)
     return cursor;
 }
 
-/***********************************************************************
- *		set_window_cursor
- */
-void set_window_cursor( Window window, HCURSOR handle )
+static Cursor get_cursor( HCURSOR handle )
 {
     Cursor cursor, prev;
 
-    if (!handle) cursor = get_empty_cursor();
-    else if (XFindContext( gdi_display, (XID)handle, cursor_context, (char **)&cursor ))
-    {
-        /* try to create it */
-        if (!(cursor = create_cursor( handle ))) return;
+    if (!handle) return get_empty_cursor();
 
-        XLockDisplay( gdi_display );
-        if (!XFindContext( gdi_display, (XID)handle, cursor_context, (char **)&prev ))
-        {
-            /* someone else was here first */
-            XFreeCursor( gdi_display, cursor );
-            cursor = prev;
-        }
-        else
-        {
-            XSaveContext( gdi_display, (XID)handle, cursor_context, (char *)cursor );
-            TRACE( "cursor %p created %lx\n", handle, cursor );
-        }
+    XLockDisplay( gdi_display );
+    if (!XFindContext( gdi_display, (XID)handle, cursor_context, (char **)&cursor ))
+    {
         XUnlockDisplay( gdi_display );
+        return cursor;
+    }
+    XUnlockDisplay( gdi_display );
+
+    /* try to create it */
+    if (!(cursor = create_cursor( handle ))) return None;
+
+    XLockDisplay( gdi_display );
+    if (!XFindContext( gdi_display, (XID)handle, cursor_context, (char **)&prev ))
+    {
+        /* someone else was here first */
+        XFreeCursor( gdi_display, cursor );
+        cursor = prev;
     }
+    else
+    {
+        XSaveContext( gdi_display, (XID)handle, cursor_context, (char *)cursor );
+        TRACE( "cursor %p created %lx\n", handle, cursor );
+    }
+    XUnlockDisplay( gdi_display );
 
-    XDefineCursor( gdi_display, window, cursor );
+    return cursor;
+}
+
+/***********************************************************************
+ *		set_window_cursor
+ */
+void set_window_cursor( Window window, HCURSOR handle )
+{
+    XDefineCursor( gdi_display, window, get_cursor( handle ) );
     /* make the change take effect immediately */
     XFlush( gdi_display );
 }
@@ -1527,12 +1538,14 @@ void X11DRV_DestroyCursorIcon( HCURSOR handle )
 {
     Cursor cursor;
 
+    XLockDisplay( gdi_display );
     if (!XFindContext( gdi_display, (XID)handle, cursor_context, (char **)&cursor ))
     {
         TRACE( "%p xid %lx\n", handle, cursor );
         XFreeCursor( gdi_display, cursor );
         XDeleteContext( gdi_display, (XID)handle, cursor_context );
     }
+    XUnlockDisplay( gdi_display );
 }
 
 /***********************************************************************
-- 
2.37.1

