From deeb170f356860e27293ef880325a938e8bc0c9b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Fri, 22 Apr 2022 23:12:33 +0200
Subject: [PATCH 7/7] winex11.drv: Do not enable/disable XInput2 when clipping.

---
 dlls/winex11.drv/mouse.c | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/dlls/winex11.drv/mouse.c b/dlls/winex11.drv/mouse.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/mouse.c
+++ b/dlls/winex11.drv/mouse.c
@@ -454,9 +454,6 @@ static BOOL grab_clipping_window( const RECT *clip )
         return FALSE;
     }
 
-    /* enable XInput2 unless we are already clipping */
-    if (!data->clipping_cursor) X11DRV_XInput2_Enable( data->display, None, PointerMotionMask );
-
     TRACE( "clipping to %s win %lx\n", wine_dbgstr_rect(clip), clip_window );
 
     if (!data->clipping_cursor) XUnmapWindow( data->display, clip_window );
@@ -498,7 +495,6 @@ static BOOL grab_clipping_window( const RECT *clip )
 
     if (!clipping_cursor)
     {
-        X11DRV_XInput2_Enable( data->display, None, 0 );
         return FALSE;
     }
     clip_rect = *clip;
-- 
0.0.0

