From 210601101eb191bcd20a1d302a2946ba8d513e3b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Fri, 22 Apr 2022 23:12:33 +0200
Subject: [PATCH 13/13] winex11.drv: Do not enable/disable XInput2 when
 clipping.

---
 dlls/winex11.drv/mouse.c | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/dlls/winex11.drv/mouse.c b/dlls/winex11.drv/mouse.c
index b73f80e4dcd..2acf14e098d 100644
--- a/dlls/winex11.drv/mouse.c
+++ b/dlls/winex11.drv/mouse.c
@@ -449,9 +449,6 @@ static BOOL grab_clipping_window( const RECT *clip )
         return FALSE;
     }
 
-    /* enable XInput2 unless we are already clipping */
-    if (!clipping_cursor) X11DRV_XInput2_Enable( data->display, None, PointerMotionMask );
-
     TRACE( "clipping to %s win %lx\n", wine_dbgstr_rect(clip), clip_window );
 
     pos = virtual_screen_to_root( clip->left, clip->top );
@@ -477,7 +474,6 @@ static BOOL grab_clipping_window( const RECT *clip )
 
     if (!clipping_cursor)
     {
-        X11DRV_XInput2_Enable( data->display, None, 0 );
         XUnmapWindow( data->display, clip_window );
         return FALSE;
     }
@@ -507,7 +503,6 @@ void ungrab_clipping_window(void)
     if (clipping_cursor) XUngrabPointer( data->display, CurrentTime );
     clipping_cursor = FALSE;
     data->clip_reset = NtGetTickCount();
-    X11DRV_XInput2_Enable( data->display, None, 0 );
 }
 
 /***********************************************************************
-- 
2.37.2

