From 75846cd9cd236315496e953298b41c2b8dd4bdd7 Mon Sep 17 00:00:00 2001
From: Huw Davies <huw@codeweavers.com>
Date: Mon, 29 Jun 2020 11:12:12 +0100
Subject: [PATCH 02/44] server: Use the helper to update the cursor last change
 time.

Signed-off-by: Huw Davies <huw@codeweavers.com>
---
 server/queue.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/server/queue.c b/server/queue.c
index 8fcdd3841ca..4a03c37ae8f 100644
--- a/server/queue.c
+++ b/server/queue.c
@@ -1895,7 +1895,7 @@ static int queue_mouse_message( struct desktop *desktop, user_handle_t win, cons
         WM_MOUSEHWHEEL   /* 0x1000 = MOUSEEVENTF_HWHEEL */
     };
 
-    desktop->cursor.last_change = get_tick_count();
+    update_desktop_cursor_pos( desktop, desktop->cursor.x, desktop->cursor.y ); /* Update last change time */
     flags = input->mouse.flags;
     time  = input->mouse.time;
     if (!time) time = desktop->cursor.last_change;
-- 
2.35.1

