From adef300dbfaf72473f6c7c10b6b34043addb71c5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Thu, 29 Apr 2021 20:07:16 +0200
Subject: [PATCH 23/44] user32: Yield thread in peek_message when no message
 was found.

---
 dlls/user32/message.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/user32/message.c b/dlls/user32/message.c
index 11111111111..11111111111 100644
--- a/dlls/user32/message.c
+++ b/dlls/user32/message.c
@@ -2780,6 +2780,7 @@ static int peek_message( MSG *msg, HWND hwnd, UINT first, UINT last, UINT flags,
                 thread_info->wake_mask = changed_mask & (QS_SENDMESSAGE | QS_SMRESULT);
                 thread_info->changed_mask = changed_mask;
                 if (buffer != buffer_init) HeapFree( GetProcessHeap(), 0, buffer );
+                SwitchToThread();
                 return 0;
             }
             if (res != STATUS_BUFFER_OVERFLOW)
-- 
2.35.1

