From 6895ef6486d7e8b3d0eaf7e26e634d1b7f106b8d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Thu, 29 Apr 2021 20:07:16 +0200
Subject: [PATCH 23/40] user32: Yield thread in peek_message when no message
 was found.

---
 dlls/win32u/message.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/win32u/message.c b/dlls/win32u/message.c
index 11111111111..11111111111 100644
--- a/dlls/win32u/message.c
+++ b/dlls/win32u/message.c
@@ -2712,6 +2712,7 @@ static int peek_message( MSG *msg, HWND hwnd, UINT first, UINT last, UINT flags,
                 thread_info->wake_mask = changed_mask & (QS_SENDMESSAGE | QS_SMRESULT);
                 thread_info->changed_mask = changed_mask;
                 if (buffer != buffer_init) free( buffer );
+                NtYieldExecution();
                 return 0;
             }
             if (res != STATUS_BUFFER_OVERFLOW)
-- 
0.0.0

