From d68ba53859bf09683d0b2c00789601bad0b3ecc3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Mon, 18 Jan 2021 17:48:47 +0100
Subject: [PATCH 28/44] user32: Use input shared memory for GetFocus.

---
 dlls/user32/focus.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/dlls/user32/focus.c b/dlls/user32/focus.c
index ac38a4f01ba..bcd4c169708 100644
--- a/dlls/user32/focus.c
+++ b/dlls/user32/focus.c
@@ -374,14 +374,15 @@ HWND WINAPI GetActiveWindow(void)
  */
 HWND WINAPI GetFocus(void)
 {
+    volatile struct input_shared_memory *shared = get_input_shared_memory();
     HWND ret = 0;
 
-    SERVER_START_REQ( get_thread_input )
+    if (!shared) return 0;
+    SHARED_READ_BEGIN( &shared->seq )
     {
-        req->tid = GetCurrentThreadId();
-        if (!wine_server_call_err( req )) ret = wine_server_ptr_handle( reply->focus );
+        ret = wine_server_ptr_handle( shared->focus );
     }
-    SERVER_END_REQ;
+    SHARED_READ_END( &shared->seq );
     return ret;
 }
 
-- 
2.34.1

