From 9dbbc399b23177075552a54881b1844451502cfe Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Tue, 17 Aug 2021 16:25:35 +0200
Subject: [PATCH 40/44] server: Only return full keystate when requested.

---
 server/queue.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/server/queue.c b/server/queue.c
index 9511acebef1..6556b83edd3 100644
--- a/server/queue.c
+++ b/server/queue.c
@@ -3366,7 +3366,7 @@ DECL_HANDLER(get_key_state)
             desktop->shared->keystate[req->key & 0xff] &= ~0x40;
             SHARED_WRITE_END( &desktop->shared->seq );
         }
-        set_reply_data( (void *)desktop->shared->keystate, size );
+        else set_reply_data( (void *)desktop->shared->keystate, size );
         release_object( desktop );
     }
     else
@@ -3378,7 +3378,7 @@ DECL_HANDLER(get_key_state)
             sync_input_keystate( queue->input );
             reply->state = keystate[req->key & 0xff];
         }
-        set_reply_data( keystate, size );
+        else set_reply_data( keystate, size );
     }
 }
 
-- 
2.35.1

