From d197d14d834019325554640773a4dade5bb1ab50 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Wed, 1 Jun 2022 15:18:56 +0200
Subject: [PATCH 01/18] user32/tests: Fix indentation to silence a warning.

---
 dlls/user32/tests/msg.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/dlls/user32/tests/msg.c b/dlls/user32/tests/msg.c
index 11111111111..11111111111 100644
--- a/dlls/user32/tests/msg.c
+++ b/dlls/user32/tests/msg.c
@@ -15210,9 +15210,11 @@ static void test_ShowWindow(void)
         ok(wp.ptMaxPosition.x == sw[i].wp_max.x && wp.ptMaxPosition.y == sw[i].wp_max.y,
            "expected %ld,%ld got %ld,%ld\n", sw[i].wp_max.x, sw[i].wp_max.y, wp.ptMaxPosition.x, wp.ptMaxPosition.y);
 
-if (0) /* FIXME: Wine behaves completely different here */
-        ok(EqualRect(&win_rc, &wp.rcNormalPosition), "expected %s got %s\n",
-           wine_dbgstr_rect(&win_rc), wine_dbgstr_rect(&wp.rcNormalPosition));
+        if (0) /* FIXME: Wine behaves completely different here */
+        {
+            ok(EqualRect(&win_rc, &wp.rcNormalPosition), "expected %s got %s\n",
+               wine_dbgstr_rect(&win_rc), wine_dbgstr_rect(&wp.rcNormalPosition));
+        }
     }
     DestroyWindow(hwnd);
     flush_events();
-- 
0.0.0

