From: Hans Leidekker <hans@codeweavers.com>
Subject: [PATCH 1/6] secur32: Fix expected buffer size in trace message.
Message-Id: <20211028155309.2481276-1-hans@codeweavers.com>
Date: Thu, 28 Oct 2021 17:53:04 +0200

Signed-off-by: Hans Leidekker <hans@codeweavers.com>
---
 dlls/secur32/schannel.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/secur32/schannel.c b/dlls/secur32/schannel.c
index eb8106e88c7..885fabfac13 100644
--- a/dlls/secur32/schannel.c
+++ b/dlls/secur32/schannel.c
@@ -1003,7 +1003,7 @@ static SECURITY_STATUS SEC_ENTRY schan_InitializeSecurityContextW(
             if (!expected_size)
             {
                 TRACE("Expected at least %lu bytes, but buffer only contains %u bytes.\n",
-                      max(6, record_size), buffer->cbBuffer);
+                      max(ctx->header_size + 1, record_size), buffer->cbBuffer);
                 return SEC_E_INCOMPLETE_MESSAGE;
             }
         }

-- 
2.30.2

