From 42ae7b0a6b719c3d5091c8ec1bc72af6db737979 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Tue, 10 Nov 2020 13:04:45 +0100
Subject: [PATCH] secur32: Disable CHACHA20-POLY1305 ciphersuites.

---
 dlls/secur32/schannel_gnutls.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/secur32/schannel_gnutls.c b/dlls/secur32/schannel_gnutls.c
index b7ffa1822b5..4d5fb006917 100644
--- a/dlls/secur32/schannel_gnutls.c
+++ b/dlls/secur32/schannel_gnutls.c
@@ -475,7 +475,7 @@ static NTSTATUS schan_create_session( void *args )
     const struct create_session_params *params = args;
     schan_credentials *cred = params->cred;
     gnutls_session_t *s = (gnutls_session_t*)&params->transport->session;
-    char priority[128] = "NORMAL:%LATEST_RECORD_VERSION", *p;
+    char priority[128] = "NORMAL:-CHACHA20-POLY1305:%LATEST_RECORD_VERSION", *p;
     BOOL using_vers_all = FALSE, disabled;
     unsigned int i, flags = (cred->credential_use == SECPKG_CRED_INBOUND) ? GNUTLS_SERVER : GNUTLS_CLIENT;
     int err;
-- 
2.35.1

