From: Hans Leidekker <hans@codeweavers.com>
Subject: [PATCH 3/6] secur32: Clear expiry timestamp in schan_InitializeSecurityContextW().
Message-Id: <20211028155309.2481276-3-hans@codeweavers.com>
Date: Thu, 28 Oct 2021 17:53:06 +0200

Signed-off-by: Hans Leidekker <hans@codeweavers.com>
---
 dlls/secur32/schannel.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/dlls/secur32/schannel.c b/dlls/secur32/schannel.c
index 885fabfac13..fdf8e990f62 100644
--- a/dlls/secur32/schannel.c
+++ b/dlls/secur32/schannel.c
@@ -904,6 +904,12 @@ static SECURITY_STATUS SEC_ENTRY schan_InitializeSecurityContextW(
     dump_buffer_desc(pInput);
     dump_buffer_desc(pOutput);
 
+    if (ptsExpiry)
+    {
+        ptsExpiry->LowPart = 0;
+        ptsExpiry->HighPart = 0;
+    }
+
     if (!phContext || (phNewContext && !pInput))
     {
         ULONG_PTR handle;

-- 
2.30.2

