From: Hans Leidekker <hans@codeweavers.com>
Subject: [PATCH 2/6] secur32/tests: Check supported DTLS protocols.
Message-Id: <20211028155309.2481276-2-hans@codeweavers.com>
Date: Thu, 28 Oct 2021 17:53:05 +0200

Signed-off-by: Hans Leidekker <hans@codeweavers.com>
---
 dlls/secur32/tests/schannel.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/secur32/tests/schannel.c b/dlls/secur32/tests/schannel.c
index aeb38096f0c..b7f9d409d41 100644
--- a/dlls/secur32/tests/schannel.c
+++ b/dlls/secur32/tests/schannel.c
@@ -203,6 +203,8 @@ static void test_supported_protocols(CredHandle *handle, unsigned exprots)
     X(SP_PROT_TLS1_1_CLIENT, "TLS 1.1 client");
     X(SP_PROT_TLS1_2_CLIENT, "TLS 1.2 client");
     X(SP_PROT_TLS1_3_CLIENT, "TLS 1.3 client");
+    X(SP_PROT_DTLS1_0_CLIENT, "DTLS 1.0 client");
+    X(SP_PROT_DTLS1_2_CLIENT, "DTLS 1.2 client");
 #undef X
 
     if(protocols.grbitProtocol)

-- 
2.30.2

