From bffedb15a8978c4483d35073fb37c47aed95ac1c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Thu, 15 Apr 2021 12:35:12 +0200
Subject: [PATCH 07/14] explorer: Load null graphics driver as a fallback.

---
 programs/explorer/desktop.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/programs/explorer/desktop.c b/programs/explorer/desktop.c
index 6d577dd0fd1..7c90575013c 100644
--- a/programs/explorer/desktop.c
+++ b/programs/explorer/desktop.c
@@ -32,6 +32,7 @@
 #include "exdisp.h"
 
 #include "wine/debug.h"
+#include "wine/gdi_driver.h"
 #include "explorer_private.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(explorer);
@@ -39,7 +40,8 @@ WINE_DEFAULT_DEBUG_CHANNEL(explorer);
 #define DESKTOP_CLASS_ATOM ((LPCWSTR)MAKEINTATOM(32769))
 #define DESKTOP_ALL_ACCESS 0x01ff
 
-static const WCHAR default_driver[] = {'m','a','c',',','x','1','1',0};
+static const WCHAR default_driver[] = L"mac,x11,null";
+static struct user_driver_funcs null_driver;
 
 static BOOL using_root;
 
@@ -808,7 +810,6 @@ static HMODULE load_graphics_driver( const WCHAR *driver, const GUID *guid )
 
     WCHAR buffer[MAX_PATH], libname[32], *name, *next;
     WCHAR key[ARRAY_SIZE( device_keyW ) + 39];
-    BOOL null_driver = FALSE;
     HMODULE module = 0;
     HKEY hkey;
     char error[80];
@@ -835,9 +836,9 @@ static HMODULE load_graphics_driver( const WCHAR *driver, const GUID *guid )
 
         if (!wcscmp( name, L"null" ))
         {
+            __wine_set_user_driver( &null_driver, WINE_GDI_DRIVER_VERSION );
             TRACE( "display %s using null driver\n", debugstr_guid(guid) );
             wcscpy( libname, L"null" );
-            null_driver = TRUE;
             break;
         }
 
@@ -867,7 +868,7 @@ static HMODULE load_graphics_driver( const WCHAR *driver, const GUID *guid )
     if (!RegCreateKeyExW( HKEY_LOCAL_MACHINE, key, 0, NULL,
                           REG_OPTION_VOLATILE, KEY_SET_VALUE, NULL, &hkey, NULL  ))
     {
-        if (module || null_driver)
+        if (module || !wcscmp( libname, L"null" ))
             RegSetValueExW( hkey, graphics_driverW, 0, REG_SZ,
                             (BYTE *)libname, (lstrlenW(libname) + 1) * sizeof(WCHAR) );
         else
-- 
2.34.1

