From b7e424eb6ea31684c3c4845ca004e75ffe3ff227 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Mon, 25 Jan 2021 15:05:19 +0100
Subject: [PATCH 01/13] server: Move desktop local variable to wider scope.

---
 server/queue.c | 18 +++++++++---------
 1 file changed, 9 insertions(+), 9 deletions(-)

diff --git a/server/queue.c b/server/queue.c
index cb95174a938..d4d23f2b7a2 100644
--- a/server/queue.c
+++ b/server/queue.c
@@ -3616,14 +3616,16 @@ DECL_HANDLER(set_cursor)
 {
     struct msg_queue *queue = get_current_queue();
     struct thread_input *input;
+    struct desktop *desktop;
 
     if (!queue) return;
     input = queue->input;
+    desktop = input->desktop;
 
     reply->prev_handle = input->shared->cursor;
     reply->prev_count  = input->shared->cursor_count;
-    reply->prev_x      = input->desktop->shared->cursor.x;
-    reply->prev_y      = input->desktop->shared->cursor.y;
+    reply->prev_x      = desktop->shared->cursor.x;
+    reply->prev_y      = desktop->shared->cursor.y;
 
     if ((req->flags & SET_CURSOR_HANDLE) && req->handle &&
         !get_user_object( req->handle, USER_CLIENT ))
@@ -3645,12 +3647,10 @@ DECL_HANDLER(set_cursor)
     SHARED_WRITE_END( &input->shared->seq );
     if (req->flags & SET_CURSOR_POS)
     {
-        set_cursor_pos( input->desktop, req->x, req->y );
+        set_cursor_pos( desktop, req->x, req->y );
     }
     if (req->flags & (SET_CURSOR_CLIP | SET_CURSOR_NOCLIP))
     {
-        struct desktop *desktop = input->desktop;
-
         /* only the desktop owner can set the message */
         if (req->clip_msg && get_top_window_owner(desktop) == current->process)
             desktop->cursor_clip_msg = req->clip_msg;
@@ -3658,10 +3658,10 @@ DECL_HANDLER(set_cursor)
         set_clip_rectangle( desktop, (req->flags & SET_CURSOR_NOCLIP) ? NULL : &req->clip, 0 );
     }
 
-    reply->new_x       = input->desktop->shared->cursor.x;
-    reply->new_y       = input->desktop->shared->cursor.y;
-    reply->new_clip    = input->desktop->shared->cursor.clip;
-    reply->last_change = input->desktop->shared->cursor.last_change;
+    reply->new_x       = desktop->shared->cursor.x;
+    reply->new_y       = desktop->shared->cursor.y;
+    reply->new_clip    = desktop->shared->cursor.clip;
+    reply->last_change = desktop->shared->cursor.last_change;
 }
 
 /* Get the history of the 64 last cursor positions */

