From 10df80804000aa66c7ef92f1361759e309f0e321 Mon Sep 17 00:00:00 2001
From: Giovanni Mascellani <gmascellani@codeweavers.com>
Date: Mon, 15 Mar 2021 12:01:25 -0500
Subject: [PATCH] winex11.drv: Flush X connection after ungrabbing the pointer

---
 dlls/winex11.drv/mouse.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/dlls/winex11.drv/mouse.c b/dlls/winex11.drv/mouse.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/mouse.c
+++ b/dlls/winex11.drv/mouse.c
@@ -585,7 +585,11 @@ void ungrab_clipping_window(void)
     TRACE( "no longer clipping\n" );
     if (clip_window_mapped) XUnmapWindow( display, clip_window );
     clip_window_mapped = FALSE;
-    if (clipping_cursor) XUngrabPointer( display, CurrentTime );
+    if (clipping_cursor)
+    {
+        XUngrabPointer( display, CurrentTime );
+        XFlush( display );
+    }
     clipping_cursor = FALSE;
     SendNotifyMessageW( GetDesktopWindow(), WM_X11DRV_CLIP_CURSOR_NOTIFY, 0, 0 );
 }
