From: Chao Long <longchao@uniontech.com>
Subject: [PATCH] user32: Fix event queue congestion can cause WINPOS_WindowFromPoint is null.
Message-Id: <20211021033122.18818-1-longchao@uniontech.com>
Date: Thu, 21 Oct 2021 11:31:22 +0800

Signed-off-by: Chao Long <longchao@uniontech.com>
---
 dlls/user32/input.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/user32/input.c b/dlls/user32/input.c
index ea2f69c44a5..55929740f38 100644
--- a/dlls/user32/input.c
+++ b/dlls/user32/input.c
@@ -1095,7 +1095,7 @@ TrackMouseEvent (TRACKMOUSEEVENT *ptme)
             tracking_info.tme.dwHoverTime = 0;
         }
 
-        if (ptme->hwndTrack == hwnd)
+        if (ptme->hwndTrack == hwnd || (ptme->dwFlags & TME_LEAVE && hittest==HTNOWHERE))
         {
             /* Adding new mouse event to the tracking list */
             tracking_info.tme = *ptme;

-- 
2.35.1

