From: Dmitry Timoshkov <dmitry@baikal.ru>
Subject: [PATCH 2/3] comctl32/tests: Add iItem and uChanged to LVN_ITEMCHANGING/LVN_ITEMCHANGED context. (Resend)
Message-Id: <20220125115703.ea0658b4269cf32cc534bf48@baikal.ru>
Date: Tue, 25 Jan 2022 11:57:03 +0300

Signed-off-by: Dmitry Timoshkov <dmitry@baikal.ru>
---
 dlls/comctl32/tests/listview.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/dlls/comctl32/tests/listview.c b/dlls/comctl32/tests/listview.c
index e95b81f5bb1..236a9d14372 100644
--- a/dlls/comctl32/tests/listview.c
+++ b/dlls/comctl32/tests/listview.c
@@ -472,7 +472,17 @@ static LRESULT WINAPI parent_wnd_proc(HWND hwnd, UINT message, WPARAM wParam, LP
     if (defwndproc_counter) msg.flags |= defwinproc;
     msg.wParam = wParam;
     msg.lParam = lParam;
-    if (message == WM_NOTIFY && lParam) msg.id = ((NMHDR*)lParam)->code;
+    if (message == WM_NOTIFY && lParam)
+    {
+        NMLISTVIEW *nmlv = (NMLISTVIEW *)lParam;
+
+        msg.id = nmlv->hdr.code;
+        if (msg.id == LVN_ITEMCHANGING || msg.id == LVN_ITEMCHANGED)
+        {
+            msg.wParam = nmlv->iItem;
+            msg.lParam = nmlv->uChanged;
+        }
+    }
     if (message == WM_COMMAND) msg.id = HIWORD(wParam);
 
     /* log system messages, except for painting */

-- 
2.34.1

