From 287719b3326ded974245b5fda27273880c4fe278 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Fri, 8 Nov 2019 12:26:58 +0100
Subject: [PATCH 1/2] winex11.drv: Only send WM_CANCELMODE if a menu is active.

---
 dlls/winex11.drv/event.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/dlls/winex11.drv/event.c b/dlls/winex11.drv/event.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/event.c
+++ b/dlls/winex11.drv/event.c
@@ -933,6 +933,7 @@ static BOOL X11DRV_FocusIn( HWND hwnd, XEvent *xev )
 static void focus_out( Display *display , HWND hwnd, Time time )
  {
     Window focus_win;
+    GUITHREADINFO threadinfo;
     int revert;
     XIC xic;
 
@@ -947,7 +948,12 @@ static void focus_out( Display *display , HWND hwnd, Time time )
         return;
     }
     if (hwnd != NtUserGetForegroundWindow()) return;
-    send_message( hwnd, WM_CANCELMODE, 0, 0 );
+
+    threadinfo.cbSize = sizeof(threadinfo);
+    NtUserGetGUIThreadInfo(0, &threadinfo);
+
+    if (threadinfo.flags & (GUI_INMENUMODE|GUI_INMOVESIZE|GUI_POPUPMENUMODE|GUI_SYSTEMMENUMODE))
+        send_message( hwnd, WM_CANCELMODE, 0, 0 );
 
     /* don't reset the foreground window, if the window which is
        getting the focus is a Wine window */
-- 
2.36.0

