From 37eddf0bc93d213f1e2b42cc9aa15694ef391309 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Thu, 15 Apr 2021 12:35:12 +0200
Subject: [PATCH 25/25] explorer: Load null graphics driver as a fallback.

---
 programs/explorer/desktop.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/programs/explorer/desktop.c b/programs/explorer/desktop.c
index 2ba276f52cb..a67f79582c2 100644
--- a/programs/explorer/desktop.c
+++ b/programs/explorer/desktop.c
@@ -39,7 +39,7 @@ WINE_DEFAULT_DEBUG_CHANNEL(explorer);
 #define DESKTOP_CLASS_ATOM ((LPCWSTR)MAKEINTATOM(32769))
 #define DESKTOP_ALL_ACCESS 0x01ff
 
-static const WCHAR default_driver[] = {'m','a','c',',','x','1','1',0};
+static const WCHAR default_driver[] = L"mac,x11,null";
 
 static BOOL using_root;
 
@@ -837,6 +837,7 @@ static HMODULE load_graphics_driver( const WCHAR *driver, const GUID *guid )
         {
             TRACE( "display %s using null driver\n", debugstr_guid(guid) );
             null_driver = TRUE;
+            wcscpy( libname, name );
             break;
         }
 
-- 
2.34.0

