From da8306d045bf81c1885dba9a054a597684291a8e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Tue, 11 May 2021 10:51:33 +0200
Subject: [PATCH 16/18] winebus.sys: Add some VID/PID to mouse and keyboard
 devices.

---
 dlls/winebus.sys/main.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/dlls/winebus.sys/main.c b/dlls/winebus.sys/main.c
index 99f834d2f40..86ccf202215 100644
--- a/dlls/winebus.sys/main.c
+++ b/dlls/winebus.sys/main.c
@@ -68,6 +68,10 @@ struct product_desc
     const WCHAR* serialnumber;
 };
 
+#define VID_WINE            0x845e
+#define PID_WINE_MOUSE      0x0001
+#define PID_WINE_KEYBOARD   0x0002
+
 #define VID_MICROSOFT 0x045e
 
 static const WCHAR xbox360_product_string[] = {
@@ -537,6 +541,6 @@ static void mouse_device_create(void)
 static void mouse_device_create(void)
 {
     static const WCHAR busidW[] = {'W','I','N','E','M','O','U','S','E',0};
-    mouse_obj = bus_create_hid_device(busidW, 0, 0, -1, 0, 0, busidW, FALSE, &mouse_vtbl, 0);
+    mouse_obj = bus_create_hid_device(busidW, VID_WINE, PID_WINE_MOUSE, -1, 0, 0, busidW, FALSE, &mouse_vtbl, 0);
     IoInvalidateDeviceRelations(bus_pdo, BusRelations);
 }
 
@@ -605,6 +609,6 @@ static void keyboard_device_create(void)
 static void keyboard_device_create(void)
 {
     static const WCHAR busidW[] = {'W','I','N','E','K','E','Y','B','O','A','R','D',0};
-    keyboard_obj = bus_create_hid_device(busidW, 0, 0, -1, 0, 0, busidW, FALSE, &keyboard_vtbl, 0);
+    keyboard_obj = bus_create_hid_device(busidW, 0, 0, -1, 0, 0, busidW, FALSE, &keyboard_vtbl, 0);
-    keyboard_obj = bus_create_hid_device(busidW, VID_WINE, PID_WINE_KEYBOARD, -1, 0, 0, busidW, FALSE, &keyboard_vtbl, 0);
     IoInvalidateDeviceRelations(bus_pdo, BusRelations);
 }
 

