From da8306d045bf81c1885dba9a054a597684291a8e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Tue, 11 May 2021 10:51:33 +0200
Subject: [PATCH 16/18] winebus.sys: Add some VID/PID to mouse and keyboard
 devices.

---
 dlls/winebus.sys/unixlib.c | 8 ++++
 1 file changed, 8 insertions(+), 0 deletions(-)

diff --git a/dlls/winebus.sys/unixlib.c b/dlls/winebus.sys/unixlib.c
index 2337cb3706d..eda3cdabacb 100644
--- a/dlls/winebus.sys/unixlib.c
+++ b/dlls/winebus.sys/unixlib.c
@@ -37,6 +37,10 @@ WINE_DEFAULT_DEBUG_CHANNEL(plugplay);
 static struct hid_descriptor mouse_desc;
 static struct hid_descriptor keyboard_desc;
 
+#define VID_WINE            0x845e
+#define PID_WINE_MOUSE      0x0001
+#define PID_WINE_KEYBOARD   0x0002
+
 static void mouse_remove(struct unix_device *iface)
 {
 }
@@ -117,6 +121,8 @@ static const WCHAR mouse_bus_id[] = {'W','I','N','E','M','O','U','S','E',0};
 static const struct device_desc mouse_device_desc =
 {
     .busid = mouse_bus_id,
+    .vid = VID_WINE,
+    .pid = PID_WINE_MOUSE,
     .input = -1,
     .serial = {'0','0','0','0',0},
 };
@@ -210,6 +216,8 @@ static const WCHAR keyboard_bus_id[] = {'W','I','N','E','K','E','Y','B','O','A',
 static const struct device_desc keyboard_device_desc =
 {
     .busid = keyboard_bus_id,
+    .vid = VID_WINE,
+    .pid = PID_WINE_KEYBOARD,
     .input = -1,
     .serial = {'0','0','0','0',0},
 };
