From: Paul Gofman <pgofman@codeweavers.com>
Subject: [PATCH] winhttp: Fix filling TIMEVAL in netconn_create().
Message-Id: <20220311174712.1838280-1-pgofman@codeweavers.com>
Date: Fri, 11 Mar 2022 20:47:12 +0300

Signed-off-by: Paul Gofman <pgofman@codeweavers.com>
---
 dlls/winhttp/net.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/winhttp/net.c b/dlls/winhttp/net.c
index 19fafb25804..09ed5e222d1 100644
--- a/dlls/winhttp/net.c
+++ b/dlls/winhttp/net.c
@@ -229,7 +229,7 @@ DWORD netconn_create( struct hostdata *host, const struct sockaddr_storage *sock
         if (ret == WSAEWOULDBLOCK || ret == WSAEINPROGRESS)
         {
             FD_SET set;
-            TIMEVAL timeval = { 0, timeout * 1000 };
+            TIMEVAL timeval = { timeout / 1000, (timeout % 1000) * 1000 };
             int res;
 
             FD_ZERO( &set );

-- 
2.35.1

