From e00faf3b33b58eaeb272c865d7536edc721f7268 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Thu, 9 Dec 2021 19:27:31 +0100
Subject: [PATCH 3/5] winex11.drv: Pass XEvent instead of Display to set_focus.

---
 dlls/winex11.drv/event.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/dlls/winex11.drv/event.c b/dlls/winex11.drv/event.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/event.c
+++ b/dlls/winex11.drv/event.c
@@ -660,8 +660,10 @@ static void set_foreground_window( HWND hwnd )
 /**********************************************************************
  *              set_focus
  */
-static void set_focus( Display *display, Window window, HWND hwnd, Time time, BOOL check )
+static void set_focus( XEvent *xev, HWND hwnd, Time time, BOOL check )
 {
+    Display *display = xev->xany.display;
+    Window window = xev->xany.window;
     HWND focus, old_active;
     GUITHREADINFO threadinfo;
     Time user_time;
@@ -795,7 +797,7 @@ static void handle_wm_protocols( HWND hwnd, XEvent *xev )
                (int)NtUserGetWindowLongW(hwnd, GWL_STYLE),
                get_focus(), get_active_window(), NtUserGetForegroundWindow(), last_focus );
 
-        set_focus( event->display, event->window, hwnd, event_time, TRUE );
+        set_focus( xev, hwnd, event_time, TRUE );
     }
     else if (protocol == x11drv_atom(_NET_WM_PING))
     {
@@ -866,7 +868,7 @@ BOOL x11drv_handle_focus_in_event( HWND hwnd, XEvent *xev, Time time )
 
     if (use_take_focus) return TRUE;
 
-    set_focus( event->display, event->window, hwnd, time, FALSE );
+    set_focus( xev, hwnd, time, FALSE );
     return TRUE;
 }
 
-- 
0.0.0

