From 4aa49e5420efcc50c8cc7fcdcb0a1cb36afd967f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Thu, 9 Dec 2021 19:27:31 +0100
Subject: [PATCH 3/8] winex11.drv: Pass XEvent instead of Display to set_focus.

---
 dlls/winex11.drv/event.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/dlls/winex11.drv/event.c b/dlls/winex11.drv/event.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/event.c
+++ b/dlls/winex11.drv/event.c
@@ -697,11 +697,13 @@ static void set_foreground_window( HWND hwnd )
 /**********************************************************************
  *              set_focus
  */
-static void set_focus( Display *display, Window window, HWND hwnd, Time time, BOOL check )
+static void set_focus( XEvent *xev, HWND hwnd, Time time, BOOL check )
 {
-    HWND focus, old_active;
+    Display *display = xev->xany.display;
+    Window window = xev->xany.window;
     GUITHREADINFO threadinfo;
     Time user_time;
+    HWND focus, old_active;
 
     old_active = NtUserGetForegroundWindow();
 
@@ -831,7 +833,7 @@ static void handle_wm_protocols( HWND hwnd, XEvent *xev )
                hwnd, NtUserIsWindowEnabled(hwnd), NtUserIsWindowVisible(hwnd), NtUserGetWindowLongW(hwnd, GWL_STYLE),
                get_focus(), get_active_window(), NtUserGetForegroundWindow(), last_focus );
 
-        set_focus( event->display, event->window, hwnd, event_time, TRUE );
+        set_focus( xev, hwnd, event_time, TRUE );
     }
     else if (protocol == x11drv_atom(_NET_WM_PING))
     {
@@ -905,7 +907,7 @@ BOOL x11drv_handle_focus_in_event( HWND hwnd, XEvent *xev, Time time )
         return TRUE;
     }
 
-    set_focus( event->display, event->window, hwnd, time, FALSE );
+    set_focus( xev, hwnd, time, FALSE );
     return TRUE;
 }
 
-- 
2.37.1

