From: Chao Long <longchao@uniontech.com>
Subject: [PATCH] winex11.drv: Fix import XA_STRING compatible wide character.
Message-Id: <20211217091744.11043-1-longchao@uniontech.com>
Date: Fri, 17 Dec 2021 17:17:44 +0800

Signed-off-by: Chao Long <longchao@uniontech.com>
---
 dlls/winex11.drv/clipboard.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/winex11.drv/clipboard.c b/dlls/winex11.drv/clipboard.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/clipboard.c
+++ b/dlls/winex11.drv/clipboard.c
@@ -714,7 +714,7 @@ static void *import_string( Atom type, const void *data, size_t size, size_t *re
     WCHAR *ret;
 
     if (!(ret = malloc( (size * 2 + 1) * sizeof(WCHAR) ))) return NULL;
-    str_size = MultiByteToWideChar( 28591, 0, data, size, ret + size, size );
+    str_size = MultiByteToWideChar( CP_UTF8, 0, data, size, ret + size, size );
     return unicode_text_from_string( ret, ret + size, str_size, ret_size );
 }
 

-- 
2.36.0

