From: Giovanni Mascellani <gmascellani@codeweavers.com>
Subject: [PATCH] winex11.drv: Do not call old_error_handler if it is NULL.
Message-Id: <20210212115800.3997498-1-gmascellani@codeweavers.com>
Date: Fri, 12 Feb 2021 12:58:00 +0100

Signed-off-by: Giovanni Mascellani <gmascellani@codeweavers.com>
---
 dlls/winex11.drv/x11drv_main.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/dlls/winex11.drv/x11drv_main.c b/dlls/winex11.drv/x11drv_main.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/x11drv_main.c
+++ b/dlls/winex11.drv/x11drv_main.c
@@ -321,7 +321,10 @@ static int error_handler( Display *display, XErrorEvent *error_evt )
              error_evt->serial, error_evt->request_code );
         DebugBreak();  /* force an entry in the debugger */
     }
-    old_error_handler( display, error_evt );
+    /* Xlib never returns a NULL previous error handler, but other
+       implementations, like Steam's overlay, do. */
+    if (old_error_handler)
+        old_error_handler( display, error_evt );
     return 0;
 }
 

-- 
2.35.3

