From: Chao Long <longchao@uniontech.com>
Subject: [PATCH] win32u: Fix event queue congestion can cause WINPOS_WindowFromPoint is null.
Message-Id: <20211021033122.18818-1-longchao@uniontech.com>
Date: Thu, 21 Oct 2021 11:31:22 +0800

Signed-off-by: Chao Long <longchao@uniontech.com>
---
 dlls/win32u/input.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/win32u/input.c b/dlls/win32u/input.c
index 11111111111..11111111111 100644
--- a/dlls/win32u/input.c
+++ b/dlls/win32u/input.c
@@ -1360,7 +1360,7 @@ BOOL WINAPI NtUserTrackMouseEvent( TRACKMOUSEEVENT *info )
         tracking_info.info.dwFlags = 0;
         tracking_info.info.dwHoverTime = 0;
 
-        if (info->hwndTrack == hwnd)
+        if (info->hwndTrack == hwnd || (info->dwFlags & TME_LEAVE && hittest==HTNOWHERE))
         {
             /* Adding new mouse event to the tracking list */
             tracking_info.info = *info;

-- 
2.36.0

