From: Timo Zuccarello <timo@zuccarello.eu>
Subject: [PATCH] dlls/winex11.drv: Do not assign reserved virtual key 0xE0 to key codes.
Message-Id: <20220413140418.2416644-1-timo@zuccarello.eu>
Date: Wed, 13 Apr 2022 16:04:18 +0200

Signed-off-by: Timo Zuccarello <timo@zuccarello.eu>
---
 dlls/winex11.drv/keyboard.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/winex11.drv/keyboard.c b/dlls/winex11.drv/keyboard.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/keyboard.c
+++ b/dlls/winex11.drv/keyboard.c
@@ -1676,7 +1676,8 @@ void X11DRV_InitKeyboard( Display *display )
         WORD first, last;
     } vkey_ranges[] = {
         { VK_OEM_1, VK_OEM_3 },
-        { VK_OEM_4, VK_ICO_00 },
+        { VK_OEM_4, VK_OEM_8 },
+        { VK_OEM_AX, VK_ICO_00 },
         { 0xe6, 0xe6 },
         { 0xe9, 0xf5 },
         { VK_OEM_NEC_EQUAL, VK_OEM_NEC_EQUAL },

-- 
2.35.3

