From 39394ff982fe4e606be0048f3733bb54a03cc992 Mon Sep 17 00:00:00 2001
From: Zsolt Vadasz <zsolt_vadasz@protonmail.com>
Date: Wed, 26 Jul 2023 23:09:35 +0200
Subject: [PATCH 2/2] win32u: Allow fullscreen clipping when using virtual
 desktop

---
 dlls/win32u/input.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/dlls/win32u/input.c b/dlls/win32u/input.c
index 11111111111..11111111111 100644
--- a/dlls/win32u/input.c
+++ b/dlls/win32u/input.c
@@ -2564,7 +2564,6 @@ BOOL clip_fullscreen_window( HWND hwnd, BOOL reset )
     {
         RECT virtual_rect = NtUserGetVirtualScreenRect();
         if (!EqualRect( &monitor_info.rcMonitor, &virtual_rect )) return FALSE;
-        if (is_virtual_desktop()) return FALSE;
     }
 
     TRACE( "win %p clipping fullscreen\n", hwnd );
-- 
0.0.0

