From: Jinoh Kang <jinoh.kang.kr@gmail.com>
Subject: [PATCH] comctl32: Fix integer overflow in DPA_Search.
Message-Id: <67bac834-f8e3-c80b-dbb4-8d8d24d70491@gmail.com>
Date: Sat, 12 Mar 2022 20:57:13 +0900

Signed-off-by: Jinoh Kang <jinoh.kang.kr@gmail.com>
---
 dlls/comctl32/dpa.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/comctl32/dpa.c b/dlls/comctl32/dpa.c
index 11111111111..11111111111 100644
--- a/dlls/comctl32/dpa.c
+++ b/dlls/comctl32/dpa.c
@@ -868,7 +868,7 @@ INT WINAPI DPA_Search (HDPA hdpa, LPVOID pFind, INT nStart,
         r = hdpa->nItemCount - 1;
         lpPtr = hdpa->ptrs;
         while (r >= l) {
-            x = (l + r) / 2;
+            x = l + (r - l) / 2;
             n = (pfnCompare)(pFind, lpPtr[x], lParam);
             if (n == 0)
                 return x;

-- 
2.36.1

