From: Jinoh Kang <jinoh.kang.kr@gmail.com>
Subject: [PATCH] user32: Recognise RichEdit50W window class in spy.
Message-Id: <be147d41-713d-162e-561f-886a60a9e113@gmail.com>
Date: Sat, 12 Mar 2022 19:36:01 +0900

Signed-off-by: Jinoh Kang <jinoh.kang.kr@gmail.com>
---
 dlls/user32/spy.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/user32/spy.c b/dlls/user32/spy.c
index dc8e0b2a708..4d22acb43c2 100644
--- a/dlls/user32/spy.c
+++ b/dlls/user32/spy.c
@@ -1764,6 +1764,7 @@ static const CONTROL_CLASS cc_array[] = {
     {TOOLTIPS_CLASSW,   tooltips_array, ARRAY_SIZE(tooltips_array)},
     {UPDOWN_CLASSW,     updown_array,   ARRAY_SIZE(updown_array)},
     {RICHEDIT_CLASS20W, richedit_array, ARRAY_SIZE(richedit_array)},
+    {MSFTEDIT_CLASS,    richedit_array, ARRAY_SIZE(richedit_array)},
     {0, 0, 0}
 };
 

-- 
2.34.1

