From 1b28bde233159e775a20f294a39275adfd5a2d22 Mon Sep 17 00:00:00 2001
From: Zebediah Figura <z.figura12@gmail.com>
Date: Sat, 13 Mar 2021 16:20:30 -0600
Subject: [PATCH 27/28] server: Allow disabling fast synchronization support.

---
 server/fast_sync.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/server/fast_sync.c b/server/fast_sync.c
index bcc9493cbfa..ea4aa666cf3 100644
--- a/server/fast_sync.c
+++ b/server/fast_sync.c
@@ -129,6 +129,12 @@ static struct fast_sync_device *get_fast_sync_device(void)
     struct fast_sync_device *device;
     int unix_fd;
 
+    if (getenv( "WINE_DISABLE_FAST_SYNC" ) && atoi( getenv( "WINE_DISABLE_FAST_SYNC" ) ))
+    {
+        set_error( STATUS_NOT_IMPLEMENTED );
+        return NULL;
+    }
+
     if (fast_sync_device_object)
         return (struct fast_sync_device *)grab_object( fast_sync_device_object );
 
-- 
2.31.1

