From 9cb62bbd95fd782825772d65e568fb1baa667c50 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Thu, 28 Jan 2021 14:05:48 +0100
Subject: [PATCH] winex11.drv: Fix focus delay issues with desktop clipping.

---
 dlls/winex11.drv/event.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/winex11.drv/event.c b/dlls/winex11.drv/event.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/event.c
+++ b/dlls/winex11.drv/event.c
@@ -335,6 +335,7 @@ static int try_grab_pointer( Display *display )
         return 0;
 
     XUngrabPointer( display, CurrentTime );
+    XFlush( display );
     return 1;
 }
 
