diff --git a/dlls/mmdevapi/main.c b/dlls/mmdevapi/main.c
index eac1da28f18..1e5bc87753d 100644
--- a/dlls/mmdevapi/main.c
+++ b/dlls/mmdevapi/main.c
@@ -112,19 +112,20 @@ static BOOL WINAPI init_driver(INIT_ONCE *once, void *param, void **context)
     if(RegOpenKeyW(HKEY_CURRENT_USER, drv_keyW, &key) == ERROR_SUCCESS){
         DWORD size = sizeof(reg_list);
 
-        if(RegQueryValueExW(key, L"Audio", 0, NULL, (BYTE*)reg_list, &size) == ERROR_SUCCESS){
-            if(reg_list[0] == '\0'){
-                TRACE("User explicitly chose no driver\n");
-                RegCloseKey(key);
-                return TRUE;
-            }
-
+        if(RegQueryValueExW(key, L"Audio", 0, NULL, (BYTE*)reg_list, &size) == ERROR_SUCCESS)
             driver_list = reg_list;
-        }
 
         RegCloseKey(key);
     }
 
+    if ((p = _wgetenv(L"STAGING_AUDIO_DRIVER")))
+        driver_list = p;
+
+    if(driver_list[0] == '\0'){
+        TRACE("User explicitly chose no driver\n");
+        return TRUE;
+    }
+
     TRACE("Loading driver list %s\n", wine_dbgstr_w(driver_list));
     for(next = p = driver_list; next; p = next + 1){
         next = wcschr(p, ',');
