Made by Richard Yao.

diff --git a/dlls/iphlpapi/ifenum.c b/dlls/iphlpapi/ifenum.c
index c3abdf0e536..8fc79e8feb4 100644
--- a/dlls/iphlpapi/ifenum.c
+++ b/dlls/iphlpapi/ifenum.c
@@ -169,9 +169,13 @@ BOOL isIfIndexLoopback(ULONG idx)
 DWORD get_interface_indices( BOOL skip_loopback, InterfaceIndexTable **table )
 {
     DWORD count = 0, i;
-    struct if_nameindex *p, *indices = if_nameindex();
+    static struct if_nameindex *indices = NULL;
+    struct if_nameindex *p;
     InterfaceIndexTable *ret;
 
+    if (indices == NULL)
+      indices = if_nameindex();
+
     if (table) *table = NULL;
     if (!indices) return 0;
 
@@ -199,7 +203,6 @@ DWORD get_interface_indices( BOOL skip_loopback, InterfaceIndexTable **table )
     }
 
 end:
-    if_freenameindex( indices );
     return count;
 }
 
@@ -817,8 +820,23 @@ static DWORD getIPAddrRowByName(PMIB_IPADDRROW ipAddrRow, const char *ifName,
 
 #if defined(HAVE_IFADDRS_H) && defined(HAVE_GETIFADDRS)
 
+int getifaddrs_cached(struct ifaddrs **ifap) {
+  static struct ifaddrs *ifa = NULL;
+
+  if (ifa == NULL)
+    getifaddrs(&ifa);
+
+  *ifap = ifa;
+  return 0;
+}
+
+void freeifaddrs_cached(struct ifaddrs *ifa)
+{
+  return;
+}
+
 /* Counts the IPv4 addresses in the system using the return value from
- * getifaddrs, returning the count.
+ * getifaddrs_cached, returning the count.
  */
 static DWORD countIPv4Addresses(struct ifaddrs *ifa)
 {
@@ -835,10 +853,10 @@ DWORD getNumIPAddresses(void)
   DWORD numAddresses = 0;
   struct ifaddrs *ifa;
 
-  if (!getifaddrs(&ifa))
+  if (!getifaddrs_cached(&ifa))
   {
     numAddresses = countIPv4Addresses(ifa);
-    freeifaddrs(ifa);
+    freeifaddrs_cached(ifa);
   }
   return numAddresses;
 }
@@ -853,7 +871,7 @@ DWORD getIPAddrTable(PMIB_IPADDRTABLE *ppIpAddrTable, HANDLE heap, DWORD flags)
   {
     struct ifaddrs *ifa;
 
-    if (!getifaddrs(&ifa))
+    if (!getifaddrs_cached(&ifa))
     {
       DWORD size = sizeof(MIB_IPADDRTABLE);
       DWORD numAddresses = countIPv4Addresses(ifa);
@@ -882,7 +900,7 @@ DWORD getIPAddrTable(PMIB_IPADDRTABLE *ppIpAddrTable, HANDLE heap, DWORD flags)
       }
       else
         ret = ERROR_OUTOFMEMORY;
-      freeifaddrs(ifa);
+      freeifaddrs_cached(ifa);
     }
     else
       ret = ERROR_INVALID_PARAMETER;
@@ -895,7 +913,7 @@ ULONG v6addressesFromIndex(IF_INDEX index, SOCKET_ADDRESS **addrs, ULONG *num_ad
   struct ifaddrs *ifa;
   ULONG ret;
 
-  if (!getifaddrs(&ifa))
+  if (!getifaddrs_cached(&ifa))
   {
     struct ifaddrs *p;
     ULONG n;
@@ -966,7 +984,7 @@ ULONG v6addressesFromIndex(IF_INDEX index, SOCKET_ADDRESS **addrs, ULONG *num_ad
       *masks = NULL;
       ret = ERROR_SUCCESS;
     }
-    freeifaddrs(ifa);
+    freeifaddrs_cached(ifa);
   }
   else
     ret = ERROR_NO_DATA;
