diff --git a/dlls/win32u/imm.c b/dlls/win32u/imm.c
index 11111111111..11111111111 100644
--- a/dlls/win32u/imm.c
+++ b/dlls/win32u/imm.c
@@ -51,7 +51,7 @@ struct imm_thread_data
 
 static struct list thread_data_list = LIST_INIT( thread_data_list );
 static pthread_mutex_t imm_mutex = PTHREAD_MUTEX_INITIALIZER;
-static BOOL disable_ime;
+static BOOL disable_ime = TRUE;
 
 static struct imc *get_imc_ptr( HIMC handle )
 {
