diff --git a/dlls/imm32/imm.c b/dlls/imm32/imm.c
index 11111111111..11111111111 100644
--- a/dlls/imm32/imm.c
+++ b/dlls/imm32/imm.c
@@ -123,7 +123,7 @@ static CRITICAL_SECTION_DEBUG critsect_debug =
       0, 0, { (DWORD_PTR)(__FILE__ ": threaddata_cs") }
 };
 static CRITICAL_SECTION threaddata_cs = { &critsect_debug, -1, 0, 0, 0, 0 };
-static BOOL disable_ime;
+static BOOL disable_ime = TRUE;
 
 static inline BOOL is_himc_ime_unicode(const InputContextData *data)
 {
