diff --git a/dlls/win32u/sysparams.c b/dlls/win32u/sysparams.c
index f3b4e0d3b13..7405bb928cb 100644
--- a/dlls/win32u/sysparams.c
+++ b/dlls/win32u/sysparams.c
@@ -1335,7 +1335,7 @@ static void clear_display_devices(void)
 static BOOL update_display_cache_from_registry(void)
 {
     DWORD adapter_id, monitor_id, monitor_count = 0, size;
-    KEY_FULL_INFORMATION key;
+    KEY_BASIC_INFORMATION key;
     struct adapter *adapter;
     struct monitor *monitor;
     HANDLE mutex = NULL;
@@ -1347,7 +1347,7 @@ static BOOL update_display_cache_from_registry(void)
                                                   sizeof(devicemap_video_keyW) )))
         return FALSE;
 
-    status = NtQueryKey( video_key, KeyFullInformation, &key, sizeof(key), &size );
+    status = NtQueryKey( video_key, KeyBasicInformation, &key, sizeof(key), &size );
     if (status && status != STATUS_BUFFER_OVERFLOW)
         return FALSE;
 
