diff --git a/dlls/win32u/sysparams.c b/dlls/win32u/sysparams.c
index 11111111111..11111111111 100644
--- a/dlls/win32u/sysparams.c
+++ b/dlls/win32u/sysparams.c
@@ -1283,7 +1283,7 @@ static ULONGLONG get_tick_count(void)
 static BOOL update_display_cache_from_registry(void)
 {
     DWORD adapter_id, monitor_id, monitor_count = 0, size;
-    KEY_FULL_INFORMATION key;
+    KEY_BASIC_INFORMATION key;
     struct adapter *adapter;
     struct monitor *monitor;
     ULONGLONG tick_count;
@@ -1298,7 +1298,7 @@ static BOOL update_display_cache_from_registry(void)
 
     if ((tick_count = get_tick_count()) - last_update < 1000) return TRUE;
 
-    status = NtQueryKey( video_key, KeyFullInformation, &key, sizeof(key), &size );
+    status = NtQueryKey( video_key, KeyBasicInformation, &key, offsetof(KEY_BASIC_INFORMATION, Name[0]), &size );
     if (status && status != STATUS_BUFFER_OVERFLOW)
         return FALSE;
 
