diff --git a/dlls/win32u/sysparams.c b/dlls/win32u/sysparams.c
index c65f4c10afa..8410c83b075 100644
--- a/dlls/win32u/sysparams.c
+++ b/dlls/win32u/sysparams.c
@@ -1207,7 +1204,7 @@ static void clear_display_devices(void)
 static BOOL update_display_cache_from_registry(void)
 {
     DWORD adapter_id, monitor_id, monitor_count = 0, size;
-    KEY_FULL_INFORMATION key;
+    KEY_BASIC_INFORMATION key;
     struct adapter *adapter;
     struct monitor *monitor;
     HANDLE mutex = NULL;
@@ -1219,7 +1216,7 @@ static BOOL update_display_cache_from_registry(void)
                                                   sizeof(devicemap_video_keyW) )))
         return FALSE;
 
-    status = NtQueryKey( video_key, KeyFullInformation, &key, sizeof(key), &size );
+    status = NtQueryKey( video_key, KeyBasicInformation, &key, sizeof(key), &size );
     if (status && status != STATUS_BUFFER_OVERFLOW)
         return FALSE;
 
