diff --git a/dlls/appwiz.cpl/addons.c b/dlls/appwiz.cpl/addons.c
index 11111111111..11111111111 100644
--- a/dlls/appwiz.cpl/addons.c
+++ b/dlls/appwiz.cpl/addons.c
@@ -78,6 +78,7 @@ typedef struct {
     const char *dir_config_key;
     LPCWSTR dialog_template;
     const char *cache_dir_env_var;
+    const char *control_env_var;
 } addon_info_t;
 
 /* Download addon files over HTTP because Wine depends on an external library
@@ -92,7 +93,8 @@ static const addon_info_t addons_info[] = {
         "http://source.winehq.org/winegecko.php",
         "MSHTML", "GeckoUrl", "GeckoCabDir",
         MAKEINTRESOURCEW(ID_DWL_GECKO_DIALOG),
-        "WINE_GECKO_CACHE_DIR"
+        "WINE_GECKO_CACHE_DIR",
+        "WINE_SKIP_GECKO_INSTALLATION"
     },
     {
         MONO_VERSION,
@@ -102,7 +104,8 @@ static const addon_info_t addons_info[] = {
         "http://source.winehq.org/winemono.php",
         "Dotnet", "MonoUrl", "MonoCabDir",
         MAKEINTRESOURCEW(ID_DWL_MONO_DIALOG),
-        "WINE_MONO_CACHE_DIR"
+        "WINE_MONO_CACHE_DIR",
+        "WINE_SKIP_MONO_INSTALLATION"
     }
 };
 
@@ -754,11 +757,16 @@ static INT_PTR CALLBACK installer_proc(HWND hwnd, UINT msg, WPARAM wParam, LPARA
 
 BOOL install_addon(addon_t addon_type)
 {
+    const char *envar_content;
+
     if(!*GECKO_ARCH)
         return FALSE;
 
     addon = addons_info+addon_type;
 
+    if ((envar_content = getenv(addon->control_env_var)) && atoi(envar_content) != 0)
+        return FALSE;
+
     p_wine_get_dos_file_name = (void *)GetProcAddress(GetModuleHandleW(L"kernel32.dll"), "wine_get_dos_file_name");
 
     /*
