diff --git a/dlls/ntdll/loader.c b/dlls/ntdll/loader.c
index 11111111111..11111111111 100644
--- a/dlls/ntdll/loader.c
+++ b/dlls/ntdll/loader.c
@@ -3177,7 +3177,7 @@ static NTSTATUS find_builtin_without_file( const WCHAR *name, UNICODE_STRING *ne
         len = new_name->Length;
         RtlAppendUnicodeToString( new_name, L"\\dlls\\" );
         RtlAppendUnicodeToString( new_name, name );
-        if ((ext = wcsrchr( name, '.' )) && !wcscmp( ext, L".dll" )) new_name->Length -= 4 * sizeof(WCHAR);
+        if ((ext = wcsrchr( name, '.' )) && !wcsicmp( ext, L".dll" )) new_name->Length -= 4 * sizeof(WCHAR);
         RtlAppendUnicodeToString( new_name, L"\\" );
         RtlAppendUnicodeToString( new_name, name );
         status = open_dll_file( new_name, pwm, mapping, image_info, id );
#@@ -2649,6 +2812,8 @@ static NTSTATUS search_dll_file( LPCWSTR paths, LPCWSTR search, UNICODE_STRING *
#         status = STATUS_IMAGE_MACHINE_TYPE_MISMATCH;
#     else if (!wcspbrk( search, L":/\\" ))
#         status = find_builtin_without_file( search, nt_name, pwm, mapping, image_info, id );
#+    else if (wcslen( search ) > sizeof(system_dir) / sizeof(WCHAR) - 1 && !_wcsnicmp( search, system_dir, sizeof(system_dir) / sizeof(WCHAR) - 1 ))
#+        status = find_builtin_without_file( search + sizeof(system_dir) / sizeof(WCHAR) - 1, nt_name, pwm, mapping, image_info, id );
# 
# done:
#     RtlFreeHeap( GetProcessHeap(), 0, name );
#@@ -2717,6 +2882,14 @@ static NTSTATUS find_dll_file( const WCHAR *load_path, const WCHAR *libname, con
#     else if (!(status = RtlDosPathNameToNtPathName_U_WithStatus( libname, nt_name, NULL, NULL )))
#         status = open_dll_file( nt_name, pwm, mapping, image_info, id );
# 
#+    if (status == STATUS_DLL_NOT_FOUND)
#+    {
#+        if (!wcspbrk( libname, L":/\\" ))
#+            status = find_builtin_without_file( libname, nt_name, pwm, mapping, image_info, id );
#+        else if (wcslen( libname ) > sizeof(system_dir) / sizeof(WCHAR) - 1 && !_wcsnicmp( libname, system_dir, sizeof(system_dir) / sizeof(WCHAR) - 1 ))
#+            status = find_builtin_without_file( libname + sizeof(system_dir) / sizeof(WCHAR) - 1, nt_name, pwm, mapping, image_info, id );
#+    }
#+
#     if (status == STATUS_IMAGE_MACHINE_TYPE_MISMATCH) status = STATUS_INVALID_IMAGE_FORMAT;
# 
# done:
