diff --git a/dlls/mountmgr.sys/device.c b/dlls/mountmgr.sys/device.c
index 11111111111..11111111111 100644
--- a/dlls/mountmgr.sys/device.c
+++ b/dlls/mountmgr.sys/device.c
@@ -1651,7 +1651,7 @@ static NTSTATUS query_property( struct disk_device *device, IRP *irp )
 
 static NTSTATUS WINAPI fill_volume_info( FS_INFORMATION_CLASS info_class, enum fs_type fs_type,
                                          DWORD serial, WCHAR *label, void *buff, ULONG length,
-                                         ULONG_PTR *size )
+                                         ULONG_PTR *size, struct volume *volume )
 {
     NTSTATUS status;
 
@@ -1753,7 +1753,7 @@ NTSTATUS query_volume( void *buff, SIZE_T insize, SIZE_T outsize, IO_STATUS_BLOC
     if (volume)
     {
         status = fill_volume_info( info_class, volume->fs_type, volume->serial, volume->label,
-                                   buff, outsize, &iosb->Information );
+                                   buff, outsize, &iosb->Information, volume );
         release_volume( volume );
     }
     LeaveCriticalSection( &device_section );
@@ -1776,7 +1776,7 @@ NTSTATUS query_volume( void *buff, SIZE_T insize, SIZE_T outsize, IO_STATUS_BLOC
         CloseHandle( handle );
         if (status) return status;
         status = fill_volume_info( info_class, fs_type, 0, NULL, buff, outsize,
-                                   &iosb->Information );
+                                   &iosb->Information, volume );
     }
 
     return status;
@@ -1799,7 +1799,7 @@ static NTSTATUS WINAPI harddisk_query_volume( DEVICE_OBJECT *device, IRP *irp )
     volume = dev->volume;
     if (volume)
         status = fill_volume_info( info_class, volume->fs_type, volume->serial, volume->label,
-                                   buff, length, &io->Information );
+                                   buff, length, &io->Information, volume );
 
     io->u.Status = status;
     LeaveCriticalSection( &device_section );
