diff --git a/dlls/user32/sysparams.c b/dlls/user32/sysparams.c
index daff3ec40ac..11df7817166 100644
--- a/dlls/user32/sysparams.c
+++ b/dlls/user32/sysparams.c
@@ -3781,7 +3781,7 @@ BOOL WINAPI EnumDisplayDevicesW( LPCWSTR device, DWORD index, DISPLAY_DEVICEW *i
 {
     UNICODE_STRING str;
     RtlInitUnicodeString( &str, device );
-    return NtUserEnumDisplayDevices( &str, index, info, flags );
+    return NT_SUCCESS(NtUserEnumDisplayDevices( &str, index, info, flags ));
 }
 
 /**********************************************************************
diff --git a/dlls/win32u/sysparams.c b/dlls/win32u/sysparams.c
index e5b2036336d..d9603a3c6cd 100644
--- a/dlls/win32u/sysparams.c
+++ b/dlls/win32u/sysparams.c
@@ -1418,8 +1418,8 @@ static struct adapter *find_adapter( UNICODE_STRING *name )
 /***********************************************************************
  *	     NtUserEnumDisplayDevices    (win32u.@)
  */
-BOOL WINAPI NtUserEnumDisplayDevices( UNICODE_STRING *device, DWORD index,
-                                      DISPLAY_DEVICEW *info, DWORD flags )
+NTSTATUS WINAPI NtUserEnumDisplayDevices( UNICODE_STRING *device, DWORD index,
+                                          DISPLAY_DEVICEW *info, DWORD flags )
 {
     struct display_device *found = NULL;
     struct adapter *adapter;
@@ -1427,7 +1427,7 @@ BOOL WINAPI NtUserEnumDisplayDevices( UNICODE_STRING *device, DWORD index,
 
     TRACE( "%s %u %p %#x\n", debugstr_us( device ), index, info, flags );
 
-    if (!lock_display_devices()) return FALSE;
+    if (!lock_display_devices()) return STATUS_NOT_FOUND;
 
     if (!device || !device->Length)
     {
@@ -1469,7 +1469,7 @@ BOOL WINAPI NtUserEnumDisplayDevices( UNICODE_STRING *device, DWORD index,
             lstrcpyW( info->DeviceKey, found->device_key );
     }
     unlock_display_devices();
-    return !!found;
+    return found ? STATUS_SUCCESS : STATUS_NOT_FOUND;
 }
 
 #define _X_FIELD(prefix, bits)                              \
diff --git a/dlls/win32u/win32u_private.h b/dlls/win32u/win32u_private.h
index 298e2171212..a63f41b75ee 100644
--- a/dlls/win32u/win32u_private.h
+++ b/dlls/win32u/win32u_private.h
@@ -200,7 +200,7 @@ struct unix_funcs
     LONG     (WINAPI *pNtUserChangeDisplaySettings)( UNICODE_STRING *devname, DEVMODEW *devmode, HWND hwnd,
                                                      DWORD flags, void *lparam );
     INT      (WINAPI *pNtUserCountClipboardFormats)(void);
-    BOOL     (WINAPI *pNtUserEnumDisplayDevices)( UNICODE_STRING *device, DWORD index,
+    NTSTATUS (WINAPI *pNtUserEnumDisplayDevices)( UNICODE_STRING *device, DWORD index,
                                                   DISPLAY_DEVICEW *info, DWORD flags );
     BOOL     (WINAPI *pNtUserEnumDisplayMonitors)( HDC hdc, RECT *rect, MONITORENUMPROC proc, LPARAM lp );
     BOOL     (WINAPI *pNtUserEnumDisplaySettings)( UNICODE_STRING *device, DWORD mode,
diff --git a/dlls/win32u/wrappers.c b/dlls/win32u/wrappers.c
index 63432f13ca7..b4413fd1a23 100644
--- a/dlls/win32u/wrappers.c
+++ b/dlls/win32u/wrappers.c
@@ -622,8 +622,8 @@ INT WINAPI NtUserCountClipboardFormats(void)
     return unix_funcs->pNtUserCountClipboardFormats();
 }
 
-BOOL WINAPI NtUserEnumDisplayDevices( UNICODE_STRING *device, DWORD index,
-                                      DISPLAY_DEVICEW *info, DWORD flags )
+NTSTATUS WINAPI NtUserEnumDisplayDevices( UNICODE_STRING *device, DWORD index,
+                                          DISPLAY_DEVICEW *info, DWORD flags )
 {
     return unix_funcs->pNtUserEnumDisplayDevices( device, index, info, flags );
 }
diff --git a/include/ntuser.h b/include/ntuser.h
index 1d1bcf5d088..22ee29ef200 100644
--- a/include/ntuser.h
+++ b/include/ntuser.h
@@ -105,8 +105,8 @@ HDESK   WINAPI NtUserCreateDesktopEx( OBJECT_ATTRIBUTES *attr, UNICODE_STRING *d
                                       ULONG heap_size );
 HWINSTA WINAPI NtUserCreateWindowStation( OBJECT_ATTRIBUTES *attr, ACCESS_MASK mask, ULONG arg3,
                                           ULONG arg4, ULONG arg5, ULONG arg6, ULONG arg7 );
-BOOL    WINAPI NtUserEnumDisplayDevices( UNICODE_STRING *device, DWORD index,
-                                         DISPLAY_DEVICEW *info, DWORD flags );
+NTSTATUS WINAPI NtUserEnumDisplayDevices( UNICODE_STRING *device, DWORD index,
+                                          DISPLAY_DEVICEW *info, DWORD flags );
 BOOL    WINAPI NtUserEnumDisplayMonitors( HDC hdc, RECT *rect, MONITORENUMPROC proc, LPARAM lp );
 BOOL    WINAPI NtUserEnumDisplaySettings( UNICODE_STRING *device, DWORD mode,
                                           DEVMODEW *dev_mode, DWORD flags );
