diff --git a/server/handle.c b/server/handle.c
index 11111111111..11111111111 100644
--- a/server/handle.c
+++ b/server/handle.c
@@ -478,6 +478,9 @@ unsigned int close_handle( struct process *process, obj_handle_t handle )
 /* retrieve the object corresponding to one of the magic pseudo-handles */
 static inline struct object *get_magic_handle( obj_handle_t handle )
 {
+    if (!(handle & 0xc0000000))
+        return NULL;
+
     switch(handle)
     {
         case 0xfffffffa:  /* current thread impersonation token pseudo-handle */
