diff --git a/tools/makedep.c b/tools/makedep.c
index 11111111111..11111111111 100644
--- a/tools/makedep.c
+++ b/tools/makedep.c
@@ -3164,6 +3164,7 @@ static void output_source_one_arch( struct makefile *make, struct incl_file *sou
 
     output_filenames( cpp_flags );
     output_filename( arch_make_variable( "CFLAGS", arch ));
+    output_filename( arch_make_variable( strmake( "%s_CFLAGS", basename( obj ) ), arch ));
     output( "\n" );
 
     if (make->testdll && !is_dll_src && strendswith( source->name, ".c" ) &&
