diff --git a/programs/wineboot/wineboot.c b/programs/wineboot/wineboot.c
index 25b09d765f1..6d3ae7cf592 100644
--- a/programs/wineboot/wineboot.c
+++ b/programs/wineboot/wineboot.c
@@ -312,6 +312,8 @@ static UINT64 read_tsc_frequency(void)
         else
         {
             freq = (freq0 + freq1) / 2;
+            freq += 5000000 / 2;
+            freq = freq - freq % 5000000;
             TRACE("TSC frequency calibration complete, found %I64u Hz\n", freq);
         }
     }
