diff --git a/dlls/ntdll/reg.c b/dlls/ntdll/reg.c
index 11111111111..11111111111 100644
--- a/dlls/ntdll/reg.c
+++ b/dlls/ntdll/reg.c
@@ -93,7 +93,7 @@ NTSTATUS WINAPI RtlpNtEnumerateSubKey( HANDLE handle, UNICODE_STRING *out, ULONG
 
   if (out->Length)
   {
-    dwLen = out->Length + sizeof(KEY_BASIC_INFORMATION);
+    dwLen = out->Length + offsetof(KEY_BASIC_INFORMATION, Name[0]);
     info = RtlAllocateHeap( GetProcessHeap(), 0, dwLen );
     if (!info)
       return STATUS_NO_MEMORY;
@@ -105,7 +105,7 @@ NTSTATUS WINAPI RtlpNtEnumerateSubKey( HANDLE handle, UNICODE_STRING *out, ULONG
   }
 
   ret = NtEnumerateKey( handle, index, KeyBasicInformation, info, dwLen, &dwResultLen );
-  dwResultLen -= sizeof(KEY_BASIC_INFORMATION);
+  dwResultLen -= offsetof(KEY_BASIC_INFORMATION, Name[0]);
 
   if (ret == STATUS_BUFFER_OVERFLOW)
     out->Length = dwResultLen;
