diff --git a/dlls/winex11.drv/window.c b/dlls/winex11.drv/window.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/window.c
+++ b/dlls/winex11.drv/window.c
@@ -2290,7 +2290,7 @@ BOOL X11DRV_CreateWindow( HWND hwnd )
     if (hwnd == NtUserGetDesktopWindow())
     {
         struct x11drv_thread_data *data = x11drv_init_thread_data();
-        XSetWindowAttributes attr;
+        XSetWindowAttributes attr = {0};
 
         if (NtUserGetWindowThread( NtUserGetDesktopWindow(), NULL ) == GetCurrentThreadId())
         {
