Modified version of a patch made by gonX.

diff --git a/dlls/winepulse.drv/pulse.c b/dlls/winepulse.drv/pulse.c
index 0360d62f471..d32caf7fb74 100644
--- a/dlls/winepulse.drv/pulse.c
+++ b/dlls/winepulse.drv/pulse.c
@@ -858,11 +858,19 @@ static void pulse_probe_settings(pa_mainloop *ml, pa_context *ctx, int render, W
     if (length)
         pulse_def_period[!render] = pulse_min_period[!render] = pa_bytes_to_usec(10 * length, &ss);
 
-    if (pulse_min_period[!render] < MinimumPeriod)
-        pulse_min_period[!render] = MinimumPeriod;
+    const char* penv = getenv("STAGING_AUDIO_PERIOD");
+    int val;
+    if (!penv) {
+        if (pulse_min_period[!render] < MinimumPeriod)
+            pulse_min_period[!render] = MinimumPeriod;
 
-    if (pulse_def_period[!render] < DefaultPeriod)
-        pulse_def_period[!render] = DefaultPeriod;
+        if (pulse_def_period[!render] < DefaultPeriod)
+            pulse_def_period[!render] = DefaultPeriod;
+    }
+	 else if ((val = atoi(penv)) >= 0) {
+        pulse_def_period[!render] = pulse_min_period[!render] = val;
+        printf("Audio period set to %d.\n", val);
+    }
 
     wfx->wFormatTag = WAVE_FORMAT_EXTENSIBLE;
     wfx->cbSize = sizeof(WAVEFORMATEXTENSIBLE) - sizeof(WAVEFORMATEX);
@@ -1260,6 +1268,13 @@ static NTSTATUS pulse_create_stream(void *args)
     if (duration < 3 * period)
         duration = 3 * period;
 
+    const char* denv = getenv("STAGING_AUDIO_DURATION");
+    if (denv) {
+        int val = atoi(denv);
+        duration = val;
+        printf("Audio duration set to %d.\n", val);
+    }
+
     stream->period_bytes = pa_frame_size(&stream->ss) * muldiv(period, stream->ss.rate, 10000000);
 
     stream->bufsize_frames = ceil((duration / 10000000.) * params->fmt->nSamplesPerSec);
