diff --git a/server/Makefile.in b/server/Makefile.in
index 739d0517339..267175afce6 100644
--- a/server/Makefile.in
+++ b/server/Makefile.in
@@ -49,6 +53,7 @@ MANPAGES = \
 	wineserver.man.in
 
 EXTRAINCL = $(DBUS_CFLAGS)
-EXTRALIBS = $(LDEXECFLAGS) $(RT_LIBS) $(INOTIFY_LIBS) $(PROCSTAT_LIBS) $(DBUS_LIBS)
+EXTRALIBS = $(LDEXECFLAGS) $(RT_LIBS) $(INOTIFY_LIBS) $(PROCSTAT_LIBS) $(DBUS_LIBS) -flto=auto -flto-partition=one -fdevirtualize-at-ltrans
+EXTRADEFS = -flto=auto -flto-partition=one -fdevirtualize-at-ltrans
 
 unicode_EXTRADEFS = -DNLSDIR="\"${nlsdir}\"" -DBIN_TO_NLSDIR=\"`${MAKEDEP} -R ${bindir} ${nlsdir}`\"
