diff --git a/server/object.c b/server/object.c
index 030478073ec..770dad0feaa 100644
--- a/server/object.c
+++ b/server/object.c
@@ -103,6 +103,7 @@ void close_objects(void)
 /*****************************************************************/
 
 /* mark a block of memory as uninitialized for debugging purposes */
+#ifndef NDEBUG
 static inline void mark_block_uninitialized( void *ptr, size_t size )
 {
     memset( ptr, 0x55, size );
@@ -112,12 +113,18 @@ static inline void mark_block_uninitialized( void *ptr, size_t size )
     VALGRIND_DISCARD( VALGRIND_MAKE_WRITABLE( ptr, size ));
 #endif
 }
+#endif
 
 /* malloc replacement */
 void *mem_alloc( size_t size )
 {
     void *ptr = malloc( size );
-    if (ptr) mark_block_uninitialized( ptr, size );
+    if (ptr)
+#ifndef NDEBUG
+        mark_block_uninitialized( ptr, size );
+#else
+        ;
+#endif
     else set_error( STATUS_NO_MEMORY );
     return ptr;
 }
diff --git a/server/object.h b/server/object.h
index ca374bbfdec..64a28458a92 100644
--- a/server/object.h
+++ b/server/object.h
@@ -29,7 +29,9 @@
 #include "wine/server_protocol.h"
 #include "wine/list.h"
 
+#ifndef NDEBUG
 #define DEBUG_OBJECTS
+#endif
 
 /* kernel objects */
 
