Modified version of a patch made by gonX.

diff --git a/dlls/winepulse.drv/pulse.c b/dlls/winepulse.drv/pulse.c
index 11111111111..11111111111 100644
--- a/dlls/winepulse.drv/pulse.c
+++ b/dlls/winepulse.drv/pulse.c
@@ -926,11 +926,19 @@ static void pulse_probe_settings(pa_mainloop *ml, pa_context *ctx, int render, c
     if (length)
         *def_period = *min_period = pa_bytes_to_usec(10 * length, &ss);
 
-    if (*min_period < MinimumPeriod)
-        *min_period = MinimumPeriod;
+    const char* penv = getenv("STAGING_AUDIO_PERIOD");
+    int val;
+    if (!penv) {
+        if (*min_period < MinimumPeriod)
+            *min_period = MinimumPeriod;
 
-    if (*def_period < DefaultPeriod)
-        *def_period = DefaultPeriod;
+        if (*def_period < DefaultPeriod)
+            *def_period = DefaultPeriod;
+    }
+	   else if ((val = atoi(penv)) >= 0) {
+        *def_period = *min_period = val;
+        printf("Audio period set to %d.\n", val);
+    }
 
     wfx->wFormatTag = WAVE_FORMAT_EXTENSIBLE;
     wfx->cbSize = sizeof(WAVEFORMATEXTENSIBLE) - sizeof(WAVEFORMATEX);
@@ -1343,6 +1351,13 @@ static NTSTATUS pulse_create_stream(void *args)
 
     stream->def_period = period;
 
+    const char* denv = getenv("STAGING_AUDIO_DURATION");
+    if (denv) {
+        int val = atoi(denv);
+        duration = val;
+        printf("Audio duration set to %d.\n", val);
+    }
+
     stream->period_bytes = pa_frame_size(&stream->ss) * muldiv(period, stream->ss.rate, 10000000);
 
     stream->bufsize_frames = ceil((duration / 10000000.) * params->fmt->nSamplesPerSec);
