diff --git a/dlls/win32u/winstation.c b/dlls/win32u/winstation.c
index 11111111111..11111111111 100644
--- a/dlls/win32u/winstation.c
+++ b/dlls/win32u/winstation.c
@@ -102,19 +102,25 @@ BOOL SYSCALL NtUserCloseWindowStation( HWINSTA handle )
     return ret;
 }
 
+static HWINSTA winsta_cache = INVALID_HANDLE_VALUE;
+
 /***********************************************************************
  *           NtUSerGetProcessWindowStation  (win32u.@)
  */
 HWINSTA SYSCALL NtUserGetProcessWindowStation(void)
 {
+    HWINSTA cache = winsta_cache;
     HWINSTA ret = 0;
 
+    if (cache != INVALID_HANDLE_VALUE) return cache;
+
     SERVER_START_REQ( get_process_winstation )
     {
         if (!wine_server_call_err( req ))
             ret = wine_server_ptr_handle( reply->handle );
     }
     SERVER_END_REQ;
+    winsta_cache = ret;
     return ret;
 }
 
@@ -125,6 +131,8 @@ BOOL SYSCALL NtUserSetProcessWindowStation( HWINSTA handle )
 {
     BOOL ret;
 
+    winsta_cache = INVALID_HANDLE_VALUE;
+
     SERVER_START_REQ( set_process_winstation )
     {
         req->handle = wine_server_obj_handle( handle );
