From: Alex Henrie <alexhenrie24@gmail.com>
Subject: [PATCH resend 1/4] include: Add IN6_PKTINFO
Message-Id: <20210607054723.155611-1-alexhenrie24@gmail.com>
Date: Sun,  6 Jun 2021 23:47:00 -0600

Signed-off-by: Alex Henrie <alexhenrie24@gmail.com>
---
 include/ws2ipdef.h | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/include/ws2ipdef.h b/include/ws2ipdef.h
index fc465804962..4e1ec373f71 100644
--- a/include/ws2ipdef.h
+++ b/include/ws2ipdef.h
@@ -177,6 +177,11 @@ typedef struct WS(in_pktinfo) {
     UINT    ipi_ifindex;
 } IN_PKTINFO, *PIN_PKTINFO;
 
+typedef struct WS(in6_pktinfo) {
+    IN6_ADDR ipi6_addr;
+    ULONG    ipi6_ifindex;
+} IN6_PKTINFO, *PIN6_PKTINFO;
+
 #ifndef USE_WS_PREFIX
 #define IPV6_OPTIONS                    1
 #define IPV6_HDRINCL                    2

-- 
2.31.1

