From 9936c5de2c495584cb6773a8120d59e192d08be4 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Fri, 6 Nov 2020 18:44:10 +0100
Subject: [PATCH 09/11] win32u: Print full bitmap size in the font list.

---
 dlls/win32u/font.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/dlls/win32u/font.c b/dlls/win32u/font.c
index 1887b1a66f3..74c60423634 100644
--- a/dlls/win32u/font.c
+++ b/dlls/win32u/font.c
@@ -94,6 +94,12 @@ struct gdi_font_face
     struct wine_rb_entry       full_name_entry;
 };
 
+static inline const char *debugstr_bitmap_font_size(struct bitmap_font_size *size)
+{
+    return wine_dbg_sprintf("width %2d, height %2d, size %2d, xppem %4d, yppem %4d, il %2d",
+                            size->width, size->height, size->size, size->x_ppem, size->y_ppem, size->internal_leading);
+}
+
 static const struct font_backend_funcs *font_funcs;
 
 static const MAT2 identity = { {0,1}, {0,0}, {0,0}, {0,1} };
@@ -1023,7 +1029,7 @@ static void dump_gdi_font_list(void)
         {
             TRACE( "\t%s\t%s\t%08x", debugstr_w(face->style_name), debugstr_w(face->full_name),
                    face->fs.fsCsb[0] );
-            if (!face->scalable) TRACE(" %d", face->size.height );
+            if (!face->scalable) TRACE(" %s", debugstr_bitmap_font_size(&face->size) );
             TRACE("\n");
 	}
     }
-- 
2.35.1

