From: "Aurélien Inacio" <aur36@mailo.com>
Subject: [PATCH] win32u: use 7-bit ascii for unhandled encodings
Message-Id: <20211102180346.25600-1-aur36@mailo.com>
Date: Tue,  2 Nov 2021 19:03:46 +0100

Signed-off-by: Aurélien Inacio <aur36@mailo.com>
---
 dlls/win32u/opentype.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/dlls/win32u/opentype.c b/dlls/win32u/opentype.c
index 10234c6d5aa..c206d6dae57 100644
--- a/dlls/win32u/opentype.c
+++ b/dlls/win32u/opentype.c
@@ -344,8 +344,8 @@ static UINT get_name_record_codepage( enum OPENTYPE_PLATFORM_ID platform, USHORT
         case TT_NAME_MAC_ENCODING_THAI:
             return 10021;
         default:
-            WARN( "encoding %u not handled, platform %d.\n", encoding, platform );
-            break;
+            WARN( "default ascii encoding used for encoding %d, platform %d\n", encoding, platform );
+            return 20127;
         }
         break;
     case OPENTYPE_PLATFORM_WIN:
@@ -366,8 +366,8 @@ static UINT get_name_record_codepage( enum OPENTYPE_PLATFORM_ID platform, USHORT
         case TT_NAME_WIN_ENCODING_JOHAB:
             return 1361;
         default:
-            FIXME( "encoding %u not handled, platform %d.\n", encoding, platform );
-            break;
+            WARN( "default ascii encoding used for encoding %d, platform %d\n", encoding, platform );
+            return 20127;
         }
         break;
     default:

-- 
2.30.2

