From a2cde919a94954b42af9b5fd2f587e605923c2c7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Sat, 7 Nov 2020 03:48:02 +0100
Subject: [PATCH 08/11] win32u: Lazily initialize the font list.

---
 dlls/win32u/font.c | 58 +++++++++++++++++++++++++++++++++++++++++-----
 1 file changed, 52 insertions(+), 6 deletions(-)

diff --git a/dlls/win32u/font.c b/dlls/win32u/font.c
index b8213ac9e6a..1887b1a66f3 100644
--- a/dlls/win32u/font.c
+++ b/dlls/win32u/font.c
@@ -499,6 +499,10 @@ static inline BOOL is_dbcs_ansi_cp(UINT ansi_cp)
 
 static pthread_mutex_t font_lock = PTHREAD_MUTEX_INITIALIZER;
 
+pthread_once_t font_list_init_once = PTHREAD_ONCE_INIT;
+DWORD font_list_init_disposition;
+DWORD font_list_init(void);
+
 #ifndef WINE_FONT_DIR
 #define WINE_FONT_DIR "fonts"
 #endif
@@ -2866,6 +2870,7 @@ static BOOL CDECL font_CreateDC( PHYSDEV *dev, LPCWSTR device, LPCWSTR output,
 {
     struct font_physdev *physdev;
 
+    pthread_once( &font_list_init_once, font_list_init );
     if (!font_funcs) return TRUE;
     if (!(physdev = calloc( 1, sizeof(*physdev) ))) return FALSE;
     push_dc_driver( dev, &physdev->dev, &font_driver );
@@ -3155,6 +3160,8 @@ static BOOL CDECL font_EnumFonts( PHYSDEV dev, LOGFONTW *lf, FONTENUMPROCW proc,
     struct enum_charset enum_charsets[32];
     DWORD count, charset;
 
+    pthread_once( &font_list_init_once, font_list_init );
+
     charset = lf ? lf->lfCharSet : DEFAULT_CHARSET;
 
     count = create_enum_charset_list( charset, enum_charsets );
@@ -4182,6 +4189,7 @@ static HFONT CDECL font_SelectFont( PHYSDEV dev, HFONT hfont, UINT *aa_flags )
     struct gdi_font *font = NULL, *prev = physdev->font;
     DC *dc = get_physdev_dc( dev );
 
+    pthread_once( &font_list_init_once, font_list_init );
     if (hfont)
     {
         LOGFONTW lf;
@@ -5907,6 +5915,7 @@ BOOL CDECL __wine_get_file_outline_text_metric( const WCHAR *path, OUTLINETEXTME
 {
     struct gdi_font *font = NULL;
 
+    pthread_once( &font_list_init_once, font_list_init );
     if (!path || !font_funcs) return FALSE;
 
     if (!(font = alloc_gdi_font( path, NULL, 0 ))) goto done;
@@ -6368,8 +6377,41 @@ UINT font_init(void)
     if (!dpi) return 96;
     update_codepage( dpi );
 
+
+    attr.Attributes = OBJ_OPENIF;
+    attr.ObjectName = &name;
+    name.Buffer = wine_font_mutexW;
+    name.Length = name.MaximumLength = sizeof(wine_font_mutexW);
+
+    if (NtCreateMutant( &mutex, MUTEX_ALL_ACCESS, &attr, FALSE ) < 0) return dpi;
+    NtWaitForSingleObject( mutex, FALSE, NULL );
+
+    wine_fonts_cache_key = reg_create_key( wine_fonts_key, cacheW, sizeof(cacheW),
+                                           REG_OPTION_VOLATILE, &disposition );
+
+    if (disposition == REG_CREATED_NEW_KEY)
+    {
+        font_list_init_disposition = disposition;
+        pthread_once( &font_list_init_once, font_list_init );
+    }
+
+    NtReleaseMutant( mutex, NULL );
+    return dpi;
+}
+
+DWORD font_list_init(void)
+{
+    OBJECT_ATTRIBUTES attr = { sizeof(attr) };
+    UNICODE_STRING name;
+    DWORD disposition = font_list_init_disposition;
+    HANDLE mutex;
+
+    static WCHAR wine_font_mutexW[] =
+        {'\\','B','a','s','e','N','a','m','e','d','O','b','j','e','c','t','s',
+         '\\','_','_','W','I','N','E','_','F','O','N','T','_','M','U','T','E','X','_','_'};
+
     if (!(font_funcs = init_freetype_lib()))
-        return dpi;
+        return TRUE;
 
     load_system_bitmap_fonts();
     load_file_system_fonts();
@@ -6380,12 +6422,9 @@ UINT font_init(void)
     name.Buffer = wine_font_mutexW;
     name.Length = name.MaximumLength = sizeof(wine_font_mutexW);
 
-    if (NtCreateMutant( &mutex, MUTEX_ALL_ACCESS, &attr, FALSE ) < 0) return dpi;
+    if (NtCreateMutant( &mutex, MUTEX_ALL_ACCESS, &attr, FALSE ) < 0) return FALSE;
     NtWaitForSingleObject( mutex, FALSE, NULL );
 
-    wine_fonts_cache_key = reg_create_key( wine_fonts_key, cacheW, sizeof(cacheW),
-                                           REG_OPTION_VOLATILE, &disposition );
-
     if (disposition == REG_CREATED_NEW_KEY)
     {
         load_registry_fonts();
@@ -6406,7 +6445,8 @@ UINT font_init(void)
     load_system_links();
     dump_gdi_font_list();
     dump_gdi_font_subst();
-    return dpi;
+
+    return TRUE;
 }
 
 /***********************************************************************
@@ -6415,6 +6455,7 @@ UINT font_init(void)
 INT WINAPI NtGdiAddFontResourceW( const WCHAR *str, ULONG size, ULONG files, DWORD flags,
                                   DWORD tid, void *dv )
 {
+    pthread_once( &font_list_init_once, font_list_init );
     if (!font_funcs) return 1;
     return add_font_resource( str, flags );
 }
@@ -6434,6 +6475,7 @@ HANDLE WINAPI NtGdiAddFontMemResourceEx( void *ptr, DWORD size, void *dv, ULONG
         SetLastError(ERROR_INVALID_PARAMETER);
         return NULL;
     }
+    pthread_once( &font_list_init_once, font_list_init );
     if (!font_funcs) return NULL;
     if (!(copy = malloc( size ))) return NULL;
     memcpy( copy, ptr, size );
@@ -6483,6 +6525,7 @@ BOOL WINAPI NtGdiRemoveFontMemResourceEx( HANDLE handle )
 BOOL WINAPI NtGdiRemoveFontResourceW( const WCHAR *str, ULONG size, ULONG files, DWORD flags,
                                       DWORD tid, void *dv )
 {
+    pthread_once( &font_list_init_once, font_list_init );
     if (!font_funcs) return TRUE;
     return remove_font_resource( str, flags );
 }
@@ -6561,6 +6604,7 @@ BOOL WINAPI NtGdiGetRealizationInfo( HDC hdc, struct font_realization_info *info
  */
 BOOL WINAPI NtGdiGetRasterizerCaps( RASTERIZER_STATUS *status, UINT size )
 {
+    pthread_once( &font_list_init_once, font_list_init );
     status->nSize = sizeof(RASTERIZER_STATUS);
     status->wFlags = font_funcs ? (TT_AVAILABLE | TT_ENABLED) : 0;
     status->nLanguageID = 0;
@@ -6577,6 +6621,7 @@ BOOL WINAPI NtGdiGetFontFileData( DWORD instance_id, DWORD file_index, UINT64 *o
     DWORD tag = 0, size;
     BOOL ret = FALSE;
 
+    pthread_once( &font_list_init_once, font_list_init );
     if (!font_funcs) return FALSE;
     pthread_mutex_lock( &font_lock );
     if ((font = get_font_from_handle( instance_id )))
@@ -6602,6 +6647,7 @@ BOOL WINAPI NtGdiGetFontFileInfo( DWORD instance_id, DWORD file_index, struct fo
     struct gdi_font *font;
     BOOL ret = FALSE;
 
+    pthread_once( &font_list_init_once, font_list_init );
     pthread_mutex_lock( &font_lock );
 
     if ((font = get_font_from_handle( instance_id )))
-- 
2.35.1

