From ac88d93d5c75d8fa7e0367933884b485dd96eefb Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Sat, 29 Jan 2022 19:10:37 +0100
Subject: [PATCH 06/16] user32/tests: Ignore internal messages in
 GetMessagePos.

---
 dlls/user32/tests/win.c | 9 ++-------
 1 file changed, 2 insertions(+), 7 deletions(-)

diff --git a/dlls/user32/tests/win.c b/dlls/user32/tests/win.c
index 11111111111..11111111111 100644
--- a/dlls/user32/tests/win.c
+++ b/dlls/user32/tests/win.c
@@ -11169,13 +11169,8 @@ static void test_GetMessagePos(void)
 
     SetTimer(button, 1, 250, NULL);
     SetCursorPos(330, 350);
-    GetMessageA(&msg, button, 0, 0);
-    while (msg.message == WM_PAINT)
-    {
-        UpdateWindow( button );
-        GetMessageA(&msg, button, 0, 0);
-    }
-    ok(msg.message == WM_TIMER, "msg.message = %x\n", msg.message);
+    while (GetMessageA(&msg, button, 0, 0) && msg.message != WM_TIMER)
+        if (msg.message == WM_PAINT) UpdateWindow(button);
     pos = GetMessagePos();
     ok(pos == MAKELONG(330, 350), "pos = %08lx\n", pos);
     KillTimer(button, 1);
-- 
2.36.0

