From 4c28f5ed84153e25cc8e45bb3ba276e15aa28aa1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Sat, 7 May 2022 01:50:58 +0200
Subject: [PATCH 6/8] winex11.drv: Track focus globally, in desktop thread.

---
 dlls/winex11.drv/event.c | 28 ++++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/dlls/winex11.drv/event.c b/dlls/winex11.drv/event.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/event.c
+++ b/dlls/winex11.drv/event.c
@@ -601,6 +601,24 @@ static HWND get_foreground_window( Display *display, Window *window )
 }
 
 
+static BOOL has_global_focus = FALSE;
+
+static void x11drv_global_focus_out( Display *display )
+{
+    if (!has_global_focus) return;
+    TRACE("global focus lost\n");
+    has_global_focus = FALSE;
+}
+
+
+static void x11drv_global_focus_in( Display *display )
+{
+    if (has_global_focus) return;
+    TRACE("global focus gained\n");
+    has_global_focus = TRUE;
+}
+
+
 /**********************************************************************
  *              set_input_focus
  *
@@ -854,6 +872,8 @@ BOOL x11drv_handle_focus_in_event( HWND hwnd, XEvent *xev, Time time )
     TRACE( "win %p xwin %lx detail=%s mode=%s\n", hwnd, event->window, focus_details[event->detail], focus_modes[event->mode] );
 
     if (event->detail == NotifyPointer) return FALSE;
+    if (is_virtual_desktop() && event->window == root_window && event->detail != NotifyInferior)
+        x11drv_global_focus_in( event->display );
     if (hwnd == NtUserGetDesktopWindow()) return FALSE;
 
     switch (event->mode)
@@ -947,6 +967,8 @@ BOOL x11drv_handle_focus_out_event( HWND hwnd, XEvent *xev, Time time )
         return TRUE;
     }
     if (!hwnd) return FALSE;
+    if (is_virtual_desktop() && event->window == root_window && event->detail != NotifyInferior)
+        x11drv_global_focus_out( event->display );
 
     switch (event->mode)
     {
@@ -1451,9 +1473,15 @@ static BOOL X11DRV_PropertyNotify( HWND hwnd, XEvent *xev )
             return FALSE;
 
         if (!(active_window = get_foreground_window( event->display, NULL )))
+        {
             set_foreground_window( NtUserGetDesktopWindow() );
+            x11drv_global_focus_out( event->display );
+        }
         else
+        {
             set_foreground_window( active_window );
+            x11drv_global_focus_in( event->display );
+        }
 
         return FALSE;
     }
-- 
2.37.1

