From 2bf790c114b3b3c61a227d0d70fe85b786d26b41 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Fri, 13 May 2022 10:30:44 +0200
Subject: [PATCH 01/30] DEBUG: winegstreamer: Disable vaapi with an env var.

---
 dlls/winegstreamer/wg_transform.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/winegstreamer/wg_transform.c b/dlls/winegstreamer/wg_transform.c
index 11111111111..11111111111 100644
--- a/dlls/winegstreamer/wg_transform.c
+++ b/dlls/winegstreamer/wg_transform.c
@@ -293,6 +293,7 @@ static GstElement *transform_find_element(GstElementFactoryListType type, GstCap
     for (tmp = transforms; tmp != NULL && element == NULL; tmp = tmp->next)
     {
         name = gst_plugin_feature_get_name(GST_PLUGIN_FEATURE(tmp->data));
+if (getenv("NOVAAPI") && !strncmp(name, "vaapi", 5)) continue;
         if (!(element = gst_element_factory_create(GST_ELEMENT_FACTORY(tmp->data), NULL)))
             GST_WARNING("Failed to create %s element.", name);
     }
-- 
2.37.1

