From: Francois Gouget <fgouget@free.fr>
Subject: [PATCH] ws2_32: Add a trailing linefeed to a FIXME().
Message-Id: <E1mBxSx-0000nm-H3@amboise>
Date: Fri, 6 Aug 2021 12:52:39 +0200 (CEST)

Signed-off-by: Francois Gouget <fgouget@free.fr>
---
 dlls/ws2_32/unixlib.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/ws2_32/unixlib.c b/dlls/ws2_32/unixlib.c
index 240389aeeaa..f0480c5865b 100644
--- a/dlls/ws2_32/unixlib.c
+++ b/dlls/ws2_32/unixlib.c
@@ -348,7 +348,7 @@ static unsigned int errno_from_unix( int err )
         case EREMOTE:           return WSAEREMOTE;
 #endif
         default:
-            FIXME( "unknown error: %s", strerror( err ) );
+            FIXME( "unknown error: %s\n", strerror( err ) );
             return WSAEFAULT;
     }
 }

-- 
2.20.1

