From 1aae126f5ce62ab68a82da47e7da4fe763f0de49 Mon Sep 17 00:00:00 2001
From: Torge Matthies <tmatthies@codeweavers.com>
Date: Sun, 5 Mar 2023 09:47:25 +0100
Subject: [PATCH 3/4] [HACK] win32u: Don't call font_init in font_CreateDC.

---
 dlls/win32u/font.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/dlls/win32u/font.c b/dlls/win32u/font.c
index 11111111111..11111111111 100644
--- a/dlls/win32u/font.c
+++ b/dlls/win32u/font.c
@@ -3250,8 +3250,6 @@ static BOOL font_CreateDC( PHYSDEV *dev, LPCWSTR device, LPCWSTR output, const D
 {
     struct font_physdev *physdev;
 
-    pthread_once( &font_init_once, font_init );
-    if (!font_funcs) return TRUE;
     if (!(physdev = calloc( 1, sizeof(*physdev) ))) return FALSE;
     push_dc_driver( dev, &physdev->dev, &font_driver );
     return TRUE;
-- 
0.0.0

