From d511219ce13077cd853afaf24d51152e1d07f9cf Mon Sep 17 00:00:00 2001
From: Keming Liang <liangkeming@kylinos.cn>
Date: Mon, 6 Mar 2023 14:45:00 +0800
Subject: [PATCH 3/3] add a skip log when there is no simsun font in test env.

---
 dlls/gdi32/tests/font.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/gdi32/tests/font.c b/dlls/gdi32/tests/font.c
index 11111111111..11111111111 100644
--- a/dlls/gdi32/tests/font.c
+++ b/dlls/gdi32/tests/font.c
@@ -7861,6 +7861,8 @@ START_TEST(font)
             test_EnumFontFamiliesW(L"SimSun", DEFAULT_CHARSET);
             test_EnumFontFamiliesW(L"MS Shell Dlg", DEFAULT_CHARSET);
         }
+        else
+            skip("SimSun is not installed\n");
     }
 
     /* On Windows Arial has a lot of default charset aliases such as Arial Cyr,
-- 
2.40.0

