From d7fa3d2baa18786920ad5b77d420edb69d805295 Mon Sep 17 00:00:00 2001
From: Torge Matthies <tmatthies@codeweavers.com>
Date: Sun, 5 Mar 2023 07:53:12 +0100
Subject: [PATCH 1/4] win32u: Factor out initialization of font registry keys.

---
 dlls/win32u/font.c | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/dlls/win32u/font.c b/dlls/win32u/font.c
index 11111111111..11111111111 100644
--- a/dlls/win32u/font.c
+++ b/dlls/win32u/font.c
@@ -6787,6 +6787,16 @@ static HKEY open_hkcu(void)
     return reg_open_key( NULL, bufferW, len * sizeof(WCHAR) );
 }
 
+static void init_reg_keys(void)
+{
+    static const WCHAR wine_fonts_keyW[] =
+        {'S','o','f','t','w','a','r','e','\\','W','i','n','e','\\','F','o','n','t','s'};
+
+    if (!hkcu_key && !(hkcu_key = open_hkcu())) return;
+    if (!wine_fonts_key)
+        wine_fonts_key = reg_create_key( hkcu_key, wine_fonts_keyW, sizeof(wine_fonts_keyW), 0, NULL );
+}
+
 /***********************************************************************
  *              font_init
  */
@@ -6801,12 +6811,10 @@ UINT font_init(void)
     static WCHAR wine_font_mutexW[] =
         {'\\','B','a','s','e','N','a','m','e','d','O','b','j','e','c','t','s',
          '\\','_','_','W','I','N','E','_','F','O','N','T','_','M','U','T','E','X','_','_'};
-    static const WCHAR wine_fonts_keyW[] =
-        {'S','o','f','t','w','a','r','e','\\','W','i','n','e','\\','F','o','n','t','s'};
     static const WCHAR cacheW[] = {'C','a','c','h','e'};
 
-    if (!(hkcu_key = open_hkcu())) return 0;
-    wine_fonts_key = reg_create_key( hkcu_key, wine_fonts_keyW, sizeof(wine_fonts_keyW), 0, NULL );
+    init_reg_keys();
+    if (!hkcu_key) return 0;
     if (wine_fonts_key) dpi = init_font_options();
     if (!dpi) return 96;
     update_codepage( dpi );
-- 
2.40.0

