From: Serge Gautherie <winehq-git_serge_180711@gautherie.fr>
Subject: [PATCH 2/2] ntdll: RtlGetNtProductType() should not validate 'type' parameter.
Message-Id: <20200228111118.3832-1-winehq-git_serge_180711@gautherie.fr>
Date: Fri, 28 Feb 2020 12:11:18 +0100

Signed-off-by: Serge Gautherie <winehq-git_serge_180711@gautherie.fr>
---
 dlls/ntdll/version.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/ntdll/version.c b/dlls/ntdll/version.c
index 31f7b5503cc..dedb5eaad4f 100644
--- a/dlls/ntdll/version.c
+++ b/dlls/ntdll/version.c
@@ -627,7 +627,7 @@ void WINAPI RtlGetNtVersionNumbers( LPDWORD major, LPDWORD minor, LPDWORD build
  */
 BOOLEAN WINAPI RtlGetNtProductType( LPDWORD type )
 {
-    if (type) *type = current_version->wProductType;
+    *type = current_version->wProductType;
     return TRUE;
 }
 
