From: Serge Gautherie <winehq-git_serge_180711@gautherie.fr>
Subject: [PATCH] msiexec: Check registry value type again for consistency.
Message-Id: <20191206165732.2020-1-winehq-git_serge_180711@gautherie.fr>
Date: Fri,  6 Dec 2019 17:57:32 +0100

Signed-off-by: Serge Gautherie <winehq-git_serge_180711@gautherie.fr>
---
 programs/msiexec/msiexec.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/programs/msiexec/msiexec.c b/programs/msiexec/msiexec.c
index 71ecc0e590a..3aa07b6759a 100644
--- a/programs/msiexec/msiexec.c
+++ b/programs/msiexec/msiexec.c
@@ -557,7 +557,7 @@ static BOOL process_args_from_reg( const WCHAR *ident, int *pargc, WCHAR ***parg
 {
 	LONG r;
 	HKEY hkey;
-	DWORD sz = 0, type = 0;
+        DWORD sz, type;
 	WCHAR *buf;
 	BOOL ret = FALSE;
 
@@ -565,7 +565,7 @@ static BOOL process_args_from_reg( const WCHAR *ident, int *pargc, WCHAR ***parg
 			L"Software\\Microsoft\\Windows\\CurrentVersion\\Installer\\RunOnceEntries", &hkey);
 	if(r != ERROR_SUCCESS)
 		return FALSE;
-	r = RegQueryValueExW(hkey, ident, 0, &type, 0, &sz);
+        r = RegQueryValueExW(hkey, ident, NULL, &type, NULL, &sz);
 	if(r == ERROR_SUCCESS && type == REG_SZ)
 	{
 		int len = lstrlenW( *pargv[0] );
@@ -576,8 +576,8 @@ static BOOL process_args_from_reg( const WCHAR *ident, int *pargc, WCHAR ***parg
 		}
 		memcpy( buf, *pargv[0], len * sizeof(WCHAR) );
 		buf[len++] = ' ';
-		r = RegQueryValueExW(hkey, ident, 0, &type, (LPBYTE)(buf + len), &sz);
-		if( r == ERROR_SUCCESS )
+                r = RegQueryValueExW(hkey, ident, NULL, &type, (LPBYTE)(buf + len), &sz);
+                if (r == ERROR_SUCCESS && type == REG_SZ)
 		{
 			process_args(buf, pargc, pargv);
 			ret = TRUE;
