From: Zebediah Figura <z.figura12@gmail.com>
Subject: [PATCH] server: Free the old process image name if a second process image is mapped (Valgrind).
Message-Id: <20210612214253.882670-1-z.figura12@gmail.com>
Date: Sat, 12 Jun 2021 16:42:53 -0500

Signed-off-by: Zebediah Figura <z.figura12@gmail.com>
---
 server/mapping.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/server/mapping.c b/server/mapping.c
index 9cb24a4213f..cd918224caa 100644
--- a/server/mapping.c
+++ b/server/mapping.c
@@ -376,6 +376,9 @@ static void add_process_view( struct thread *thread, struct memory_view *view )
             /* main exe */
             set_process_machine( process, view );
             list_add_head( &process->views, &view->entry );
+
+            free( process->image );
+            process->image = NULL;
             if (get_view_nt_name( view, &name ) && (process->image = memdup( name.str, name.len )))
                 process->imagelen = name.len;
             return;

-- 
2.30.2

