From f82a41c353d85e8a59f55bb1fe6d914e65b552b7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Sat, 6 Mar 2021 18:12:26 +0100
Subject: [PATCH 5/6] ntdll: Initialize USR1 handler XSTATE buffer (valgrind).

---
 dlls/ntdll/unix/signal_x86_64.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/ntdll/unix/signal_x86_64.c b/dlls/ntdll/unix/signal_x86_64.c
index 11111111111..11111111111 100644
--- a/dlls/ntdll/unix/signal_x86_64.c
+++ b/dlls/ntdll/unix/signal_x86_64.c
@@ -3094,6 +3094,7 @@ static void usr1_handler( int signal, siginfo_t *siginfo, void *ucontext )
         DECLSPEC_ALIGN(64) XSTATE xs;
         context.c.ContextFlags = CONTEXT_FULL;
         context_init_xstate( &context.c, &xs );
+        memset( &xs, 0, offsetof(XSTATE, YmmContext) );
 
         NtGetContextThread( GetCurrentThread(), &context.c );
         wait_suspend( &context.c );
-- 
2.37.1

