From: Paul Gofman <pgofman@codeweavers.com>
Subject: [PATCH v3 1/3] server: Update system regs from set_thread_context handler only.
Message-Id: <20210527101600.348718-1-pgofman@codeweavers.com>
Date: Thu, 27 May 2021 13:15:58 +0300

Signed-off-by: Paul Gofman <pgofman@codeweavers.com>
---
    Supersedes 206374-206376.
v3:
    - replaced patch.

 server/thread.c | 7 +------
 1 file changed, 1 insertion(+), 6 deletions(-)

diff --git a/server/thread.c b/server/thread.c
index 0c7f11c0da1..6de36d817c1 100644
--- a/server/thread.c
+++ b/server/thread.c
@@ -1659,12 +1659,7 @@ DECL_HANDLER(select)
              current->context && current->suspend_cookie == req->cookie)
     {
         if (current->context->regs.flags)
-        {
-            unsigned int system_flags = get_context_system_regs(current->process->machine) &
-                                        current->context->regs.flags;
-            if (system_flags) set_thread_context( current, &current->context->regs, system_flags );
             set_reply_data( &current->context->regs, sizeof(context_t) );
-        }
         release_object( current->context );
         current->context = NULL;
     }
@@ -1870,7 +1865,7 @@ DECL_HANDLER(set_thread_context)
         unsigned int system_flags = get_context_system_regs( context->machine ) & context->flags;
 
         if (thread != current) stop_thread( thread );
-        else if (system_flags) set_thread_context( thread, context, system_flags );
+        if (system_flags) set_thread_context( thread, context, system_flags );
         if (thread->context && !get_error())
         {
             copy_context( &thread->context->regs, context, context->flags );

-- 
2.31.1

