From: Xu Wei <xuwei@uniontech.com>
Subject: [PATCH v3 1/2] kernel32/tests: Fix stack overflow by DefineDosDeviceW().
Message-Id: <20210909084537.2589-1-xuwei@uniontech.com>
Date: Thu,  9 Sep 2021 16:45:36 +0800

Signed-off-by: Xu Wei <xuwei@uniontech.com>
---
 dlls/kernel32/tests/volume.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/dlls/kernel32/tests/volume.c b/dlls/kernel32/tests/volume.c
index 11111111111..11111111111 100644
--- a/dlls/kernel32/tests/volume.c
+++ b/dlls/kernel32/tests/volume.c
@@ -113,6 +113,7 @@ static void test_dos_devices(void)
 {
     char buf[MAX_PATH], buf2[400];
     char drivestr[3];
+    WCHAR drivestrW[3];
     HANDLE file;
     BOOL ret;
 
@@ -187,6 +188,21 @@ static void test_dos_devices(void)
     ret = QueryDosDeviceA( drivestr, buf, sizeof(buf) );
     ok(!ret, "expected failure\n");
     ok(GetLastError() == ERROR_FILE_NOT_FOUND, "got error %lu\n", GetLastError());
+
+    SetLastError(0);
+    ret = DefineDosDeviceW( DDD_RAW_TARGET_PATH, drivestrW, NULL );
+    ok(!ret, "expected failure\n");
+    ok(GetLastError() == ERROR_INVALID_PARAMETER, "got error %lu\n", GetLastError());
+
+    SetLastError(0);
+    ret = DefineDosDeviceW( DDD_RAW_TARGET_PATH, NULL, drivestrW );
+    ok(ret, "got error %lu\n", GetLastError());
+
+    SetLastError(0);
+    DefineDosDeviceW( DDD_EXACT_MATCH_ON_REMOVE|DDD_RAW_TARGET_PATH|DDD_REMOVE_DEFINITION,
+            "{00000001-0000-0000-0000-4E6574446576}",
+            "\\Device\\{00000001-0000-0000-0000-4E6574446576}" );
+    ok(ret, "got error %lu\n", GetLastError());
 }
 
 static void test_FindFirstVolume(void)

-- 
2.39.0

