From: Eric Pouech <eric.pouech@gmail.com>
Subject: [PATCH] kernelbase: always return NULL in case of error in GetModuleHandleA (GCC11)
Message-Id: <163276099958.288139.244684879015273005.stgit@euterpe>
Date: Mon, 27 Sep 2021 18:43:21 +0200

Signed-off-by: Eric Pouech <eric.pouech@gmail.com>

---
 dlls/kernelbase/loader.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/kernelbase/loader.c b/dlls/kernelbase/loader.c
index ac463528e3c..5d70c9d0aed 100644
--- a/dlls/kernelbase/loader.c
+++ b/dlls/kernelbase/loader.c
@@ -334,7 +334,7 @@ done:
  */
 HMODULE WINAPI DECLSPEC_HOTPATCH GetModuleHandleA( LPCSTR module )
 {
-    HMODULE ret;
+    HMODULE ret = NULL;
 
     GetModuleHandleExA( GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT, module, &ret );
     return ret;

