From: Jacek Caban <jacek@codeweavers.com>
Subject: [PATCH 2/3] winebuild: Don't use iretq in syscall dispatcher.
Message-Id: <33f078c1-237a-23fb-6404-4e98065a9340@codeweavers.com>
Date: Thu, 3 Jun 2021 15:02:13 +0200


Signed-off-by: Jacek Caban <jacek@codeweavers.com>
---
  tools/winebuild/import.c | 7 +++----
  1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/tools/winebuild/import.c b/tools/winebuild/import.c
index 472eeba0491..c5451014bc6 100644
--- a/tools/winebuild/import.c
+++ b/tools/winebuild/import.c
@@ -1689,12 +1689,11 @@ static void output_syscall_dispatcher(void)
         output_cfi( ".cfi_same_value %%rsi" );
         output( "\tmovq -0x90(%%rbp),%%rbx\n" );
         output_cfi( ".cfi_same_value %%rbx" );
-        output( "\tleaq -0x28(%%rbp),%%rsp\n" );
         output_cfi( ".cfi_def_cfa_register %%rsp" );
-        output_cfi( ".cfi_adjust_cfa_offset 40" );
-        output( "\tmovq (%%rbp),%%rbp\n" );
+        output( "\tleave\n" );
+        output_cfi( ".cfi_adjust_cfa_offset -8" );
         output_cfi( ".cfi_same_value %%rbp" );
-        output( "\tiretq\n" );
+        output( "\tretq\n" );
         output( "5:\tmovl $0x%x,%%eax\n", invalid_param );
         output( "\tjmp 2b\n" );
         break;

