From: Alex Henrie <alexhenrie24@gmail.com>
Subject: [PATCH 2/9] adsldp: Fix memory leak in search_ExecuteSearch (Coverity)
Message-Id: <20211230055808.61907-1-alexhenrie24@gmail.com>
Date: Wed, 29 Dec 2021 22:58:08 -0700

Signed-off-by: Alex Henrie <alexhenrie24@gmail.com>
---
 dlls/adsldp/adsldp.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/dlls/adsldp/adsldp.c b/dlls/adsldp/adsldp.c
index 3d5ea29882c..c8f63a25e0c 100644
--- a/dlls/adsldp/adsldp.c
+++ b/dlls/adsldp/adsldp.c
@@ -1302,7 +1302,11 @@ static HRESULT WINAPI search_ExecuteSearch(IDirectorySearch *iface, LPWSTR filte
         props = NULL;
     else
     {
-        if (count && !names) return E_ADS_BAD_PARAMETER;
+        if (count && !names)
+        {
+            heap_free(ldap_ctx);
+            return E_ADS_BAD_PARAMETER;
+        }
 
         props = heap_alloc((count + 1) * sizeof(props[0]));
         if (!props)

-- 
2.34.1

