From: Alex Henrie <alexhenrie24@gmail.com>
Subject: [PATCH 3/9] msi: Fix memory leak in get_fusion_filename (Coverity)
Message-Id: <20211230060454.78139-1-alexhenrie24@gmail.com>
Date: Wed, 29 Dec 2021 23:04:54 -0700

Signed-off-by: Alex Henrie <alexhenrie24@gmail.com>
---
 dlls/msi/package.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/msi/package.c b/dlls/msi/package.c
index 7488c095080..596affbaae4 100644
--- a/dlls/msi/package.c
+++ b/dlls/msi/package.c
@@ -553,6 +553,7 @@ static LPWSTR get_fusion_filename(MSIPACKAGE *package)
         GetWindowsDirectoryW(windir, MAX_PATH);
         len = lstrlenW(windir) + lstrlenW(L"Microsoft.NET\\Framework\\") + lstrlenW(L"v2.0.50727") +
               lstrlenW(L"fusion.dll") + 3;
+        msi_free(filename);
         if (!(filename = msi_alloc(len * sizeof(WCHAR)))) return NULL;
 
         lstrcpyW(filename, windir);

-- 
2.35.1

