From: Alex Henrie <alexhenrie24@gmail.com>
Subject: [PATCH 7/9] combase: Don't use attr.ObjectName after calling create_key (cppcheck)
Message-Id: <20211230062119.95326-1-alexhenrie24@gmail.com>
Date: Wed, 29 Dec 2021 23:21:17 -0700

Signed-off-by: Alex Henrie <alexhenrie24@gmail.com>
---
 dlls/combase/combase.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/combase/combase.c b/dlls/combase/combase.c
index 11111111111..11111111111 100644
--- a/dlls/combase/combase.c
+++ b/dlls/combase/combase.c
@@ -229,7 +229,7 @@ static HKEY create_classes_root_hkey(DWORD access)
     RtlInitUnicodeString(&name, L"\\Registry\\Machine\\Software\\Classes");
 
     if (create_key( &hkey, access, &attr )) return 0;
-    TRACE( "%s -> %p\n", debugstr_w(attr.ObjectName->Buffer), hkey );
+    TRACE( "%s -> %p\n", debugstr_w(name.Buffer), hkey );
 
     if (!(access & KEY_WOW64_64KEY))
     {

-- 
2.37.1

