From: Alex Henrie <alexhenrie24@gmail.com>
Subject: [PATCH 6/7] ntdll: Remove extra call to getenv from virtual_init (scan-build)
Message-Id: <20211221050337.105216-1-alexhenrie24@gmail.com>
Date: Mon, 20 Dec 2021 22:03:37 -0700

Signed-off-by: Alex Henrie <alexhenrie24@gmail.com>
---
 dlls/ntdll/unix/virtual.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/ntdll/unix/virtual.c b/dlls/ntdll/unix/virtual.c
index 11111111111..11111111111 100644
--- a/dlls/ntdll/unix/virtual.c
+++ b/dlls/ntdll/unix/virtual.c
@@ -3308,7 +3308,7 @@ static void *alloc_virtual_heap( SIZE_T size )
 void virtual_init(void)
 {
     const struct preload_info **preload_info = dlsym( RTLD_DEFAULT, "wine_main_preload_info" );
-    const char *preload = getenv( "WINEPRELOADRESERVE" );
+    const char *preload;
     size_t size;
     int i;
     pthread_mutexattr_t attr;

-- 
0.0.0

