From: Floris Renaud <jkfloris@dds.nl>
Subject: [PATCH] winecfg: Add the command line options to the man page.
Message-Id: <20211109232812.48680-1-jkfloris@dds.nl>
Date: Wed, 10 Nov 2021 00:28:12 +0100

Signed-off-by: Floris Renaud <jkfloris@dds.nl>
---
 programs/winecfg/winecfg.man.in | 48 +++++++++++++++++++++++++++++++--
 1 file changed, 46 insertions(+), 2 deletions(-)

diff --git a/programs/winecfg/winecfg.man.in b/programs/winecfg/winecfg.man.in
index 522d2a0e812..2d2e89dfc9b 100644
--- a/programs/winecfg/winecfg.man.in
+++ b/programs/winecfg/winecfg.man.in
@@ -1,14 +1,58 @@
-.TH WINECFG 1 "November 2010" "@PACKAGE_STRING@" "Wine Programs"
+.TH WINECFG 1 "November 2021" "@PACKAGE_STRING@" "Wine Programs"
 .SH NAME
 winecfg \- Wine Configuration Editor
 .SH SYNOPSIS
-.BR "winecfg"
+.B winecfg
+.RI [ OPTION ]
 .SH DESCRIPTION
 .B winecfg
 is the Wine configuration editor. It allows you to change several settings, such as DLL load order
 (native versus builtin), enable a virtual desktop, setup disk drives, and change the Wine audio driver,
 among others. Many of these settings can be made on a per application basis, for example, preferring native
 riched20.dll for wordpad.exe, but not for notepad.exe.
+.SH OPTIONS
+.RS
+.IP "\fB[no option]\fR" 16
+Launch the graphical version of this program.
+.IP "\fB/v\fR, \fB-v\fR"
+Display the current global Windows version.
+.IP "\fB/v \fIversion\fR, \fB-v \fIversion\fR"
+Set global Windows version to
+.IR version .
+.IP "\fB/?\fR, \fB-?\fR"
+Display valid versions for
+.I version
+and available options.
+.RE
+.PP
+Valid versions for
+.I version
+are:
+.TS
+l l r
+---
+l l r.
+version	Mimic Windows version	WINEARCH
+win10	Windows 10 Ultimate N, Build 17763	win32 & win64
+win81	Windows 8.1 Ultimate N, Build 9600	win32 & win64
+win8	Windows 8 Ultimate N, Build 9200	win32 & win64
+win2008r2	Windows Server 2008 R2 Standard, Build 7601, Service Pack 1	win32 & win64
+win7	Windows 7 Ultimate N, Build 7601, Service Pack 1	win32 & win64
+win2008	Windows Server 2008 Standard, Build 6002, Service Pack 2	win32 & win64
+vista	Windows Vista Ultimate N, Build 6002, Service Pack 2	win32 & win64
+win2003	Windows Server 2003 Standard Edition, Build 3790, Service Pack 2	win32 & win64
+winxp64	Windows XP Professional, Build 3790, Service Pack 2	win64
+winxp	Windows XP Professional, Build 2600, Service Pack 3	win32
+win2k	Windows 2000 Professional, Build 2195, Service Pack 4	win32
+winme	Windows ME, V4.90, Build 3000	win32
+win98	Windows 98 Second Edition, V4.10, Build 2222	win32
+win95	Windows 95, V4.00, Build 950	win32
+nt40	Windows NT Workstation, V4.0, Build 1381, Service Pack 6a	win32
+nt351	Windows NT Workstation, V3.51, Build 1057, Service pack 5	win32
+win31	Windows 3.1	win32
+win30	Windows 3.0	win32
+win20	Windows 2.0	win32
+.TE
 .SH BUGS
 Bugs can be reported on the
 .UR https://bugs.winehq.org

-- 
2.33.1

