From bd0cf6316fb975b1e3106d225b413e5ac8d7e17d Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Sat, 20 Nov 2021 19:15:08 +0100
Subject: [PATCH 3/3] WIP: ntdll: Add a fast path for ASCII lower/upper case.

---
 dlls/ntdll/unix/unix_private.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/ntdll/unix/unix_private.h b/dlls/ntdll/unix/unix_private.h
index d05b6e38a80..6255d815139 100644
--- a/dlls/ntdll/unix/unix_private.h
+++ b/dlls/ntdll/unix/unix_private.h
@@ -481,11 +481,15 @@ static inline SIZE_T ntdll_wcscspn( const WCHAR *str, const WCHAR *reject )
 
 static inline WCHAR ntdll_towupper( WCHAR ch )
 {
+    if (ch >= 'a' && ch <= 'z') return ch - 'a' + 'A';
+    if (ch <= 0x7f) return ch;
     return ch + uctable[uctable[uctable[ch >> 8] + ((ch >> 4) & 0x0f)] + (ch & 0x0f)];
 }
 
 static inline WCHAR ntdll_towlower( WCHAR ch )
 {
+    if (ch >= 'A' && ch <= 'Z') return ch - 'A' + 'a';
+    if (ch <= 0x7f) return ch;
     return ch + lctable[lctable[lctable[ch >> 8] + ((ch >> 4) & 0x0f)] + (ch & 0x0f)];
 }
 
-- 
2.34.1

