From: Alex Henrie <alexhenrie24@gmail.com>
Subject: [PATCH 4/7] ntdll: Fix memory leak on error path in NtCreateFile (scan-build)
Message-Id: <20211221050248.101468-1-alexhenrie24@gmail.com>
Date: Mon, 20 Dec 2021 22:02:48 -0700

Signed-off-by: Alex Henrie <alexhenrie24@gmail.com>
---
 dlls/ntdll/unix/file.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/ntdll/unix/file.c b/dlls/ntdll/unix/file.c
index 11111111111..11111111111 100644
--- a/dlls/ntdll/unix/file.c
+++ b/dlls/ntdll/unix/file.c
@@ -4194,7 +4194,7 @@ NTSTATUS WINAPI NtCreateFile( HANDLE *handle, ACCESS_MASK access, OBJECT_ATTRIBU
 {
     OBJECT_ATTRIBUTES new_attr;
     UNICODE_STRING nt_name;
-    char *unix_name;
+    char *unix_name = 0;
     BOOL created = FALSE;
     NTSTATUS status;
     ULONG new_attributes, new_disposition;
@@ -4290,6 +4290,7 @@ NTSTATUS WINAPI NtCreateFile( HANDLE *handle, ACCESS_MASK access, OBJECT_ATTRIBU
     if (status != STATUS_SUCCESS)
     {
         WARN( "%s not found (%x)\n", debugstr_us(attr->ObjectName), io->u.Status );
+        free( unix_name );
         return status;
     }
 

-- 
2.37.2

