From 4054919d99b37edb6691484703562e65cb99f978 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Sat, 20 Nov 2021 19:14:04 +0100
Subject: [PATCH 2/4] WIP: ntdll: Lookup short file name extension from the
 end.

---
 dlls/ntdll/unix/file.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/dlls/ntdll/unix/file.c b/dlls/ntdll/unix/file.c
index 73dd4eba504..7702fe26aee 100644
--- a/dlls/ntdll/unix/file.c
+++ b/dlls/ntdll/unix/file.c
@@ -1290,7 +1290,9 @@ static ULONG hash_short_file_name( const WCHAR *name, int length, LPWSTR buffer
     }
 
     /* Find last dot for start of the extension */
-    for (p = name + 1, ext = NULL; p < end - 1; p++) if (*p == '.') ext = p;
+    for (end = name + length, ext = end - 1; ext > name; ext--)
+        if (*ext == '.') break;
+    if (ext == name) ext = NULL;
 
     /* Copy first 4 chars, replacing invalid chars with '_' */
     for (i = 4, p = name, dst = buffer; i > 0; i--, p++)
-- 
2.34.1

