From: Jinoh Kang <jinoh.kang.kr@gmail.com>
Subject: [PATCH v6 1/6] include: Move WIN32_MEMORY_RANGE_ENTRY from winnt.h to winbase.h.
Message-Id: <a6221984-7278-3a24-4908-b77a4149bd6b@gmail.com>
Date: Fri, 3 Dec 2021 23:35:45 +0900

Signed-off-by: Jinoh Kang <jinoh.kang.kr@gmail.com>
---
 include/winbase.h | 6 ++++++
 include/winnt.h   | 6 ------
 2 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/include/winbase.h b/include/winbase.h
index 8e8a185b4ff..a1efec3db4b 100644
--- a/include/winbase.h
+++ b/include/winbase.h
@@ -1757,6 +1757,12 @@ typedef struct _WIN32_FIND_STREAM_DATA {
     WCHAR cStreamName[MAX_PATH + 36];
 } WIN32_FIND_STREAM_DATA,*PWIN32_FIND_STREAM_DATA;
 
+typedef struct _WIN32_MEMORY_RANGE_ENTRY
+{
+    PVOID  VirtualAddress;
+    SIZE_T NumberOfBytes;
+} WIN32_MEMORY_RANGE_ENTRY, *PWIN32_MEMORY_RANGE_ENTRY;
+
 WINBASEAPI BOOL        WINAPI ActivateActCtx(HANDLE,ULONG_PTR *);
 WINADVAPI  BOOL        WINAPI AddAccessAllowedAce(PACL,DWORD,DWORD,PSID);
 WINADVAPI  BOOL        WINAPI AddAccessAllowedAceEx(PACL,DWORD,DWORD,DWORD,PSID);
diff --git a/include/winnt.h b/include/winnt.h
index 2ab74ab123c..3e8a92fc063 100644
--- a/include/winnt.h
+++ b/include/winnt.h
@@ -776,12 +776,6 @@ typedef struct DECLSPEC_ALIGN(8) MEM_EXTENDED_PARAMETER {
     } DUMMYUNIONNAME;
 } MEM_EXTENDED_PARAMETER, *PMEM_EXTENDED_PARAMETER;
 
-typedef struct _WIN32_MEMORY_RANGE_ENTRY
-{
-    PVOID  VirtualAddress;
-    SIZE_T NumberOfBytes;
-} WIN32_MEMORY_RANGE_ENTRY, *PWIN32_MEMORY_RANGE_ENTRY;
-
 #define	PAGE_NOACCESS		0x01
 #define	PAGE_READONLY		0x02
 #define	PAGE_READWRITE		0x04

-- 
2.31.1

