From: Alex Henrie <alexhenrie24@gmail.com>
Subject: [PATCH 5/9] mciwave: Fix bad free in MCI_SendCommandAsync (Coverity)
Message-Id: <20211230060748.92445-1-alexhenrie24@gmail.com>
Date: Wed, 29 Dec 2021 23:07:48 -0700

Signed-off-by: Alex Henrie <alexhenrie24@gmail.com>
---
 dlls/mciwave/mciwave.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/mciwave/mciwave.c b/dlls/mciwave/mciwave.c
index 13dd927cd42..4f4e69cd0a2 100644
--- a/dlls/mciwave/mciwave.c
+++ b/dlls/mciwave/mciwave.c
@@ -124,7 +124,7 @@ static	DWORD MCI_SendCommandAsync(UINT wDevID, async_cmd cmd, DWORD_PTR dwParam1
 	WARN("Couldn't allocate thread for async command handling, sending synchronously\n");
         if (handles[1]) CloseHandle(handles[1]);
         sca->evt = NULL;
-	return MCI_SCAStarter(&sca);
+        return MCI_SCAStarter(sca);
     }
 
     SetThreadPriority(handles[0], THREAD_PRIORITY_TIME_CRITICAL);

-- 
2.34.1

