From: Alex Henrie <alexhenrie24@gmail.com>
Subject: [PATCH 6/9] shell32: Fix sizeof argument in DoPaste (Coverity)
Message-Id: <20211230060920.92501-1-alexhenrie24@gmail.com>
Date: Wed, 29 Dec 2021 23:09:20 -0700

Signed-off-by: Alex Henrie <alexhenrie24@gmail.com>
---
 dlls/shell32/shlview_cmenu.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/shell32/shlview_cmenu.c b/dlls/shell32/shlview_cmenu.c
index 292d6c94886..f9bcf1f0dd9 100644
--- a/dlls/shell32/shlview_cmenu.c
+++ b/dlls/shell32/shlview_cmenu.c
@@ -1702,7 +1702,7 @@ static HRESULT DoPaste(ContextMenu *This)
 
 	      TRACE("CF_HDROP=%p\n", medium.u.hGlobal);
 	      count = DragQueryFileW(medium.u.hGlobal, -1, NULL, 0);
-	      pidls = SHAlloc(count*sizeof(ITEMIDLIST**));
+	      pidls = SHAlloc(count*sizeof(ITEMIDLIST*));
 	      if (pidls)
 	      {
 	        for (i = 0; i < count; i++)

-- 
2.35.1

