From: Alex Henrie <alexhenrie24@gmail.com>
Subject: [PATCH] wined3d: Fix memory leak on error path in wined3d_buffer_vk_create_buffer_object (Coverity)
Message-Id: <20211229040203.186207-1-alexhenrie24@gmail.com>
Date: Tue, 28 Dec 2021 21:02:03 -0700

Signed-off-by: Alex Henrie <alexhenrie24@gmail.com>
---
 dlls/wined3d/buffer.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/wined3d/buffer.c b/dlls/wined3d/buffer.c
index c078ba7e634..b42c7bd919a 100644
--- a/dlls/wined3d/buffer.c
+++ b/dlls/wined3d/buffer.c
@@ -1498,6 +1498,7 @@ static BOOL wined3d_buffer_vk_create_buffer_object(struct wined3d_buffer_vk *buf
             vk_memory_type_from_access_flags(resource->access, resource->usage), bo_vk)))
     {
         WARN("Failed to create Vulkan buffer.\n");
+        heap_free(bo_vk);
         return FALSE;
     }
 

-- 
2.34.1

