From 450200319c3abe0d35bffaafd3b2a46f216b7c6f Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Thu, 31 Mar 2022 22:11:46 +0200
Subject: [PATCH 1/2] ntdll, esync: Use Precise system time for calculations.

Signed-off-by: Torge Matthies <openglfreak@googlemail.com>
---
 dlls/ntdll/unix/esync.c | 8 ++------
 1 file changed, 2 insertions(+), 6 deletions(-)

diff --git a/dlls/ntdll/unix/esync.c b/dlls/ntdll/unix/esync.c
index 11111111111..11111111111 100644
--- a/dlls/ntdll/unix/esync.c
+++ b/dlls/ntdll/unix/esync.c
@@ -728,8 +728,7 @@ static LONGLONG update_timeout( ULONGLONG end )
     LARGE_INTEGER now;
     LONGLONG timeleft;
 
-    NtQuerySystemTime( &now );
-    timeleft = end - now.QuadPart;
+    timeleft = end - RtlGetSystemTimePrecise();
     if (timeleft < 0) timeleft = 0;
     return timeleft;
 }
@@ -813,7 +812,6 @@ static NTSTATUS __esync_wait_objects( DWORD count, const HANDLE *handles, BOOLEA
     int has_esync = 0, has_server = 0;
     BOOL msgwait = FALSE;
     LONGLONG timeleft;
-    LARGE_INTEGER now;
     DWORD pollcount;
     ULONGLONG end;
     int64_t value;
@@ -842,7 +840,6 @@ static NTSTATUS __esync_wait_objects( DWORD count, const HANDLE *handles, BOOLEA
         ntdll_get_thread_data()->esync_apc_fd = fd;
     }
 
-    NtQuerySystemTime( &now );
     if (timeout)
     {
         if (timeout->QuadPart == TIMEOUT_INFINITE)
@@ -850,7 +847,7 @@ static NTSTATUS __esync_wait_objects( DWORD count, const HANDLE *handles, BOOLEA
         else if (timeout->QuadPart >= 0)
             end = timeout->QuadPart;
         else
-            end = now.QuadPart - timeout->QuadPart;
+            end = RtlGetSystemTimePrecise() - timeout->QuadPart;
     }
 
     for (i = 0; i < count; i++)
@@ -1058,7 +1055,6 @@ static NTSTATUS __esync_wait_objects( DWORD count, const HANDLE *handles, BOOLEA
 
                 /* If we got here, someone else stole (or reset, etc.) whatever
                  * we were waiting for. So keep waiting. */
-                NtQuerySystemTime( &now );
             }
             else
                 goto err;
-- 
2.37.2

