From: Mohamad Al-Jaf <mohamadaljaf@gmail.com>
Subject: [PATCH v11 1/8] wdscore: Fix spec entries.
Message-Id: <20220412030023.405531-1-mohamadaljaf@gmail.com>
Date: Mon, 11 Apr 2022 23:00:16 -0400

Signed-off-by: Mohamad Al-Jaf <mohamadaljaf@gmail.com>
---
 dlls/wdscore/wdscore.spec | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/dlls/wdscore/wdscore.spec b/dlls/wdscore/wdscore.spec
index 11111111111..11111111111 100644
--- a/dlls/wdscore/wdscore.spec
+++ b/dlls/wdscore/wdscore.spec
@@ -116,7 +116,7 @@
 @ stub WdsGetBlackboardValue
 @ stub WdsGetCurrentExecutionGroup
 @ stub WdsGetSetupLog
-@ stub WdsTempDir
+@ stub WdsGetTempDir
 @ stub WdsInitialize
 @ stub WdsInitializeCallbackArray
 @ stub WdsInitializeDataBinary
@@ -139,7 +139,7 @@
 @ stub WdsPackCollection
 @ stub WdsPublish
 @ stub WdsPublishEx
-@ stub WdsPublishImmediateAsynch
+@ stub WdsPublishImmediateAsync
 @ stub WdsPublishImmediateEx
 @ stub WdsPublishOffline
 @ stub WdsSeqAlloc
@@ -150,8 +150,8 @@
 @ stub WdsSetUILanguage
 @ stub WdsSetupLogDestroy
 @ stub WdsSetupLogInit
-@ stub WdsSetLogMessageA
-@ stub WdsSetLogMessageW
+@ stub WdsSetupLogMessageA
+@ stub WdsSetupLogMessageW
 @ stub WdsSubscribeEx
 @ stub WdsTerminate
 @ stub WdsUnlockExecutionGroup

-- 
2.39.0

