From 0f5cdcd3d0500a58f017ecbf584f3ef2e7c02df4 Mon Sep 17 00:00:00 2001
From: Torge Matthies <tmatthies@codeweavers.com>
Date: Thu, 21 Sep 2023 02:51:16 +0200
Subject: [PATCH 3/5] win32u: Capitalize font directory name.

---
 dlls/win32u/font.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/dlls/win32u/font.c b/dlls/win32u/font.c
index 11111111111..11111111111 100644
--- a/dlls/win32u/font.c
+++ b/dlls/win32u/font.c
@@ -505,7 +505,7 @@ static void get_fonts_data_dir_path( const WCHAR *file, WCHAR *path )
 
 static void get_fonts_win_dir_path( const WCHAR *file, WCHAR *path )
 {
-    asciiz_to_unicode( path, "\\??\\C:\\windows\\fonts\\" );
+    asciiz_to_unicode( path, "\\??\\C:\\windows\\Fonts\\" );
     if (file) lstrcatW( path, file );
 }
 
@@ -5920,8 +5920,8 @@ static inline int get_line_width( DC *dc, int metric_size )
  *           NtGdiExtTextOutW    (win32u.@)
  *
  * Draws text using the currently selected font, background color, and text color.
- * 
- * 
+ *
+ *
  * PARAMS
  *    x,y    [I] coordinates of string
  *    flags  [I]
-- 
0.0.0

