From aeaa888230d529f5d22c77bed9921c5d7c2fd7c7 Mon Sep 17 00:00:00 2001
From: Claire Girka <claire@sitedethib.com>
Date: Fri, 15 Jul 2022 22:09:57 +0200
Subject: [PATCH 2/4] mmdevapi: decode ContainerId property to CLSID in
 MMDevice_GetPropValue.

---
 dlls/mmdevapi/devenum.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/dlls/mmdevapi/devenum.c b/dlls/mmdevapi/devenum.c
index 11111111111..11111111111 100644
--- a/dlls/mmdevapi/devenum.c
+++ b/dlls/mmdevapi/devenum.c
@@ -189,6 +189,21 @@ static HRESULT MMDevice_GetPropValue(const GUID *devguid, DWORD flow, REFPROPERT
             break;
     }
     RegCloseKey(regkey);
+
+    /* Special case ContainerID as CLSID */
+    if(pv->vt == VT_BLOB && pv->blob.pBlobData && pv->blob.cbSize == 24 && pv->blob.pBlobData[0] == VT_CLSID && IsEqualPropertyKey(*key, DEVPKEY_Device_ContainerId)) {
+        GUID *guid = CoTaskMemAlloc(sizeof(GUID));
+        if (!guid) {
+            PropVariantClear(pv);
+            hr = E_OUTOFMEMORY;
+        } else {
+            memcpy(guid, pv->blob.pBlobData + 8, sizeof(GUID));
+            CoTaskMemFree(pv->blob.pBlobData);
+            pv->vt = VT_CLSID;
+            pv->puuid = guid;
+        }
+    }
+
     return hr;
 }
 


