From 337a59ff8d922de40942ad553f50dcd495e06b57 Mon Sep 17 00:00:00 2001
From: Torge Matthies <tmatthies@codeweavers.com>
Date: Fri, 2 Jun 2023 11:53:21 +0200
Subject: [PATCH 6/7] makedep: Support architecture-specific EXTRAASMFLAGS.

Signed-off-by: Torge Matthies <tmatthies@codeweavers.com>
---
 tools/makedep.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/tools/makedep.c b/tools/makedep.c
index 11111111111..11111111111 100644
--- a/tools/makedep.c
+++ b/tools/makedep.c
@@ -2727,6 +2727,7 @@ static void output_source_asm_one_arch( struct makefile *make, struct incl_file
                                         unsigned int arch, int is_dll_src )
 {
     const char *obj_name;
+    char buffer[64];
 
     if (make->disabled[arch] && !(source->file->flags & FLAG_C_IMPLIB)) return;
 
@@ -2764,6 +2765,11 @@ static void output_source_asm_one_arch( struct makefile *make, struct incl_file
     output_filenames( cpp_flags );
     output_filenames( get_expanded_make_var_array( make, "EXTRAASMFLAGS" ));
     output_filenames( get_expanded_file_local_var( make, obj, "EXTRAASMFLAGS" ));
+    strcpy( buffer, archs.str[arch] );
+    strcat( buffer, "_" );
+    strcat( buffer, "EXTRAASMFLAGS" );
+    output_filenames( get_expanded_make_var_array( make, buffer ));
+    output_filenames( get_expanded_file_local_var( make, obj, buffer ));
     output( " %s\n", source->filename );
 }
 


