From 0fb9bf1ccd55a2db36c7ddbd60e7204241728b44 Mon Sep 17 00:00:00 2001
From: Paul Gofman <pgofman@codeweavers.com>
Date: Tue, 3 May 2022 16:28:09 -0500
Subject: [PATCH 1/4] ntdll: Fix allocation size in grow_logical_proc_ex_buf().

Signed-off-by: Paul Gofman <pgofman@codeweavers.com>
---
 dlls/ntdll/unix/system.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/ntdll/unix/system.c b/dlls/ntdll/unix/system.c
index 11111111111..11111111111 100644
--- a/dlls/ntdll/unix/system.c
+++ b/dlls/ntdll/unix/system.c
@@ -621,8 +621,8 @@ static BOOL grow_logical_proc_ex_buf( SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX **
 {
     SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX *new_dataex;
     DWORD new_len = *max_len * 2;
-    if (!(new_dataex = realloc( *pdataex, new_len * sizeof(*new_dataex) ))) return FALSE;
-    memset( new_dataex + *max_len, 0, (new_len - *max_len) * sizeof(*new_dataex) );
+    if (!(new_dataex = realloc( *pdataex, new_len ))) return FALSE;
+    memset( (char *)new_dataex + *max_len, 0, new_len - *max_len );
     *pdataex = new_dataex;
     *max_len = new_len;
     return TRUE;


