From 68604557a54ba43f31a63022e80f70095ada4696 Mon Sep 17 00:00:00 2001
From: Torge Matthies <tmatthies@codeweavers.com>
Date: Tue, 4 Apr 2023 01:15:52 +0200
Subject: [PATCH] ntdll: Exclude more environment variables from Windows
 environment.

Signed-off-by: Torge Matthies <tmatthies@codeweavers.com>
---
 dlls/ntdll/unix/env.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/dlls/ntdll/unix/env.c b/dlls/ntdll/unix/env.c
index 11111111111..11111111111 100644
--- a/dlls/ntdll/unix/env.c
+++ b/dlls/ntdll/unix/env.c
@@ -342,6 +342,9 @@ static BOOL is_special_env_var( const char *var )
             STARTS_WITH( var, "HOME=" ) ||
             STARTS_WITH( var, "TEMP=" ) ||
             STARTS_WITH( var, "TMP=" ) ||
+            STARTS_WITH( var, "TZ=" ) ||
+            STARTS_WITH( var, "LANG=" ) ||
+            STARTS_WITH( var, "LC_" ) ||
             STARTS_WITH( var, "QT_" ) ||
             STARTS_WITH( var, "VK_" ));
 }
-- 
0.0.0

