From 38a045f4684a916d32ff0362f07ddd15a939bbfc Mon Sep 17 00:00:00 2001
From: Torge Matthies <tmatthies@codeweavers.com>
Date: Tue, 2 May 2023 01:36:12 +0200
Subject: [PATCH] libs/libjpeg: Set default DCT algorithm to fastest.

---
 libs/jpeg/jconfig.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/libs/jpeg/jconfig.h b/libs/jpeg/jconfig.h
index 11111111111..11111111111 100644
--- a/libs/jpeg/jconfig.h
+++ b/libs/jpeg/jconfig.h
@@ -17,6 +17,8 @@
 /* #undef NEED_SHORT_EXTERNAL_NAMES */
 /* Define this if you get warnings about undefined structures. */
 /* #undef INCOMPLETE_TYPES_BROKEN */
+#define JDCT_DEFAULT JDCT_FASTEST
+#define JDCT_FASTEST JDCT_IFAST
 
 /* Define "boolean" as unsigned char, not enum, on Windows systems. */
 #ifdef _WIN32
-- 
0.0.0

