From f956b78206dd0180db5adb17fbee28544c10aa26 Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Fri, 16 Dec 2022 19:37:35 +0100
Subject: [PATCH] ntdll: Increase LFH block size category count.

---
 dlls/ntdll/heap.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/ntdll/heap.c b/dlls/ntdll/heap.c
index 11111111111..11111111111 100644
--- a/dlls/ntdll/heap.c
+++ b/dlls/ntdll/heap.c
@@ -157,7 +157,7 @@ C_ASSERT( sizeof(struct entry) <= HEAP_MIN_BLOCK_SIZE );
 #define BLOCK_SIZE_CATEGORY_SMALL(size)   (((size) - HEAP_MIN_BLOCK_SIZE) / BLOCK_ALIGN)
 #define BLOCK_SIZE_CATEGORY_MEDIUM(size)  (BLOCK_SIZE_CATEGORY_SMALL(BLOCK_SIZE_SMALL_MAX) + ((size) - (BLOCK_SIZE_SMALL_MAX)) / BLOCK_SIZE_MEDIUM_STEP)
 #define BLOCK_SIZE_CATEGORY(size)         (((size) >= BLOCK_SIZE_SMALL_MAX) ? BLOCK_SIZE_CATEGORY_MEDIUM(size) : BLOCK_SIZE_CATEGORY_SMALL(size))
-#define BLOCK_SIZE_CATEGORY_COUNT         BLOCK_SIZE_CATEGORY(0x400 * BLOCK_ALIGN)
+#define BLOCK_SIZE_CATEGORY_COUNT         BLOCK_SIZE_CATEGORY(0x1500 * BLOCK_ALIGN)
 #define BLOCK_SIZE_MEDIUM_MIN             (BLOCK_SIZE_SMALL_MAX + BLOCK_SIZE_MEDIUM_STEP - BLOCK_ALIGN)
 
 #define BLOCK_CATEGORY_SIZE(category)     (category >= BLOCK_SIZE_CATEGORY_MEDIUM(BLOCK_SIZE_SMALL_MAX) \
@@ -167,7 +167,7 @@ C_ASSERT( sizeof(struct entry) <= HEAP_MIN_BLOCK_SIZE );
 C_ASSERT( BLOCK_SIZE_CATEGORY(HEAP_MIN_BLOCK_SIZE) == 0 );
 C_ASSERT( BLOCK_SIZE_CATEGORY_SMALL(BLOCK_SIZE_SMALL_MAX - 1) == 44 );
 C_ASSERT( BLOCK_SIZE_CATEGORY_MEDIUM(BLOCK_SIZE_SMALL_MAX) == 45 );
-C_ASSERT( BLOCK_SIZE_CATEGORY_COUNT == 126 );
+C_ASSERT( BLOCK_SIZE_CATEGORY_COUNT == 489 );
 
 /* difference between block classes must fit into block tail_size */
 C_ASSERT( BLOCK_CATEGORY_SIZE(126) - BLOCK_CATEGORY_SIZE(125) <= FIELD_MAX( struct block, tail_size ) );
-- 
2.39.0

