From e59d96f59742a9074ae44576aa350764f50675dd Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Fri, 20 Jan 2023 18:31:18 +0100
Subject: [PATCH 7/8] STAGING: server: Propagate main thread scheduling priority to
 shm request threads.

---
 server/thread.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/server/thread.c b/server/thread.c
index 11111111111..11111111111 100644
--- a/server/thread.c
+++ b/server/thread.c
@@ -601,6 +601,8 @@ struct thread *create_thread( int fd, struct process *process, const struct secu
     int request_pipe[2];
 #ifdef __linux__
     pthread_t pthread;
+    int policy;
+    struct sched_param param;
 #endif
 
     if (memory_barrier_obj)
@@ -702,6 +704,9 @@ struct thread *create_thread( int fd, struct process *process, const struct secu
         release_object( thread );
         return NULL;
     }
+    if (!pthread_getschedparam( pthread_self(), &policy, &param ) && (policy || param.sched_priority))
+        if (pthread_setschedparam( pthread, policy | SCHED_RESET_ON_FORK, &param ))
+            pthread_setschedparam( pthread, policy, &param );
     pthread_detach( pthread );
     thread->request_shm_thread_running = 1;
 #endif
-- 
0.0.0

