From 08fba3115ad459fe0566c044f802672cfd2a4051 Mon Sep 17 00:00:00 2001
From: Torge Matthies <tmatthies@codeweavers.com>
Date: Thu, 8 Dec 2022 20:33:30 +0100
Subject: [PATCH 1/5] include: Add SYSCALL calling convention.

---
 include/windef.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/windef.h b/include/windef.h
index 11111111111..11111111111 100644
--- a/include/windef.h
+++ b/include/windef.h
@@ -200,6 +200,8 @@ extern "C" {
 #define APIENTRY    WINAPI
 #define CONST       __ONLY_IN_WINELIB(const)
 
+#define SYSCALL WINAPI
+
 /* Misc. constants. */
 
 #ifndef NULL
-- 
2.39.0

