From f6a61b2aeb68cdb0d7a86fc877e4abf1b5a6ffc4 Mon Sep 17 00:00:00 2001
From: Giovanni Mascellani <gmascellani@codeweavers.com>
Date: Thu, 28 Apr 2022 13:26:58 +0200
Subject: [PATCH] Silence some warnings.

---
 dlls/sharedgpures.sys/shared_resource.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/sharedgpures.sys/shared_resource.c b/dlls/sharedgpures.sys/shared_resource.c
index 11111111111..11111111111 100644
--- a/dlls/sharedgpures.sys/shared_resource.c
+++ b/dlls/sharedgpures.sys/shared_resource.c
@@ -282,7 +282,7 @@ static NTSTATUS WINAPI dispatch_ioctl(DEVICE_OBJECT *device, IRP *irp)
     struct shared_resource **res = (struct shared_resource **) &stack->FileObject->FsContext;
     NTSTATUS status;
 
-    TRACE( "ioctl %x insize %u outsize %u\n",
+    TRACE( "ioctl %lx insize %lu outsize %lu\n",
            stack->Parameters.DeviceIoControl.IoControlCode,
            stack->Parameters.DeviceIoControl.InputBufferLength,
            stack->Parameters.DeviceIoControl.OutputBufferLength );
@@ -314,7 +314,7 @@ static NTSTATUS WINAPI dispatch_ioctl(DEVICE_OBJECT *device, IRP *irp)
                                       &irp->IoStatus );
             break;
     default:
-        FIXME( "ioctl %x not supported\n", stack->Parameters.DeviceIoControl.IoControlCode );
+        FIXME( "ioctl %lx not supported\n", stack->Parameters.DeviceIoControl.IoControlCode );
         status = STATUS_NOT_SUPPORTED;
         break;
     }
-- 
2.37.1

