From 7b6953f2d8e858858ebd7c744d02f53acdda4d1f Mon Sep 17 00:00:00 2001
From: Torge Matthies <tmatthies@codeweavers.com>
Date: Fri, 2 Jun 2023 11:53:21 +0200
Subject: [PATCH 4/7] configure: Search for nasm-compatible assembler.

Signed-off-by: Torge Matthies <tmatthies@codeweavers.com>
---
 configure    | 55 ++++++++++++++++++++++++++++++++++++++++++++++++++++
 configure.ac |  6 ++++++
 2 files changed, 61 insertions(+)

diff --git a/configure b/configure
index 11111111111..11111111111 100755
--- a/configure
+++ b/configure
@@ -795,6 +795,7 @@ RANLIB
 STRIP
 ac_ct_AR
 AR
+NASM
 BISON
 FLEX
 SED_CMD
@@ -6623,6 +6624,59 @@ fi
 printf "%s\n" "$wine_cv_recent_bison" >&6; }
 test $wine_cv_recent_bison != no || as_fn_error $? "Your bison version is too old. Please install bison version 3.0 or newer." "$LINENO" 5
 
+for ac_prog in nasm yasm
+do
+  # Extract the first word of "$ac_prog", so it can be a program name with args.
+set dummy $ac_prog; ac_word=$2
+{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for $ac_word" >&5
+printf %s "checking for $ac_word... " >&6; }
+if test ${ac_cv_prog_NASM+y}
+then :
+  printf %s "(cached) " >&6
+else $as_nop
+  if test -n "$NASM"; then
+  ac_cv_prog_NASM="$NASM" # Let the user override the test.
+else
+as_save_IFS=$IFS; IFS=$PATH_SEPARATOR
+for as_dir in $PATH
+do
+  IFS=$as_save_IFS
+  case $as_dir in #(((
+    '') as_dir=./ ;;
+    */) ;;
+    *) as_dir=$as_dir/ ;;
+  esac
+    for ac_exec_ext in '' $ac_executable_extensions; do
+  if as_fn_executable_p "$as_dir$ac_word$ac_exec_ext"; then
+    ac_cv_prog_NASM="$ac_prog"
+    printf "%s\n" "$as_me:${as_lineno-$LINENO}: found $as_dir$ac_word$ac_exec_ext" >&5
+    break 2
+  fi
+done
+  done
+IFS=$as_save_IFS
+
+fi
+fi
+NASM=$ac_cv_prog_NASM
+if test -n "$NASM"; then
+  { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: $NASM" >&5
+printf "%s\n" "$NASM" >&6; }
+else
+  { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: no" >&5
+printf "%s\n" "no" >&6; }
+fi
+
+
+  test -n "$NASM" && break
+done
+test -n "$NASM" || NASM="none"
+
+if test "$NASM" = "none"
+then
+  as_fn_error $? "no suitable nasm found. Please install the 'nasm' or 'yasm' package." "$LINENO" 5
+fi
+
 if test -n "$ac_tool_prefix"; then
   for ac_prog in ar gar
   do
@@ -23253,6 +23307,7 @@ RUNTESTFLAGS = $RUNTESTFLAGS
 SED_CMD = $SED_CMD
 FLEX = $FLEX
 BISON = $BISON
+NASM = $NASM
 AR = $AR
 ac_ct_AR = $ac_ct_AR
 STRIP = $STRIP
diff --git a/configure.ac b/configure.ac
index 11111111111..11111111111 100644
--- a/configure.ac
+++ b/configure.ac
@@ -290,6 +290,12 @@ EOF
     fi])
 test $wine_cv_recent_bison != no || AC_MSG_ERROR([Your bison version is too old. Please install bison version 3.0 or newer.])
 
+AC_CHECK_PROGS(NASM,[nasm yasm],none)
+if test "$NASM" = "none"
+then
+  AC_MSG_ERROR([no suitable nasm found. Please install the 'nasm' or 'yasm' package.])
+fi
+
 AC_CHECK_TOOLS(AR,[ar gar],ar)
 AC_CHECK_TOOL(STRIP,strip,strip)
 AC_PROG_RANLIB


