From 9b8468a90b53c092a277a6c07f598da14789a307 Mon Sep 17 00:00:00 2001
From: Torge Matthies <tmatthies@codeweavers.com>
Date: Thu, 21 Sep 2023 02:51:26 +0200
Subject: [PATCH 4/5] setupapi: Capitalize font directory name.

---
 dlls/setupapi/dirid.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/setupapi/dirid.c b/dlls/setupapi/dirid.c
index 11111111111..11111111111 100644
--- a/dlls/setupapi/dirid.c
+++ b/dlls/setupapi/dirid.c
@@ -111,7 +111,7 @@ static const WCHAR *create_system_dirid( int dirid )
         break;
     case DIRID_FONTS:
         GetWindowsDirectoryW( buffer, MAX_PATH );
-        lstrcatW( buffer, L"\\fonts" );
+        lstrcatW( buffer, L"\\Fonts" );
         break;
     case DIRID_VIEWERS:
         GetSystemDirectoryW( buffer, MAX_PATH );
-- 
0.0.0

