From 2fe9c84be64a02dc42d0cb81909dc8f8492891c2 Mon Sep 17 00:00:00 2001
From: Paul Gofman <pgofman@codeweavers.com>
Date: Tue, 7 Feb 2023 10:16:36 -0600
Subject: [PATCH 3/4] loader/wine.inf: Add known dlls key.

---
 loader/wine.inf.in | 45 +++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 45 insertions(+)

diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index 11111111111..11111111111 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -503,6 +503,51 @@ HKLM,%Control%\Session Manager\Environment,"windir",0x00020000,"%SystemRoot%"
 HKLM,%Control%\Session Manager\Environment,"winsysdir",,"%11%"
 HKLM,%Control%\Session Manager\Memory Management,PagingFiles,,"%24%\pagefile.sys 27 77"
 HKLM,%Control%\Session Manager\Memory Management,WriteWatch,0x00040002,1
+;;KnownDLLs
+HKLM,%Control%\Session Manager\KnownDLLs,"_wow64cpu",,"wow64cpu.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"_wowarmhw",,"wowarmhw.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"_xtajit",,"_xtajit.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"advapi32",,"advapi32.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"clbcatq",,"clbcatq.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"combase",,"combase.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"COMDLG32",,"COMDLG32.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"coml2",,"coml2.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"DifxApi",,"difxapi.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"gdi32",,"gdi32.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"gdiplus",,"gdiplus.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"IMAGEHLP",,"IMAGEHLP.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"IMM32",,"IMM32.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"kernel32",,"kernel32.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"MSCTF",,"MSCTF.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"MSVCRT",,"MSVCRT.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"NORMALIZ",,"NORMALIZ.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"NSI",,"NSI.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"ole32",,"ole32.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"OLEAUT32",,"OLEAUT32.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"PSAPI",,"PSAPI.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"rpcrt4",,"rpcrt4.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"sechost",,"sechost.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"Setupapi",,"Setupapi.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"SHCORE",,"SHCORE.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"SHELL32",,"SHELL32.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"SHLWAPI",,"SHLWAPI.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"user32",,"user32.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"wow64",,"wow64.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"wow64win",,"wow64win.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"WS2_32",,"WS2_32.dll"
+;;KnownDLLs not present in registry on Windows but present in \\KnownDLLs directory
+HKLM,%Control%\Session Manager\KnownDLLs,"ucrtbase",,"ucrtbase.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"msvcp_win",,"msvcp_win.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"bcrypt",,"bcrypt.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"COMCTL32",,"COMCTL32.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"cfgmgr32",,"cfgmgr32.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"ntdll",,"ntdll.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"bcryptPrimitives",,"bcryptPrimitives.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"win32u",,"win32u.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"gdi32full",,"gdi32full.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"WINTRUST",,"WINTRUST.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"CRYPT32",,"CRYPT32.dll"
+HKLM,%Control%\Session Manager\KnownDLLs,"WLDAP32",,"WLDAP32.dll"
 
 [Fonts]
 HKLM,%FontSubStr%,"Arial Baltic,186",,"Arial,186"


