From 3598e7d534049183c7c09467939b1b26d795dd33 Mon Sep 17 00:00:00 2001
From: Torge Matthies <tmatthies@codeweavers.com>
Date: Tue, 22 Nov 2022 18:18:21 +0100
Subject: [PATCH 10/11] user32: Remove NtUserDriverCallback* kernel callbacks.

Signed-off-by: Torge Matthies <tmatthies@codeweavers.com>
---
 dlls/wow64win/user.c | 61 --------------------------------------------
 include/ntuser.h     |  3 ---
 2 files changed, 64 deletions(-)

diff --git a/dlls/wow64win/user.c b/dlls/wow64win/user.c
index 11111111111..11111111111 100644
--- a/dlls/wow64win/user.c
+++ b/dlls/wow64win/user.c
@@ -1444,56 +1444,6 @@ static NTSTATUS WINAPI wow64_NtUserThunkLock( void *arg, ULONG size )
     return dispatch_callback( NtUserThunkLock, arg, size );
 }
 
-static NTSTATUS WINAPI wow64_NtUserDriverCallbackFirst0( void *arg, ULONG size )
-{
-    return dispatch_callback( NtUserDriverCallbackFirst + 0, arg, size );
-}
-
-static NTSTATUS WINAPI wow64_NtUserDriverCallbackFirst1( void *arg, ULONG size )
-{
-    return dispatch_callback( NtUserDriverCallbackFirst + 1, arg, size );
-}
-
-static NTSTATUS WINAPI wow64_NtUserDriverCallbackFirst2( void *arg, ULONG size )
-{
-    return dispatch_callback( NtUserDriverCallbackFirst + 2, arg, size );
-}
-
-static NTSTATUS WINAPI wow64_NtUserDriverCallbackFirst3( void *arg, ULONG size )
-{
-    return dispatch_callback( NtUserDriverCallbackFirst + 3, arg, size );
-}
-
-static NTSTATUS WINAPI wow64_NtUserDriverCallbackFirst4( void *arg, ULONG size )
-{
-    return dispatch_callback( NtUserDriverCallbackFirst + 4, arg, size );
-}
-
-static NTSTATUS WINAPI wow64_NtUserDriverCallbackFirst5( void *arg, ULONG size )
-{
-    return dispatch_callback( NtUserDriverCallbackFirst + 5, arg, size );
-}
-
-static NTSTATUS WINAPI wow64_NtUserDriverCallbackFirst6( void *arg, ULONG size )
-{
-    return dispatch_callback( NtUserDriverCallbackFirst + 6, arg, size );
-}
-
-static NTSTATUS WINAPI wow64_NtUserDriverCallbackFirst7( void *arg, ULONG size )
-{
-    return dispatch_callback( NtUserDriverCallbackFirst + 7, arg, size );
-}
-
-static NTSTATUS WINAPI wow64_NtUserDriverCallbackFirst8( void *arg, ULONG size )
-{
-    return dispatch_callback( NtUserDriverCallbackFirst + 8, arg, size );
-}
-
-static NTSTATUS WINAPI wow64_NtUserDriverCallbackFirst9( void *arg, ULONG size )
-{
-    return dispatch_callback( NtUserDriverCallbackFirst + 9, arg, size );
-}
-
 user_callback user_callbacks[] =
 {
     /* user32 callbacks */
@@ -1520,17 +1470,6 @@ user_callback user_callbacks[] =
     /* win16 hooks */
     wow64_NtUserCallFreeIcon,
     wow64_NtUserThunkLock,
-    /* Driver-specific callbacks */
-    wow64_NtUserDriverCallbackFirst0,
-    wow64_NtUserDriverCallbackFirst1,
-    wow64_NtUserDriverCallbackFirst2,
-    wow64_NtUserDriverCallbackFirst3,
-    wow64_NtUserDriverCallbackFirst4,
-    wow64_NtUserDriverCallbackFirst5,
-    wow64_NtUserDriverCallbackFirst6,
-    wow64_NtUserDriverCallbackFirst7,
-    wow64_NtUserDriverCallbackFirst8,
-    wow64_NtUserDriverCallbackFirst9,
 };
 
 C_ASSERT( ARRAYSIZE(user_callbacks) == NtUserCallCount );
diff --git a/include/ntuser.h b/include/ntuser.h
index 11111111111..11111111111 100644
--- a/include/ntuser.h
+++ b/include/ntuser.h
@@ -52,9 +52,6 @@ enum
     /* win16 hooks */
     NtUserCallFreeIcon,
     NtUserThunkLock,
-    /* Driver-specific callbacks */
-    NtUserDriverCallbackFirst,
-    NtUserDriverCallbackLast = NtUserDriverCallbackFirst + 9,
     NtUserCallCount
 };
 
-- 
0.0.0

