From 5c6859860b4121848546dd1e86123a330195b178 Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Sat, 26 Nov 2022 21:24:03 +0100
Subject: [PATCH 1/2] kernelbase: Optimize free_sortkey_state.

---
 dlls/kernelbase/locale.c | 22 ++++++++++++++--------
 1 file changed, 14 insertions(+), 8 deletions(-)

diff --git a/dlls/kernelbase/locale.c b/dlls/kernelbase/locale.c
index 11111111111..11111111111 100644
--- a/dlls/kernelbase/locale.c
+++ b/dlls/kernelbase/locale.c
@@ -3640,16 +3640,22 @@ static BOOL remove_unneeded_weights( const struct sortguid *sortid, struct sortk
     return TRUE;
 }
 
+static inline void free_buff( void *buf )
+{
+    if (buf)
+        RtlFreeHeap( GetProcessHeap(), 0, buf );
+}
+
 static void free_sortkey_state( struct sortkey_state *s )
 {
-    RtlFreeHeap( GetProcessHeap(), 0, s->key_primary.new_buf );
-    RtlFreeHeap( GetProcessHeap(), 0, s->key_diacritic.new_buf );
-    RtlFreeHeap( GetProcessHeap(), 0, s->key_case.new_buf );
-    RtlFreeHeap( GetProcessHeap(), 0, s->key_special.new_buf );
-    RtlFreeHeap( GetProcessHeap(), 0, s->key_extra[0].new_buf );
-    RtlFreeHeap( GetProcessHeap(), 0, s->key_extra[1].new_buf );
-    RtlFreeHeap( GetProcessHeap(), 0, s->key_extra[2].new_buf );
-    RtlFreeHeap( GetProcessHeap(), 0, s->key_extra[3].new_buf );
+    free_buff( s->key_primary.new_buf );
+    free_buff( s->key_diacritic.new_buf );
+    free_buff( s->key_case.new_buf );
+    free_buff( s->key_special.new_buf );
+    free_buff( s->key_extra[0].new_buf );
+    free_buff( s->key_extra[1].new_buf );
+    free_buff( s->key_extra[2].new_buf );
+    free_buff( s->key_extra[3].new_buf );
 }
 
 static int append_weights( const struct sortguid *sortid, DWORD flags,
-- 
2.40.0

