From 88713da4cfcb91dc04c8dd4b3d42aec3dd2188db Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Thu, 14 Apr 2022 02:56:03 +0200
Subject: [PATCH] winepulse.drv: Don't add a period to real_bufsize_bytes.

---
 dlls/winepulse.drv/pulse.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/winepulse.drv/pulse.c b/dlls/winepulse.drv/pulse.c
index 11111111111..11111111111 100644
--- a/dlls/winepulse.drv/pulse.c
+++ b/dlls/winepulse.drv/pulse.c
@@ -1335,7 +1335,7 @@ static NTSTATUS pulse_create_stream(void *args)
         dump_attr(attr);
         if (stream->dataflow == eRender) {
             size = stream->real_bufsize_bytes =
-                stream->bufsize_frames * 2 * pa_frame_size(&stream->ss);
+                stream->bufsize_frames * pa_frame_size(&stream->ss);
             if (NtAllocateVirtualMemory(GetCurrentProcess(), (void **)&stream->local_buffer,
                                         zero_bits(), &size, MEM_COMMIT, PAGE_READWRITE))
                 hr = E_OUTOFMEMORY;
-- 
2.37.2

