From 8d5a16dd8df4eee4739ce019f6e9899fbff411a1 Mon Sep 17 00:00:00 2001
From: li wenzhe <584592187@qq.com>
Date: Thu, 9 Mar 2023 10:27:42 +0000
Subject: [PATCH] Update dlls/ntdll/wcstring.c

---
 dlls/ntdll/wcstring.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/dlls/ntdll/wcstring.c b/dlls/ntdll/wcstring.c
index 11111111111..11111111111 100644
--- a/dlls/ntdll/wcstring.c
+++ b/dlls/ntdll/wcstring.c
@@ -353,7 +353,12 @@ int __cdecl wcsncmp( LPCWSTR str1, LPCWSTR str2, size_t n )
 LPWSTR __cdecl wcsncpy( LPWSTR s1, LPCWSTR s2, size_t n )
 {
     WCHAR *ret = s1;
-    for ( ; n; n--) if (!(*s1++ = *s2++)) break;
+    for ( ; n; n--) {
+        if (!(*s1++ = *s2++)) {
+            n--;
+            break;
+        }
+    }
     for ( ; n; n--) *s1++ = 0;
     return ret;
 }
-- 
2.40.0

