From de08512c9c50983279d1979542c51db1bf031bda Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Sat, 26 Nov 2022 21:44:45 +0100
Subject: [PATCH 2/2] kernelbase: Increase buffer size in compare_string.

---
 dlls/kernelbase/locale.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/kernelbase/locale.c b/dlls/kernelbase/locale.c
index 11111111111..11111111111 100644
--- a/dlls/kernelbase/locale.c
+++ b/dlls/kernelbase/locale.c
@@ -3829,8 +3829,8 @@ static int compare_string( const struct sortguid *sortid, DWORD flags,
 {
     struct sortkey_state s1;
     struct sortkey_state s2;
-    BYTE primary1[32];
-    BYTE primary2[32];
+    BYTE primary1[256];
+    BYTE primary2[256];
     int i, ret, len, pos1 = 0, pos2 = 0;
     BOOL have_extra1, have_extra2;
     BYTE case_mask = 0x3f;
-- 
2.39.0

