From 7e14c3d0957b5126c03df58fe0e1e8a20dd2254c Mon Sep 17 00:00:00 2001
From: Torge Matthies <tmatthies@codeweavers.com>
Date: Wed, 14 Dec 2022 09:04:48 +0100
Subject: [PATCH 2/3] include: Add _AddressOfReturnAddress intrinsic.

Signed-off-by: Torge Matthies <tmatthies@codeweavers.com>
---
 include/winnt.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/include/winnt.h b/include/winnt.h
index 11111111111..11111111111 100644
--- a/include/winnt.h
+++ b/include/winnt.h
@@ -6582,6 +6582,7 @@ typedef enum _FIRMWARE_TYPE
 #pragma intrinsic(_InterlockedOr)
 #pragma intrinsic(_InterlockedXor)
 #pragma intrinsic(_ReturnAddress)
+#pragma intrinsic(_AddressOfReturnAddress)
 #pragma intrinsic(__fastfail)
 
 BOOLEAN   _BitScanForward(unsigned long*,unsigned long);
@@ -6598,6 +6599,7 @@ short     _InterlockedIncrement16(short volatile*);
 long      _InterlockedOr(long volatile *,long);
 long      _InterlockedXor(long volatile *,long);
 void     *_ReturnAddress(void);
+void     *_AddressOfReturnAddress(void);
 DECLSPEC_NORETURN void __fastfail(unsigned int);
 
 #ifndef __i386__
@@ -6956,6 +6958,11 @@ static FORCEINLINE void *_ReturnAddress(void)
     return __builtin_extract_return_addr(__builtin_return_address(0));
 }
 
+static FORCEINLINE void *_AddressOfReturnAddress(void)
+{
+    return ((void*)((void**)__builtin_frame_address(0) + 1));
+}
+
 static FORCEINLINE DECLSPEC_NORETURN void __fastfail(unsigned int code)
 {
 #if defined(__x86_64__) || defined(__i386__)
-- 
2.40.0

