From 903d5e37b4fe509f3fd79155bba273187aaaf04b Mon Sep 17 00:00:00 2001
From: Adam Rehn <adam@tensorworks.com.au>
Date: Wed, 20 Sep 2023 21:00:30 +1000
Subject: [PATCH] wineserver: Report non-zero exit code for abnormal process
 termination

Signed-off-by: Adam Rehn <adam@tensorworks.com.au>
---
 server/process.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/server/process.c b/server/process.c
index 11111111111..11111111111 100644
--- a/server/process.c
+++ b/server/process.c
@@ -930,7 +930,7 @@ static void process_poll_event( struct fd *fd, int event )
     struct process *process = get_fd_user( fd );
     assert( process->obj.ops == &process_ops );
 
-    if (event & (POLLERR | POLLHUP)) kill_process( process, 0 );
+    if (event & (POLLERR | POLLHUP)) kill_process( process, !process->is_terminating );
     else if (event & POLLIN) receive_fd( process );
 }
 
-- 
0.0.0

