From f0f8924a99b1472f2ea355342d55e4523c09480b Mon Sep 17 00:00:00 2001
From: Jinoh Kang <jinoh.kang.kr@gmail.com>
Date: Mon, 18 Apr 2022 19:38:25 +0900
Subject: [PATCH 5/6] win32u/tests: Add tests for NtUserSetObjectInformation.

Signed-off-by: Jinoh Kang <jinoh.kang.kr@gmail.com>
---
 dlls/win32u/tests/win32u.c | 35 +++++++++++++++++++++++++++++++++++
 1 file changed, 35 insertions(+)

diff --git a/dlls/win32u/tests/win32u.c b/dlls/win32u/tests/win32u.c
index 11111111111..11111111111 100644
--- a/dlls/win32u/tests/win32u.c
+++ b/dlls/win32u/tests/win32u.c
@@ -2023,6 +2023,39 @@ static void test_wndproc_hook(void)
     UnregisterClassW( L"TestLParamClass", NULL );
 }
 
+static void test_NtUserSetObjectInformation(void)
+{
+    SetLastError( 0xdeadbeef );
+    NtUserSetObjectInformation( GetProcessWindowStation(), UOI_FLAGS, (void *)NULL, sizeof(USEROBJECTFLAGS) );
+    todo_wine
+    ok( GetLastError() == ERROR_NOACCESS, "NtUserSetObjectInformation error %lu\n", GetLastError() );
+
+    SetLastError( 0xdeadbeef );
+    NtUserSetObjectInformation( GetThreadDesktop(GetCurrentThreadId()), UOI_FLAGS, (void *)NULL, sizeof(USEROBJECTFLAGS) );
+    todo_wine
+    ok( GetLastError() == ERROR_NOACCESS, "NtUserSetObjectInformation error %lu\n", GetLastError() );
+
+    SetLastError( 0xdeadbeef );
+    NtUserSetObjectInformation( GetCurrentProcess(), UOI_TIMERPROC_EXCEPTION_SUPPRESSION, (void *)TRUE, 0 );
+    todo_wine
+    ok( GetLastError() == 0xdeadbeef, "NtUserSetObjectInformation error %lu\n", GetLastError() );
+
+    SetLastError( 0xdeadbeef );
+    NtUserSetObjectInformation( GetCurrentProcess(), UOI_TIMERPROC_EXCEPTION_SUPPRESSION, (void *)0xdeadbeef, 0xdeadbeef );
+    todo_wine
+    ok( GetLastError() == 0xdeadbeef, "NtUserSetObjectInformation error %lu\n", GetLastError() );
+
+    SetLastError( 0xdeadbeef );
+    NtUserSetObjectInformation( GetCurrentProcess(), UOI_TIMERPROC_EXCEPTION_SUPPRESSION, (void *)-1, -1 );
+    todo_wine
+    ok( GetLastError() == 0xdeadbeef, "NtUserSetObjectInformation error %lu\n", GetLastError() );
+
+    SetLastError( 0xdeadbeef );
+    NtUserSetObjectInformation( NULL, UOI_TIMERPROC_EXCEPTION_SUPPRESSION, (void *)-1, -1 );
+    todo_wine
+    ok( GetLastError() == 0xdeadbeef, "NtUserSetObjectInformation error %lu\n", GetLastError() );
+}
+
 START_TEST(win32u)
 {
     char **argv;
@@ -2066,4 +2099,6 @@ START_TEST(win32u)
 
     test_NtUserEnableMouseInPointer( argv, FALSE );
     test_NtUserEnableMouseInPointer( argv, TRUE );
+
+    test_NtUserSetObjectInformation();
 }


