From 8af68e03ee61164e9472ee376d124c6ef03623ec Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Thu, 28 Jul 2022 11:47:03 +0200
Subject: [PATCH 10/11] wineboot: Relax TSC frequency calibration error to 500ppm.

---
 programs/wineboot/wineboot.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/programs/wineboot/wineboot.c b/programs/wineboot/wineboot.c
index 11111111111..11111111111 100644
--- a/programs/wineboot/wineboot.c
+++ b/programs/wineboot/wineboot.c
@@ -317,7 +317,7 @@ static UINT64 read_tsc_frequency( BOOL has_rdtscp )
             freq1 = (tsc3 - tsc1) * 10000000 / (time1 - time0);
             error = llabs( (freq1 - freq0) * 1000000 / min( freq1, freq0 ) );
         }
-        while (error > 100 && --retries);
+        while (error > 500 && --retries);
 
         if (!retries) WARN( "TSC frequency calibration failed, unstable TSC?\n" );
         else
