From fc832b136e39b17391dced0b78141f1e864862ff Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Sun, 14 Nov 2021 14:38:45 +0100
Subject: [PATCH 4/9] winex11.drv: Reset XInput2 valuators only on DeviceChange
 events.

---
 dlls/winex11.drv/mouse.c | 5 -----
 1 file changed, 5 deletions(-)

diff --git a/dlls/winex11.drv/mouse.c b/dlls/winex11.drv/mouse.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/mouse.c
+++ b/dlls/winex11.drv/mouse.c
@@ -353,11 +353,6 @@ static void disable_xinput2(void)
     mask.deviceid = XIAllMasterDevices;
 
     pXISelectEvents( data->display, DefaultRootWindow( data->display ), &mask, 1 );
-
-    data->x_valuator.number = -1;
-    data->y_valuator.number = -1;
-    data->x_valuator.value = 0;
-    data->y_valuator.value = 0;
 #endif
 }
 
-- 
2.36.0

